From f2391b9da6127e4acd5b54b7ae6c2d585df0e2a0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Joaqu=C3=ADn=20Cuenca=20Abela?= <e98cuenc@gmail.com>
Date: Wed, 29 May 2013 16:22:09 +0200
Subject: [PATCH 02/22] Don't notify continuously when MySQL slave SQL thread
 is running

Signed-off-by: Florian Forster <octo@collectd.org>
---
 src/mysql.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/mysql.c b/src/mysql.c
index 8b3cd21..f93a442 100644
--- a/src/mysql.c
+++ b/src/mysql.c
@@ -689,7 +689,7 @@ static int mysql_read_slave_stats (mysql_database_t *db, MYSQL *con)
 			ssnprintf (n.message, sizeof (n.message),
 					"slave SQL thread started");
 			plugin_dispatch_notification (&n);
-			db->slave_sql_running = 0;
+			db->slave_sql_running = 1;
 		}
 	}
 
-- 
1.9.3

