From 6bd48f2346c5072dc22da58c7b7cd8e8ceb83fc5 Mon Sep 17 00:00:00 2001
From: Marc Fournier <marc.fournier@camptocamp.com>
Date: Sat, 29 Mar 2014 13:37:36 +0100
Subject: [PATCH 10/22] indent wh_init() to be consistent with the rest of the
 file

---
 src/write_http.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/write_http.c b/src/write_http.c
index 04c637b..0a13444 100644
--- a/src/write_http.c
+++ b/src/write_http.c
@@ -658,10 +658,10 @@ static int wh_config (oconfig_item_t *ci) /* {{{ */
 
 static int wh_init (void) /* {{{ */
 {
-  /* Call this while collectd is still single-threaded to avoid
-   * initialization issues in libgcrypt. */
-  curl_global_init (CURL_GLOBAL_SSL);
-  return (0);
+        /* Call this while collectd is still single-threaded to avoid
+         * initialization issues in libgcrypt. */
+        curl_global_init (CURL_GLOBAL_SSL);
+        return (0);
 } /* }}} int wh_init */
 
 void module_register (void) /* {{{ */
-- 
1.9.3

