From 5dbc1cfb4baff831f950b3c9de8c332bc9aa2b6e Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Wed, 17 Sep 2014 11:25:17 +0200
Subject: [PATCH 18/22] openvpn: Remove boguous file handler check

---
 src/openvpn.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/src/openvpn.c b/src/openvpn.c
index 7d4e4a0..9598abc 100644
--- a/src/openvpn.c
+++ b/src/openvpn.c
@@ -316,9 +316,6 @@ static int multi1_read (char *name, FILE *fh)
 		}
 	}
 
-	if (ferror (fh))
-		return (0);
-
 	if (collect_user_count)
 		numusers_submit(name, name, sum_users);
 
-- 
1.9.3

