From 41253b68808deade2a1866f0c24f4bbc029a92c2 Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Wed, 24 Sep 2014 14:38:04 +0200
Subject: [PATCH 21/22] openvpn: Fix copy-and-paste error

---
 src/openvpn.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/openvpn.c b/src/openvpn.c
index f686721..2db3677 100644
--- a/src/openvpn.c
+++ b/src/openvpn.c
@@ -293,7 +293,7 @@ static int multi1_read (char *name, FILE *fh)
 			continue;
 
 		// Ignore not yet fully established connections
-		if (strcmp(fields[1], "UNDEF") == 0)
+		if (strcmp(fields[0], "UNDEF") == 0)
 			continue;
 
 		if (collect_user_count)
-- 
1.9.3

