diff -Naur gcc-13.1.0.org/gcc/config/riscv/genrvv-type-indexer.cc gcc-13.1.0/gcc/config/riscv/genrvv-type-indexer.cc
--- gcc-13.1.0.org/gcc/config/riscv/genrvv-type-indexer.cc	2023-04-26 09:09:39.000000000 +0200
+++ gcc-13.1.0/gcc/config/riscv/genrvv-type-indexer.cc	2023-08-02 11:28:53.615913583 +0200
@@ -115,9 +115,9 @@
   if (sew == eew)
     elmul_log2 = lmul_log2;
   else if (sew > eew)
-    elmul_log2 = lmul_log2 - std::log2 (sew / eew);
+    elmul_log2 = lmul_log2 - log2 (sew / eew);
   else /* sew < eew */
-    elmul_log2 = lmul_log2 + std::log2 (eew / sew);
+    elmul_log2 = lmul_log2 + log2 (eew / sew);
 
   if (float_p)
     return floattype (eew, elmul_log2);
