diff -Naur iperf-3.9.org/src/Makefile.am iperf-3.9/src/Makefile.am
--- iperf-3.9.org/src/Makefile.am	2020-08-15 00:08:27.000000000 +0200
+++ iperf-3.9/src/Makefile.am	2020-12-22 22:16:14.488228000 +0100
@@ -58,9 +58,9 @@
 iperf3_profile_SOURCES  = main.c \
                           $(libiperf_la_SOURCES)
 
-iperf3_profile_CFLAGS   = -pg -g
+iperf3_profile_CFLAGS   = -g
 iperf3_profile_LDADD    = libiperf.la
-iperf3_profile_LDFLAGS  = -pg -g
+iperf3_profile_LDFLAGS  = -g
 endif
 
 # Specify the sources and various flags for the test cases
diff -Naur iperf-3.9.org/src/Makefile.in iperf-3.9/src/Makefile.in
--- iperf-3.9.org/src/Makefile.in	2020-08-15 00:08:27.000000000 +0200
+++ iperf-3.9/src/Makefile.in	2020-12-22 22:16:41.934746000 +0100
@@ -680,9 +680,9 @@
 @ENABLE_PROFILING_TRUE@iperf3_profile_SOURCES = main.c \
 @ENABLE_PROFILING_TRUE@                          $(libiperf_la_SOURCES)
 
-@ENABLE_PROFILING_TRUE@iperf3_profile_CFLAGS = -pg -g
+@ENABLE_PROFILING_TRUE@iperf3_profile_CFLAGS = -g
 @ENABLE_PROFILING_TRUE@iperf3_profile_LDADD = libiperf.la
-@ENABLE_PROFILING_TRUE@iperf3_profile_LDFLAGS = -pg -g
+@ENABLE_PROFILING_TRUE@iperf3_profile_LDFLAGS = -g
 
 # Specify the sources and various flags for the test cases
 t_timer_SOURCES = t_timer.c
