--- sarg-2.3.11/configure.ac~	2019-04-26 22:34:33.499022406 +0000
+++ sarg-2.3.11/configure.ac	2019-04-26 22:35:11.886556020 +0000
@@ -29,35 +29,6 @@
 # Report more warnings to improve code quality.
 CFLAGS="${CFLAGS} -Wall -Wno-sign-compare"
 
-dnl Check for supported compiler options
-
-AC_MSG_CHECKING([for extra warnings flag in $CC])
-saved_CFLAGS="${CFLAGS}"
-CFLAGS="${CFLAGS} -Wextra -Wno-unused-parameter"
-AC_COMPILE_IFELSE([AC_LANG_SOURCE([])],[have_extra_warnings="yes"],[have_extra_warnings="no"])
-AC_MSG_RESULT($have_extra_warnings)
-if test "$have_extra_warnings" = "no" ; then
-    CFLAGS="${saved_CFLAGS}"
-fi
-
-AC_MSG_CHECKING([for implicit-function-declaration error flag in $CC])
-saved_CFLAGS="${CFLAGS}"
-CFLAGS="${CFLAGS} -Werror=implicit-function-declaration"
-AC_COMPILE_IFELSE([AC_LANG_SOURCE([])],[have_implicit_function_declaration="yes"],[have_implicit_function_declaration="no"])
-AC_MSG_RESULT($have_implicit_function_declaration)
-if test "$have_implicit_function_declaration" = "no" ; then
-    CFLAGS="${saved_CFLAGS}"
-fi
-
-AC_MSG_CHECKING([for format error flag in $CC])
-saved_CFLAGS="${CFLAGS}"
-CFLAGS="${CFLAGS} -Werror=format"
-AC_COMPILE_IFELSE([AC_LANG_SOURCE([])],[have_error_format="yes"],[have_error_format="no"])
-AC_MSG_RESULT($have_error_format)
-if test "$have_error_format" = "no" ; then
-    CFLAGS="${saved_CFLAGS}"
-fi
-
 case "$host" in
    *-solaris*)
    LDFLAGS="${LDFLAGS} -lsocket -lnsl"
