diff -Naur mdadm-4.0.org/Makefile mdadm-4.0/Makefile
--- mdadm-4.0.org/Makefile	2017-01-09 22:46:34.000000000 +0100
+++ mdadm-4.0/Makefile	2018-01-19 11:18:42.868219725 +0100
@@ -43,7 +43,7 @@
 
 CC ?= $(CROSS_COMPILE)gcc
 CXFLAGS ?= -ggdb
-CWFLAGS = -Wall -Werror -Wstrict-prototypes -Wextra -Wno-unused-parameter
+CWFLAGS = -Wall -Werror -Werror=implicit-fallthrough=0 -Werror=format-overflow=0 -Werror=format-truncation=0 -Wstrict-prototypes -Wextra -Wno-unused-parameter
 ifdef WARN_UNUSED
 CWFLAGS += -Wp,-D_FORTIFY_SOURCE=2 -O3
 endif
