From a5b116fe3107a56e1d881906e77d9731b0c6b2c2 Mon Sep 17 00:00:00 2001
From: Michael Adam <obnox@samba.org>
Date: Sat, 1 Jun 2013 02:14:41 +0200
Subject: [PATCH] shadow_copy2: implement disk_free

Signed-off-by: Michael Adam <obnox@samba.org>
---
 source3/modules/vfs_shadow_copy2.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/source3/modules/vfs_shadow_copy2.c b/source3/modules/vfs_shadow_copy2.c
index fedfb53..7fd4dd5 100644
--- a/source3/modules/vfs_shadow_copy2.c
+++ b/source3/modules/vfs_shadow_copy2.c
@@ -944,6 +944,16 @@ static int shadow_copy2_get_shadow_copy2_data(vfs_handle_struct *handle,
 	return 0;
 }
 
+static uint64_t shadow_copy2_disk_free(vfs_handle_struct *handle,
+				       const char *fname, bool small_query,
+				       uint64_t *bsize, uint64_t *dfree,
+				       uint64_t *dsize)
+{
+	SHADOW2_NEXT(DISK_FREE,
+		     (handle, name, small_query, bsize, dfree, dsize),
+		     uint64_t, 0);
+}
+
 static struct vfs_fn_pointers vfs_shadow_copy2_fns = {
         .opendir = shadow_copy2_opendir,
         .mkdir = shadow_copy2_mkdir,
@@ -975,6 +985,7 @@ static struct vfs_fn_pointers vfs_shadow_copy2_fns = {
         .get_nt_acl = shadow_copy2_get_nt_acl,
         .chmod_acl = shadow_copy2_chmod_acl,
 	.get_shadow_copy_data = shadow_copy2_get_shadow_copy2_data,
+	.disk_free = shadow_copy2_disk_free,
 };
 
 NTSTATUS vfs_shadow_copy2_init(void);
-- 
2.1.0

