Index: samba-3.6.22/source3/winbindd/winbindd_group.c
===================================================================
--- samba-3.6.22.orig/source3/winbindd/winbindd_group.c
+++ samba-3.6.22/source3/winbindd/winbindd_group.c
@@ -69,7 +69,7 @@ bool fill_grent(TALLOC_CTX *mem_ctx, str
 	/* Group name and password */
 
 	safe_strcpy(gr->gr_name, full_group_name, sizeof(gr->gr_name) - 1);
-	safe_strcpy(gr->gr_passwd, "x", sizeof(gr->gr_passwd) - 1);
+	safe_strcpy(gr->gr_passwd, "*", sizeof(gr->gr_passwd) - 1);
 
 	return True;
 }
