Index: samba-3.6.6/source3/winbindd/idmap.c
===================================================================
--- samba-3.6.6.orig/source3/winbindd/idmap.c
+++ samba-3.6.6/source3/winbindd/idmap.c
@@ -129,7 +129,7 @@ NTSTATUS smb_register_idmap(int version,
 
 	for (entry = backends; entry != NULL; entry = entry->next) {
 		if (strequal(entry->name, name)) {
-			DEBUG(0,("Idmap module %s already registered!\n",
+			DEBUG(5,("Idmap module %s already registered!\n",
 				 name));
 			return NT_STATUS_OBJECT_NAME_COLLISION;
 		}
Index: samba-3.6.6/source3/winbindd/nss_info.c
===================================================================
--- samba-3.6.6.orig/source3/winbindd/nss_info.c
+++ samba-3.6.6/source3/winbindd/nss_info.c
@@ -66,7 +66,7 @@ static struct nss_function_entry *nss_ge
 	}
 
 	if ( nss_get_backend(name) ) {
-		DEBUG(0,("smb_register_idmap_nss: idmap module %s "
+		DEBUG(5,("smb_register_idmap_nss: idmap module %s "
 			 "already registered!\n", name));
 		return NT_STATUS_OBJECT_NAME_COLLISION;
 	}
