From: Yi Zou <yi.zou@intel.com>
Subject: [FcOE] change fcoe_sw sg_tablesize to SG_ALL
References: bnc #459142

Signed-off-by: Yi Zou <yi.zou@intel.com>
Acked-by: Bernhard Walle <bwalle@suse.de>
---

 drivers/scsi/fcoe/fcoe_sw.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)


--- a/drivers/scsi/fcoe/fcoe_sw.c
+++ b/drivers/scsi/fcoe/fcoe_sw.c
@@ -100,7 +100,7 @@ static struct scsi_host_template fcoe_sw
 	.cmd_per_lun = 32,
 	.can_queue = FCOE_MAX_OUTSTANDING_COMMANDS,
 	.use_clustering = ENABLE_CLUSTERING,
-	.sg_tablesize = 4,
+	.sg_tablesize = SG_ALL,
 	.max_sectors = 0xffff,
 };
 
