From: Yi Zou <yi.zou@intel.com>
Subject: [FcOE] check return for fc_set_mfs
References: bnc #459142

Signed-off-by: Yi Zou <yi.zou@intel.com>
Acked-by: Bernhard Walle <bwalle@suse.de>
---

 drivers/scsi/fcoe/fcoe_sw.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)


--- a/drivers/scsi/fcoe/fcoe_sw.c
+++ b/drivers/scsi/fcoe/fcoe_sw.c
@@ -178,7 +178,8 @@ static int fcoe_sw_netdev_config(struct
 	 */
 	mfs = fc->real_dev->mtu - (sizeof(struct fcoe_hdr) +
 				   sizeof(struct fcoe_crc_eof));
-	fc_set_mfs(lp, mfs);
+	if (fc_set_mfs(lp, mfs))
+		return -EINVAL;
 
 	lp->link_status = ~FC_PAUSE & ~FC_LINK_UP;
 	if (!fcoe_link_ok(lp))
