diff -up wrk/genisoimage/exclude.c.wrk wrk/genisoimage/exclude.c
--- wrk/genisoimage/exclude.c.wrk	2015-02-25 13:16:28.054237196 +0100
+++ wrk/genisoimage/exclude.c	2015-02-25 13:33:06.761312002 +0100
@@ -39,7 +39,7 @@ exclude(char *fn)
 {
 	register int	i;
 
-	for (i = 0; excl[i] && i < MAXEXCL; i++)
+	for (i = 0; i < MAXEXCL && excl[i]; i++)
 		;
 
 	if (i == MAXEXCL) {
@@ -69,7 +69,7 @@ is_excluded(char *fn)
 	/*
 	 * very dumb search method ...
 	 */
-	for (i = 0; excl[i] && i < MAXEXCL; i++) {
+	for (i = 0; i < MAXEXCL && excl[i]; i++) {
 		if (strcmp(excl[i], fn) == 0) {
 			return (1);	/* found -> excluded filenmae */
 		}
diff -up wrk/icedax/toc.c.wrk wrk/icedax/toc.c
