diff -up cdrkit-1.1.11/icedax/icedax.c.handler cdrkit-1.1.11/icedax/icedax.c
--- cdrkit-1.1.11/icedax/icedax.c.handler	2011-06-02 16:37:02.446206285 +0200
+++ cdrkit-1.1.11/icedax/icedax.c	2011-06-02 16:37:25.444206279 +0200
@@ -1054,7 +1054,7 @@ static void exit_wrapper(int status)
 
 	if (child_pid != 0) {
 		SCSI *usalp = get_scsi_p();
-		if (usalp->running) {
+		if (usalp && usalp->running) {
 			usalp->cb_fun = on_exitscsi;
 			usalp->cb_arg = (void *) (uintptr_t) status;
 		} else {
