diff -up wrk/libusal/scsi-linux-sg.c.wrk wrk/libusal/scsi-linux-sg.c
--- wrk/libusal/scsi-linux-sg.c.wrk	2015-05-29 13:30:41.088896572 +0200
+++ wrk/libusal/scsi-linux-sg.c	2015-05-29 13:38:25.776303388 +0200
@@ -509,7 +509,9 @@ usalo_open(SCSI *usalp, char *device)
 		/* scan and maybe keep one open, sg_setup decides */
 #define HDX 0
 #define SCD 1
-#define SG 2
+/* second scd option included because of Fedora naming convention /dev/srN */
+#define SCD2 2
+#define SG 3
 		int h;
 /*
 retry_scan_open:
@@ -533,6 +535,15 @@ retry_scan_open:
 						first=0;
 						last=255;
 						break;
+					}
+				case(SCD2):
+					{
+						if(!check_linux_26())
+							continue;
+						pattern="/dev/sr%d";
+						first=0;
+						last=255;
+						break;
 					}
 				case(SG):
 					{
