diff -up cdrkit-1.1.11/librols/CMakeLists.txt.usalinst cdrkit-1.1.11/librols/CMakeLists.txt
--- cdrkit-1.1.11/librols/CMakeLists.txt.usalinst	2012-07-31 10:51:30.270669381 +0200
+++ cdrkit-1.1.11/librols/CMakeLists.txt	2012-07-31 10:52:02.365378257 +0200
@@ -3,5 +3,11 @@ INCLUDE_DIRECTORIES(../include ../libhfs
 
 SET(LIBROLS_SRCS  astoi.c astoll.c astoull.c breakline.c cmpbytes.c comerr.c error.c fexec.c fillbytes.c getargs.c getav0.c getdomainname.c geterrno.c gethostid.c gethostname.c getpagesize.c handlecond.c movebytes.c raisecond.c rename.c saveargs.c seterrno.c spawn.c strcatl.c strdup.c streql.c swabbytes.c usleep.c stdio/cvmod.c stdio/dat.c stdio/fcons.c stdio/fdown.c stdio/fdup.c stdio/ffileread.c stdio/ffilewrite.c stdio/fgetline.c stdio/file_raise.c stdio/fileclose.c stdio/fileluopen.c stdio/filemopen.c stdio/fileopen.c  stdio/filepos.c stdio/fileread.c stdio/filereopen.c stdio/fileseek.c stdio/filesize.c stdio/filestat.c stdio/filewrite.c stdio/flag.c stdio/flush.c stdio/fpipe.c stdio/niread.c stdio/niwrite.c stdio/nixread.c stdio/nixwrite.c stdio/openfd.c stdio/peekc.c default.c)
 
-ADD_LIBRARY (rols STATIC ${CMAKE_CURRENT_SOURCE_DIR}/../include/xconfig.h  ${CMAKE_CURRENT_SOURCE_DIR}/../include/align.h ${LIBROLS_SRCS})
+ADD_LIBRARY (rols SHARED ${CMAKE_CURRENT_SOURCE_DIR}/../include/xconfig.h  ${CMAKE_CURRENT_SOURCE_DIR}/../include/align.h ${LIBROLS_SRCS})
+SET_TARGET_PROPERTIES(rols PROPERTIES
+    VERSION 0.0.1
+    SOVERSION 0
+)
+
+INSTALL(TARGETS rols DESTINATION lib${LIB_SUFFIX})
 
diff -up cdrkit-1.1.11/libusal/CMakeLists.txt.usalinst cdrkit-1.1.11/libusal/CMakeLists.txt
--- cdrkit-1.1.11/libusal/CMakeLists.txt.usalinst	2012-07-31 10:51:30.302670085 +0200
+++ cdrkit-1.1.11/libusal/CMakeLists.txt	2012-07-31 10:52:09.190528942 +0200
@@ -6,5 +6,12 @@ ADD_DEFINITIONS(-DUSE_RCMD_RSH)
 #SET(LIBSCG_SRCS rdummy.c usalsettarget.c usaltimes.c scsi-linux-ata.c scsi-linux-pg.c scsi-linux-sg.c scsierrs.c scsihack.c scsihelp.c scsiopen.c scsitransp.c)
 SET(LIBSCG_SRCS usalsettarget.c usaltimes.c scsierrs.c scsihack.c scsihelp.c scsiopen.c scsitransp.c scsi-remote.c)
 LINK_DIRECTORIES(../librols)
-ADD_LIBRARY (usal STATIC ${LIBSCG_SRCS})
+ADD_LIBRARY (usal SHARED ${LIBSCG_SRCS})
+
 TARGET_LINK_LIBRARIES(usal ${SCG_SELF_LIBS})
+SET_TARGET_PROPERTIES(usal PROPERTIES
+    VERSION 0.0.1
+    SOVERSION 0
+)
+INSTALL(TARGETS usal DESTINATION lib${LIB_SUFFIX})
+INSTALL(FILES usal/scsicdb.h usal/aspi-dos.h usal/usalcmd.h usal/usalio.h usal/scsireg.h usal/usalops.h usal/scsidefs.h usal/spti-wnt.h usal/srb_os2.h usal/scsitransp.h usal/aspi-win32.h usal/scsisense.h usaltimes.h pg.h DESTINATION include/usal)
