diff -up cdrkit-1.1.11/readom/readom.c.readsegfault cdrkit-1.1.11/readom/readom.c
--- cdrkit-1.1.11/readom/readom.c.readsegfault	2011-03-07 15:51:46.889441591 +0100
+++ cdrkit-1.1.11/readom/readom.c	2011-03-07 15:51:50.702938263 +0100
@@ -906,7 +906,7 @@ read_ftoc(SCSI *usalp, parm_t *parmp, BO
 
 
 	strcpy(filename, "toc.dat");
-	if (strcmp(parmp->name, "/dev/null") != 0) {
+	if (parmp != NULL && strcmp(parmp->name, "/dev/null") != 0) {
 
 		len = strlen(parmp->name);
 		if (len > (sizeof (filename)-5)) {
