diff -up cdrkit-1.1.10/genisoimage/genisoimage.c.build-fix cdrkit-1.1.10/genisoimage/genisoimage.c
--- cdrkit-1.1.10/genisoimage/genisoimage.c.build-fix	2010-01-20 11:05:05.000000000 +0100
+++ cdrkit-1.1.10/genisoimage/genisoimage.c	2010-01-20 11:21:16.144098749 +0100
@@ -47,6 +47,7 @@
 
 #include <mconfig.h>
 #include "genisoimage.h"
+#include "checksum.h"
 #include "iso9660.h"
 #include <errno.h>
 #include <timedefs.h>
