Subject: enable extended dlm debugging
From: Lars Marovsky-Bree <lmb@suse.de>

  This patch enable extended dlm debugging

Signed-off-by: Lars Marovsky-Bree <lmb@suse.de>
---
Index: fs/dlm/Makefile
===================================================================
--- a/fs/dlm/Makefile
+++ b/fs/dlm/Makefile
@@ -16,6 +16,6 @@ dlm-y :=			ast.o \
 				recoverd.o \
 				requestqueue.o \
 				user.o \
-				util.o 
-dlm-$(CONFIG_DLM_DEBUG) +=	debug_fs.o
+				util.o \
+				debug_fs.o
 
Index: fs/dlm/dlm_internal.h
===================================================================
--- a/fs/dlm/dlm_internal.h
+++ b/fs/dlm/dlm_internal.h
@@ -587,6 +587,7 @@ void dlm_timeout_warn(struct dlm_lkb *lk
 int dlm_plock_init(void);
 void dlm_plock_exit(void);
 
+#define CONFIG_DLM_DEBUG 1
 #ifdef CONFIG_DLM_DEBUG
 int dlm_register_debugfs(void);
 void dlm_unregister_debugfs(void);
