diff -Naur linux-3.14.74.org/arch/arm/mach-omap2/cclock3xxx_data.c linux-3.14.74/arch/arm/mach-omap2/cclock3xxx_data.c
--- linux-3.14.74.org/arch/arm/mach-omap2/cclock3xxx_data.c	2016-07-27 18:56:02.000000000 +0200
+++ linux-3.14.74/arch/arm/mach-omap2/cclock3xxx_data.c	2016-07-29 01:47:45.272515907 +0200
@@ -250,7 +250,7 @@
 
 static struct clk dpll1_ck;
 
-static const struct clk_ops dpll1_ck_ops = {
+static clk_ops_no_const dpll1_ck_ops = {
 	.init		= &omap2_init_clk_clkdm,
 	.enable		= &omap3_noncore_dpll_enable,
 	.disable	= &omap3_noncore_dpll_disable,
diff -Naur linux-3.14.74.org/net/ipv6/addrconf.c linux-3.14.74/net/ipv6/addrconf.c
--- linux-3.14.74.org/net/ipv6/addrconf.c	2016-07-29 03:47:13.000000000 +0200
+++ linux-3.14.74/net/ipv6/addrconf.c	2016-07-29 00:47:00.000000000 +0200
@@ -4818,7 +4818,7 @@
 {
 	struct inet6_dev *idev = ctl->extra1;
 	int min_mtu = IPV6_MIN_MTU;
-	struct ctl_table lctl;
+	ctl_table_no_const lctl;
 
 	lctl = *ctl;
 	lctl.extra1 = &min_mtu;
