diff -Naur linux-3.14.76.org/arch/arm/mach-omap2/cclock3xxx_data.c linux-3.14.76/arch/arm/mach-omap2/cclock3xxx_data.c
--- linux-3.14.76.org/arch/arm/mach-omap2/cclock3xxx_data.c	2016-08-18 06:26:02.000000000 +0200
+++ linux-3.14.76/arch/arm/mach-omap2/cclock3xxx_data.c	2016-08-18 06:37:51.442186995 +0200
@@ -250,7 +250,7 @@
 
 static struct clk dpll1_ck;
 
-static struct clk_ops dpll1_ck_ops;
+static clk_ops_no_const dpll1_ck_ops;
 
 static struct clk_ops dpll1_ck_ops_34xx __initdata = {
 	.init		= &omap2_init_clk_clkdm,
diff -Naur linux-3.14.76.org/net/ipv6/addrconf.c linux-3.14.76/net/ipv6/addrconf.c
--- linux-3.14.76.org/net/ipv6/addrconf.c	2016-08-18 06:25:51.000000000 +0200
+++ linux-3.14.76/net/ipv6/addrconf.c	2016-08-18 06:31:51.802186824 +0200
@@ -4818,7 +4818,7 @@
 {
 	struct inet6_dev *idev = ctl->extra1;
 	int min_mtu = IPV6_MIN_MTU;
-	struct ctl_table lctl;
+	ctl_table_no_const lctl;
 
 	lctl = *ctl;
 	lctl.extra1 = &min_mtu;
