diff -Nur arm.vanilla/src/util/hostnames.py arm/src/util/hostnames.py
--- arm.vanilla/src/util/hostnames.py	2012-04-29 05:59:24.000000000 +0200
+++ arm/src/util/hostnames.py	2013-07-31 17:59:19.245591564 +0200
@@ -30,7 +30,6 @@
 import threading
 import itertools
 import Queue
-import distutils.sysconfig
 
 from util import log, sysTools
 
@@ -264,7 +263,7 @@
     # 'socket.gethostbyaddr'. The following checks if the system has the
     # gethostbyname_r function, which determines if python resolutions can be
     # done in parallel or not. If so, this is preferable.
-    isSocketResolutionParallel = distutils.sysconfig.get_config_var("HAVE_GETHOSTBYNAME_R")
+    isSocketResolutionParallel = True #distutils.sysconfig.get_config_var("HAVE_GETHOSTBYNAME_R")
     self.useSocketResolution = CONFIG["queries.hostnames.useSocketModule"] and isSocketResolutionParallel
     
     for _ in range(CONFIG["queries.hostnames.poolSize"]):
