diff -Naur backports-3.18.1-1.org/Makefile backports-3.18.1-1/Makefile
--- backports-3.18.1-1.org/Makefile	2014-12-21 22:37:13.000000000 +0100
+++ backports-3.18.1-1/Makefile	2014-12-29 15:51:19.358111370 +0100
@@ -8,9 +8,9 @@
 SHELL := /bin/bash
 BACKPORT_PWD := $(shell pwd)
 
-KMODDIR ?= updates
+KMODDIR ?= kernel
 ifneq ($(origin KLIB), undefined)
-KMODPATH_ARG := "INSTALL_MOD_PATH=$(KLIB)"
+KMODPATH_ARG :=
 else
 KLIB := /lib/modules/$(shell uname -r)/
 KMODPATH_ARG :=
diff -Naur backports-3.18.1-1.org/Makefile.real backports-3.18.1-1/Makefile.real
--- backports-3.18.1-1.org/Makefile.real	2014-12-21 22:37:13.000000000 +0100
+++ backports-3.18.1-1/Makefile.real	2014-12-29 15:51:40.934780933 +0100
@@ -92,11 +92,11 @@
 	@$(MAKE) -C $(KLIB_BUILD) M=$(BACKPORT_PWD)			\
 		INSTALL_MOD_DIR=$(KMODDIR) $(KMODPATH_ARG)		\
 		modules_install
-	@./scripts/blacklist.sh $(KLIB)/ $(KLIB)/$(KMODDIR)
+#	@./scripts/blacklist.sh $(KLIB)/ $(KLIB)/$(KMODDIR)
 	@./scripts/compress_modules.sh $(KLIB)/$(KMODDIR)
-	@./scripts/check_depmod.sh
-	@/sbin/depmod -a
-	@./scripts/update-initramfs.sh $(KLIB)
+#	@./scripts/check_depmod.sh
+#	@/sbin/depmod -a
+#	@./scripts/update-initramfs.sh $(KLIB)
 	@echo
 	@echo Your backported driver modules should be installed now.
 	@echo Reboot.
