diff -Naur backports-4.2.6-1.org/drivers/net/wireless/ath/ath10k/htt_rx.c backports-4.2.6-1/drivers/net/wireless/ath/ath10k/htt_rx.c
--- backports-4.2.6-1.org/drivers/net/wireless/ath/ath10k/htt_rx.c	2015-11-15 22:19:40.000000000 +0100
+++ backports-4.2.6-1/drivers/net/wireless/ath/ath10k/htt_rx.c	2016-03-26 10:48:17.189430101 +0100
@@ -1476,7 +1476,7 @@
 	 */
 
 	if (!rx_status->freq) {
-		ath10k_warn(ar, "no channel configured; ignoring frame(s)!\n");
+		ath10k_dbg(ar,ATH10K_DBG_DATA ,"no channel configured; ignoring frame(s)!\n");
 		return false;
 	}
 
