--- bash-3.1/shell.c.login	2006-01-13 16:52:14.000000000 +0000
+++ bash-3.1/shell.c	2006-01-13 16:52:15.000000000 +0000
@@ -1543,9 +1543,10 @@
      any startup files; just try to be more like /bin/sh. */
   shell_name = argv0 ? base_pathname (argv0) : PROGRAM;
 
-  if (*shell_name == '-')
+  if (argv0 && *argv0 == '-')
     {
-      shell_name++;
+      if (*shell_name == '-')
+        shell_name++;
       login_shell++;
     }
 
