diff -ruN busybox-0.60.5.old/Config.h busybox-0.60.5/Config.h
--- busybox-0.60.5.old/Config.h	Mon Sep 16 07:30:10 2002
+++ busybox-0.60.5/Config.h	Mon Dec  8 09:57:27 2003
@@ -6,143 +6,35 @@
 //   i.e.,  //#define BB_BLAH
 //
 //
-// BusyBox Applications
-//#define BB_ADJTIMEX
-//#define BB_AR
-#define BB_ASH
-#define BB_BASENAME
-#define BB_CAT
-#define BB_CHGRP
-#define BB_CHMOD
-#define BB_CHOWN
-#define BB_CHROOT
-#define BB_CHVT
-#define BB_CLEAR
-//#define BB_CMP
-#define BB_CP
-//#define BB_CPIO
-#define BB_CUT
-#define BB_DATE
-//#define BB_DC
-#define BB_DD
-//#define BB_DEALLOCVT
+// BusyBox Applications 
+#define BB_ASH 
+#define BB_CAT 
+#define BB_CHROOT 
+#define BB_DD 
 #define BB_DF
-#define BB_DIRNAME
-#define BB_DMESG
-//#define BB_DOS2UNIX
-//#define BB_DPKG
-//#define BB_DPKG_DEB
-//#define BB_DUTMP
-#define BB_DU
-//#define BB_DUMPKMAP
-#define BB_ECHO
-#define BB_ENV
-//#define BB_EXPR
-//#define BB_FBSET
-//#define BB_FDFLUSH
-#define BB_FIND
-#define BB_FREE
-//#define BB_FREERAMDISK
-//#define BB_FSCK_MINIX
-//#define BB_GETOPT
-#define BB_GREP
-#define BB_GUNZIP
-#define BB_GZIP
-#define BB_HALT
-#define BB_HEAD
-//#define BB_HOSTID
-//#define BB_HOSTNAME
-//#define BB_HUSH
-#define BB_ID
-//#define BB_IFCONFIG
-#define BB_INIT
-//#define BB_INSMOD
-#define BB_KILL
-#define BB_KILLALL
-#define BB_KLOGD
-//#define BB_LASH
-//#define BB_LENGTH
-#define BB_LN
-//#define BB_LOADACM
-//#define BB_LOADFONT
-//#define BB_LOADKMAP
-#define BB_LOGGER
-//#define BB_LOGNAME
-//#define BB_LOSETUP
-#define BB_LS
-#define BB_LSMOD
-//#define BB_MAKEDEVS
-//#define BB_MD5SUM
-#define BB_MKDIR
-//#define BB_MKFIFO
-//#define BB_MKFS_MINIX
-#define BB_MKNOD
-#define BB_MKSWAP
-//#define BB_MKTEMP
-#define BB_MODPROBE
-#define BB_MORE
-#define BB_MOUNT
-//#define BB_MSH
-//#define BB_MT
-#define BB_MV
-//#define BB_NC
-//#define BB_NSLOOKUP
-#define BB_PIDOF
-//#define BB_PING
-//#define BB_PIVOT_ROOT
-#define BB_POWEROFF
-//#define BB_PRINTF
-#define BB_PS
-#define BB_PWD
-//#define BB_RDATE
-//#define BB_READLINK
-#define BB_REBOOT
-//#define BB_RENICE
-#define BB_RESET
-#define BB_RM
-#define BB_RMDIR
-//#define BB_RMMOD
-//#define BB_ROUTE
-//#define BB_RPM2CPIO
+#define BB_ECHO 
+#define BB_GUNZIP 
+#define BB_GZIP 
+#define BB_HALT 
+#define BB_IFCONFIG 
+#define BB_INSMOD 
+#define BB_INIT 
+#define BB_LS 
+#define BB_MKSWAP 
+#define BB_MODPROBE 
+#define BB_MOUNT 
+#define BB_REBOOT 
+#define BB_RMMOD 
 #define BB_SED
-//#define BB_SETKEYCODES
-#define BB_SLEEP
-#define BB_SORT
-//#define BB_STTY
-#define BB_SWAPONOFF
-#define BB_SYNC
+#define BB_SLEEP 
+#define BB_SWAPONOFF 
 #define BB_SYSLOGD
-#define BB_TAIL
-#define BB_TAR
-//#define BB_TEE
-//#define BB_TEST
-//#define BB_TELNET
-//#define BB_TFTP
-//#define BB_TIME
-//#define BB_TOP
-#define BB_TOUCH
-//#define BB_TR
-//#define BB_TRACEROUTE
-#define BB_TRUE_FALSE
-#define BB_TTY
-//#define BB_UNIX2DOS
-//#define BB_UUENCODE
-//#define BB_UUDECODE
-#define BB_UMOUNT
-#define BB_UNIQ
-#define BB_UNAME
-//#define BB_UPDATE
-#define BB_UPTIME
-//#define BB_USLEEP
-//#define BB_VI
-//#define BB_WATCHDOG
-#define BB_WC
-//#define BB_WGET
-#define BB_WHICH
-#define BB_WHOAMI
-#define BB_XARGS
-#define BB_YES
-// End of Applications List
+#define BB_TAR 
+#define BB_UMOUNT 
+#define BB_WGET 
+#define BB_PING 
+// End of Applications List 
+
 //
 //
 //
@@ -209,7 +101,7 @@
 #define BB_FEATURE_LS_FOLLOWLINKS
 //
 // Use color to identify different file types
-#define BB_FEATURE_LS_COLOR
+
 //
 // Disable for a smaller (but less functional) ping
 #define BB_FEATURE_FANCY_PING
@@ -333,7 +225,7 @@
 //#define BB_FEATURE_MINIX2
 //
 // Enable ifconfig status reporting output -- this feature adds 7k.
-//#define BB_FEATURE_IFCONFIG_STATUS
+#define BB_FEATURE_IFCONFIG_STATUS
 //
 // Enable ifconfig slip-specific options "keepalive" and "outfill"
 //#define BB_FEATURE_IFCONFIG_SLIP
diff -ruN busybox-0.60.5.old/insmod.c busybox-0.60.5/insmod.c
--- busybox-0.60.5.old/insmod.c	Mon Sep 16 07:30:10 2002
+++ busybox-0.60.5/insmod.c	Mon Dec  8 09:57:27 2003
@@ -3538,7 +3538,7 @@
 			 * if it is a link... */
 			if (realpath (module_dir, real_module_dir) == NULL)
 				strcpy(real_module_dir, module_dir);
-			recursive_action(real_module_dir, TRUE, FALSE, FALSE,
+			recursive_action(real_module_dir, TRUE, TRUE, FALSE,
 					check_module_name_match, 0, m_fullName);
 		}
 
@@ -3550,7 +3550,7 @@
 				strcpy(module_dir, _PATH_MODULES);
 			/* No module found under /lib/modules/`uname -r`, this
 			 * time cast the net a bit wider.  Search /lib/modules/ */
-			if (recursive_action(module_dir, TRUE, FALSE, FALSE,
+			if (recursive_action(module_dir, TRUE, TRUE, FALSE,
 						check_module_name_match, 0, m_fullName) == FALSE) 
 			{
 				if (m_filename[0] == '\0'
