diff -Naur ccache-3.1.org/ccache.c ccache-3.1/ccache.c
--- ccache-3.1.org/ccache.c	2010-09-16 19:01:09.000000000 +0200
+++ ccache-3.1/ccache.c	2010-10-21 21:41:08.000000000 +0200
@@ -805,13 +805,16 @@
 	 */
 	compilercheck = getenv("CCACHE_COMPILERCHECK");
 	if (!compilercheck) {
-		compilercheck = "mtime";
+		compilercheck = "size";
 	}
 	if (str_eq(compilercheck, "none")) {
 		/* Do nothing. */
 	} else if (str_eq(compilercheck, "content")) {
 		hash_delimiter(hash, "cc_content");
 		hash_file(hash, args->argv[0]);
+	} else if (str_eq(compilercheck, "size")) {
+		hash_delimiter(hash, "cc_size");
+		hash_int(hash, st.st_size);
 	} else if (str_eq(compilercheck, "mtime")) {
 		hash_delimiter(hash, "cc_mtime");
 		hash_int(hash, st.st_size);
