diff -Naur compat-wireless-3.5-1-snpc.org/config.mk compat-wireless-3.5/config.mk
--- compat-wireless-3.5-1-snpc.org/config.mk	2012-07-31 17:22:29.000000000 -0400
+++ compat-wireless-3.5/config.mk	2012-08-13 13:09:55.913234600 -0400
@@ -246,10 +246,12 @@
 # mac80211 test driver
 export CONFIG_MAC80211_HWSIM=m
 
+ifdef CONFIG_PCI
 export CONFIG_ATH5K=m
 # export CONFIG_ATH5K_DEBUG=y
 # export CONFIG_ATH5K_TRACER=y
 # export CONFIG_ATH5K_AHB=y
+endif #CONFIG_PCI
 
 export CONFIG_ATH9K=m
 export CONFIG_ATH9K_HW=m
