diff -Naur compat-drivers-3.8-1-u.org/drivers/net/ethernet/atheros/alx/alx_main.c compat-drivers-3.8-1-u/drivers/net/ethernet/atheros/alx/alx_main.c
--- compat-drivers-3.8-1-u.org/drivers/net/ethernet/atheros/alx/alx_main.c	2013-02-20 14:24:50.000000000 +0100
+++ compat-drivers-3.8-1-u/drivers/net/ethernet/atheros/alx/alx_main.c	2013-03-03 12:41:04.000000000 +0100
@@ -26,6 +26,7 @@
 #include <linux/bitops.h>
 #include <linux/netdevice.h>
 #include <linux/etherdevice.h>
+#include <linux/vmalloc.h>
 
 #include "alx_reg.h"
 #include "alx_hw.h"
