diff -Naur compat-wireless-2.6.34.org/include/linux/compat-2.6.33.h compat-wireless-2.6.34/include/linux/compat-2.6.33.h
--- compat-wireless-2.6.34.org/include/linux/compat-2.6.33.h	2010-05-18 03:11:59.000000000 +0200
+++ compat-wireless-2.6.34/include/linux/compat-2.6.33.h	2010-05-27 19:58:55.000000000 +0200
@@ -60,16 +60,6 @@
 /* this will never happen on older kernels */
 #define NETDEV_POST_INIT 0xffff
 
-static inline struct sk_buff *netdev_alloc_skb_ip_align(struct net_device *dev,
-                unsigned int length)
-{
-	struct sk_buff *skb = netdev_alloc_skb(dev, length + NET_IP_ALIGN);
-
-	if (NET_IP_ALIGN && skb)
-		skb_reserve(skb, NET_IP_ALIGN);
-	return skb;
-}
-
 #if defined(CONFIG_PCCARD) || defined(CONFIG_PCCARD_MODULE)
 
 #if defined(CONFIG_PCMCIA) || defined(CONFIG_PCMCIA_MODULE)
