diff -Naur compat-wireless-2.6.39-1-sn.org/compat/kfifo.c compat-wireless-2.6.39-1-sn/compat/kfifo.c
--- compat-wireless-2.6.39-1-sn.org/compat/kfifo.c	2011-05-24 01:43:48.000000000 +0200
+++ compat-wireless-2.6.39-1-sn/compat/kfifo.c	2011-05-24 17:20:06.956818313 +0200
@@ -27,6 +27,10 @@
 #include <linux/uaccess.h>
 #include <linux/kfifo.h>
 
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Stefani Seibold <stefani@seibold.net>");
+MODULE_DESCRIPTION("A generic kernel FIFO implementation");
+
 /*
  * internal helper to calculate the unused elements in a fifo
  */
