diff -Naur compat-wireless-2009-04-13.org/drivers/net/wireless/ath5k/base.c compat-wireless-2009-04-13/drivers/net/wireless/ath5k/base.c
--- compat-wireless-2009-04-13.org/drivers/net/wireless/ath5k/base.c	2009-04-13 06:14:20.000000000 +0200
+++ compat-wireless-2009-04-13/drivers/net/wireless/ath5k/base.c	2009-04-16 19:15:17.000000000 +0200
@@ -521,6 +521,7 @@
 
 	hw->wiphy->interface_modes =
 		BIT(NL80211_IFTYPE_STATION) |
+		BIT(NL80211_IFTYPE_AP) |
 		BIT(NL80211_IFTYPE_ADHOC) |
 		BIT(NL80211_IFTYPE_MESH_POINT);
 
