diff -Naur compat-drivers-3.8.3-2-snpu.org/drivers/net/wireless/ath/regd.c compat-drivers-3.8.3-2-snpu/drivers/net/wireless/ath/regd.c
--- compat-drivers-3.8.3-2-snpu.org/drivers/net/wireless/ath/regd.c	2013-03-15 22:55:09.000000000 +0100
+++ compat-drivers-3.8.3-2-snpu/drivers/net/wireless/ath/regd.c	2013-08-22 11:51:36.793297656 +0200
@@ -200,6 +200,8 @@
 	u32 bandwidth = 0;
 	int r;
 
+	return;
+
 	for (band = 0; band < IEEE80211_NUM_BANDS; band++) {
 
 		if (!wiphy->bands[band])
@@ -259,6 +261,8 @@
 	u32 bandwidth = 0;
 	int r;
 
+	return;
+
 	sband = wiphy->bands[IEEE80211_BAND_2GHZ];
 	if (!sband)
 		return;
@@ -308,6 +312,8 @@
 	struct ieee80211_channel *ch;
 	unsigned int i;
 
+	return;
+
 	if (!wiphy->bands[IEEE80211_BAND_5GHZ])
 		return;
 
@@ -514,6 +520,8 @@
 {
 	const struct ieee80211_regdomain *regd;
 
+	return 0;
+
 	wiphy->reg_notifier = reg_notifier;
 	wiphy->flags |= WIPHY_FLAG_STRICT_REGULATORY;
 
