--- cups-2.4.2/scheduler/conf.c.orig	2022-05-26 08:17:21.000000000 +0200
+++ cups-2.4.2/scheduler/conf.c	2023-04-23 15:37:48.521145700 +0200
@@ -605,11 +605,11 @@
   ClassifyOverride  = 0;
 
 #ifdef HAVE_TLS
-#  ifdef HAVE_GNUTLS
+#  if defined HAVE_GNUTLS || defined HAVE_OPENSSL
   cupsdSetString(&ServerKeychain, "ssl");
 #  else
   cupsdSetString(&ServerKeychain, "/Library/Keychains/System.keychain");
-#  endif /* HAVE_GNUTLS */
+#  endif /* HAVE_GNUTLS || HAVE_OPENSSL */
 
   _httpTLSSetOptions(_HTTP_TLS_NONE, _HTTP_TLS_1_0, _HTTP_TLS_MAX);
 #endif /* HAVE_TLS */
