diff -Naur cyrus-imapd-2.2.12.orig/doc/Makefile.dist cyrus-imapd-2.2.12/doc/Makefile.dist
--- cyrus-imapd-2.2.12.orig/doc/Makefile.dist	Wed Oct 22 20:50:04 2003
+++ cyrus-imapd-2.2.12/doc/Makefile.dist	Wed Nov  9 13:23:58 2005
@@ -13,7 +13,7 @@
 
 	rm -f groff-html-*.png pod2htm*
 
-	fig2dev -L png murder.fig murder.png
-	fig2dev -L png netnews.fig netnews.png
+#	fig2dev -L png murder.fig murder.png
+#	fig2dev -L png netnews.fig netnews.png
 
 	(cd text; make)
