--- tools/config2header	2005/04/13 08:56:16	1.1
+++ tools/config2header	2005/04/13 08:57:48
@@ -214,7 +214,6 @@
 print HFILE <<EOF
   IMAPOPT_LAST
 };
-extern struct imapopt_s imapopts[];
 
 enum enum_value {
   IMAP_ENUM_ZERO = 0,
@@ -260,6 +259,9 @@
 
 print HFILE <<EOF
 };
+
+extern struct imapopt_s imapopts[];
+
 #endif /* INCLUDED_IMAPOPTIONS_H */
 EOF
     ;
--- imap/fud.c	2005/04/13 09:19:05	1.1
+++ imap/fud.c	2005/04/13 09:20:02
@@ -202,6 +202,7 @@
     r = begin_handling();
 
     shut_down(r);
+    return 0;
 }
 
 static void cyrus_timeout(int signo __attribute__((unused)))
--- imap/smmapd.c	2005/04/13 09:19:12	1.1
+++ imap/smmapd.c	2005/04/13 09:19:50
@@ -199,6 +199,7 @@
     r = begin_handling();
 
     shut_down(r);
+    return 0;
 }
 
 int verify_user(const char *key, long quotacheck,
--- notifyd/notifyd.c	2005/04/13 09:19:23	1.1
+++ notifyd/notifyd.c	2005/04/13 09:19:39
@@ -266,4 +266,5 @@
     r = do_notify();
 
     shut_down(r);
+    return 0;
 }
