diff -Naur cyrus-imapd-2.2.12.orig/imap/spool.c cyrus-imapd-2.2.12/imap/spool.c
--- cyrus-imapd-2.2.12.orig/imap/spool.c	2004-10-27 22:40:50.000000000 +0200
+++ cyrus-imapd-2.2.12/imap/spool.c	2005-10-24 15:08:37.000000000 +0200
@@ -451,7 +451,7 @@
 	    p[1] = '\n';
 	    p[2] = '\0';
 	}
-	else if (p[0] != '\n' && (strlen(buf) < sizeof(buf)-2)) {
+	else if (p[0] != '\n' && (strlen(buf) < sizeof(buf)-3)) {
 	    /* line contained a \0 not at the end */
 	    r = IMAP_MESSAGE_CONTAINSNULL;
 	    continue;
