From 21fd4b8d26d01d622185ab8de971a9ee934220a3 Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Thu, 24 Jul 2014 13:23:36 +0200
Subject: [PATCH] Add a program prefix to syslog messages.

---
 src/ddns/__init__.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/ddns/__init__.py b/src/ddns/__init__.py
index 22764e6..6fe3a33 100644
--- a/src/ddns/__init__.py
+++ b/src/ddns/__init__.py
@@ -42,6 +42,8 @@ def setup_logging():
 	handler = logging.handlers.SysLogHandler(address="/dev/log",
 		facility=logging.handlers.SysLogHandler.LOG_DAEMON
 	)
+	formatter = logging.Formatter("ddns[%(process)d]: %(message)s")
+	handler.setFormatter(formatter)
 	handler.setLevel(logging.INFO)
 	rootlogger.addHandler(handler)
 
-- 
1.9.3

