diff -up dhcp-4.2.1b1/client/dhclient.conf.supersede dhcp-4.2.1b1/client/dhclient.conf
--- dhcp-4.2.1b1/client/dhclient.conf.supersede	2010-09-15 01:03:56.000000000 +0200
+++ dhcp-4.2.1b1/client/dhclient.conf	2011-01-27 18:38:28.000000000 +0100
@@ -4,7 +4,7 @@ send dhcp-lease-time 3600;
 supersede domain-search "fugue.com", "home.vix.com";
 prepend domain-name-servers 127.0.0.1;
 request subnet-mask, broadcast-address, time-offset, routers,
-	domain-name, domain-name-servers, host-name;
+	domain-search, domain-name-servers, host-name;
 require subnet-mask, domain-name-servers;
 timeout 60;
 retry 60;
