Fix for CAN-2005-1848 borrowed from gentoo / debian
--- client.c.orig	2005-07-10 10:16:43.000000000 +0100
+++ client.c	2005-07-10 10:22:31.000000000 +0100
@@ -129,6 +129,8 @@
        	default:
 	  if ( p[1] )
 	    {
+	      if (p + 2 + p[1] >= end)
+		  goto swend; /* Corrupt packet */
 	      if ( DhcpOptions.len[*p] == p[1] )
 	        memcpy(DhcpOptions.val[*p],p+2,p[1]);
 	      else
