diff -Naur dracut-006.org/dracut dracut-006/dracut
--- dracut-006.org/dracut	2010-06-17 10:46:29.000000000 +0200
+++ dracut-006/dracut	2010-08-07 21:00:38.000000000 +0200
@@ -308,9 +308,7 @@
 	#strip -R .comment $note "$f" || :
     done
 fi
-
-type pigz &>/dev/null && gzip=pigz || gzip=gzip
-( cd "$initdir"; find . |cpio -R 0:0 -H newc -o --quiet |$gzip -9 > "$outfile"; ) 
+( cd "$initdir"; find . |cpio -R 0:0 -H newc -o --quiet | lzma > "$outfile"; ) 
 if [ $? -ne 0 ]; then
     derror "dracut: creation of $outfile failed"
     exit 1
