diff -Naur dracut-006.org/modules.d/99base/init dracut-006/modules.d/99base/init
--- dracut-006.org/modules.d/99base/init	2010-06-17 10:46:29.000000000 +0200
+++ dracut-006/modules.d/99base/init	2013-10-31 17:05:09.135475525 +0100
@@ -175,7 +175,7 @@
 
     $UDEV_QUEUE_EMPTY >/dev/null 2>&1 || continue
 
-    modprobe scsi_wait_scan && rmmod scsi_wait_scan
+    sleep 0.1
 
     $UDEV_QUEUE_EMPTY >/dev/null 2>&1 || continue
 
