diff -Naur dracut-006.org/modules.d/99base/init dracut-006/modules.d/99base/init
--- dracut-006.org/modules.d/99base/init	2010-06-17 10:46:29.000000000 +0200
+++ dracut-006/modules.d/99base/init	2013-10-31 16:51:17.658771341 +0100
@@ -87,6 +87,9 @@
 
 ln -s /proc/self/fd /dev/fd >/dev/null 2>&1
 
+# create run dir
+mkdir run
+
 if getarg rdinitdebug; then
     getarg quiet && DRACUT_QUIET="yes"
     mkfifo /dev/initlog.pipe
