diff -Naur dracut-038.org/modules.d/90kernel-modules/module-setup.sh dracut-038/modules.d/90kernel-modules/module-setup.sh
--- dracut-038.org/modules.d/90kernel-modules/module-setup.sh	2014-06-30 12:03:12.000000000 +0200
+++ dracut-038/modules.d/90kernel-modules/module-setup.sh	2015-05-05 14:58:56.820197839 +0200
@@ -47,7 +47,7 @@
             atkbd i8042 usbhid hid-apple hid-sunplus hid-cherry hid-logitech \
             hid-logitech-dj hid-microsoft firewire-ohci \
             pcmcia usb_storage nvme hid-hyperv hv-vmbus \
-            sdhci_acpi
+            hyperv-keyboard sdhci_acpi sdhci_pci
 
         if [[ "$(uname -p)" == arm* ]]; then
             # arm specific modules
