diff -Naur org/init new/init
--- init	2010-06-17 10:46:29.000000000 +0200
+++ init	2010-08-10 17:55:41.000000000 +0200
@@ -111,8 +111,9 @@
 getarg 'rdbreak=cmdline' && emergency_shell -n cmdline "Break before cmdline"
 source_all cmdline
 
-[ -z "$root" ] && die "No or empty root= argument"
-[ -z "$rootok" ] && die "Don't know how to handle 'root=$root'"
+# Disable root argument check ...
+#[ -z "$root" ] && die "No or empty root= argument"
+#[ -z "$rootok" ] && die "Don't know how to handle 'root=$root'"
 
 # Network root scripts may need updated root= options,
 # so deposit them where they can see them (udev purges the env)
@@ -202,12 +203,15 @@
     done
     
     i=$(($i+1))
-    [ $i -gt $RDRETRY ] \
-        && { flock -s 9 ; emergency_shell "No root device found"; } 9>/.console_lock
+	# Start IPFire installer after root was not found ;)
+    [ $i -gt $RDRETRY ] && break 2;
+
 done
 unset job
 unset queuetriggered
 
+/etc/rc_installer
+
 # pre-mount happens before we try to mount the root filesystem,
 # and happens once.
 getarg 'rdbreak=pre-mount' && emergency_shell -n pre-mount "Break pre-mount"
