diff -Naur dracut-006.org/modules.d/99base/init dracut-006/modules.d/99base/init
--- dracut-006.org/modules.d/99base/init	2010-06-17 10:46:29.000000000 +0200
+++ dracut-006/modules.d/99base/init	2011-01-12 11:40:41.713190125 +0100
@@ -190,17 +190,6 @@
     # no more udev jobs and queues empty.
     sleep 0.5
 
-    # dirty hack for some cdrom drives,
-    # which report no medium for quiet
-    # some time.
-    for cdrom in /sys/block/sr*; do
-	[ -e "$cdrom" ] || continue
-	# skip, if cdrom medium was already found
-        strstr "$(udevadm info --query=env --path=${cdrom##/sys})" \
-	    ID_CDROM_MEDIA && continue
-        echo change > "$cdrom/uevent"        
-    done
-    
     i=$(($i+1))
     [ $i -gt $RDRETRY ] \
         && { flock -s 9 ; emergency_shell "No root device found"; } 9>/.console_lock
