diff -ruN e2fsprogs-1.34.old/lib/ext2fs/mkjournal.c e2fsprogs-1.34/lib/ext2fs/mkjournal.c
--- e2fsprogs-1.34.old/lib/ext2fs/mkjournal.c	Sun Feb  3 07:03:43 2002
+++ e2fsprogs-1.34/lib/ext2fs/mkjournal.c	Fri Dec 12 11:20:53 2003
@@ -312,11 +312,7 @@
 	ext2_ino_t		journal_ino;
 	struct stat		st;
 	char			jfile[1024];
-	int			fd, mount_flags, f;
-
-	if ((retval = ext2fs_check_mount_point(fs->device_name, &mount_flags,
-					       jfile, sizeof(jfile)-10)))
-		return retval;
+	int			fd, mount_flags=0, f;
 
 	if (mount_flags & EXT2_MF_MOUNTED) {
 		strcat(jfile, "/.journal");
