*** orgdriver/ide/pci/via82cxxx.c	Thu May  3 21:49:52 2007
--- drivers/ide/pci/via82cxxx.c	Thu May 10 11:51:49 2007
***************
*** 67,72 ****
--- 67,76 ----
  #define VIA_NO_UNMASK		0x080	/* Doesn't work with IRQ unmasking on */
  #define VIA_BAD_ID		0x100	/* Has wrong vendor ID (0x1107) */
  #define VIA_BAD_AST		0x200	/* Don't touch Address Setup Timing */
+ #define PCI_DEVICE_ID_VIA_VT8251 0x3287
+ #define PCI_DEVICE_ID_VIA_VT8237A 0x3337
+ #define PCI_DEVICE_ID_VIA_CX700 0x8324
+ #define PCI_DEVICE_ID_VIA_CX700_IDE 0x5324
  
  /*
   * VIA SouthBridge chips.
***************
*** 79,84 ****
--- 83,91 ----
  	u8 rev_max;
  	u16 flags;
  } via_isa_bridges[] = {
+ 	{ "vt8251", PCI_DEVICE_ID_VIA_VT8251 , 0x00, 0x2f, VIA_UDMA_133 | VIA_BAD_AST },
+ 	{ "vt8237a", PCI_DEVICE_ID_VIA_VT8237A , 0x00, 0x2f, VIA_UDMA_133 | VIA_BAD_AST },
+ 	{ "cx700", PCI_DEVICE_ID_VIA_CX700, 0x00, 0x2f, VIA_UDMA_133 | VIA_BAD_AST },
  	{ "vt6410",	PCI_DEVICE_ID_VIA_6410,     0x00, 0x2f, VIA_UDMA_133 | VIA_BAD_AST },
  	{ "vt8251",	PCI_DEVICE_ID_VIA_8251,     0x00, 0x2f, VIA_UDMA_133 | VIA_BAD_AST },
  	{ "vt8237",	PCI_DEVICE_ID_VIA_8237,     0x00, 0x2f, VIA_UDMA_133 | VIA_BAD_AST },
***************
*** 498,503 ****
--- 505,511 ----
  static struct pci_device_id via_pci_tbl[] = {
  	{ PCI_VENDOR_ID_VIA, PCI_DEVICE_ID_VIA_82C576_1, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0},
  	{ PCI_VENDOR_ID_VIA, PCI_DEVICE_ID_VIA_82C586_1, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0},
+ 	{ PCI_VENDOR_ID_VIA, PCI_DEVICE_ID_VIA_CX700_IDE, PCI_ANY_ID, PCI_ANY_ID, 0, 0,0},
  	{ PCI_VENDOR_ID_VIA, PCI_DEVICE_ID_VIA_6410,     PCI_ANY_ID, PCI_ANY_ID, 0, 0, 1},
  	{ 0, },
  };
