From d1f6789178e141664895332af79dbe7aa9e5ab68 Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Wed, 13 Jun 2018 20:58:42 +0200
Subject: [PATCH 2/2] Ignore any systems with ID "Default string"

Signed-off-by: Michael Tremer <michael.tremer@ipfire.org>
---
 src/fireinfo/system.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/fireinfo/system.py b/src/fireinfo/system.py
index e3295af..73c3882 100644
--- a/src/fireinfo/system.py
+++ b/src/fireinfo/system.py
@@ -42,6 +42,7 @@ INVALID_ID_STRINGS = (
 	"N/A", "n/a",
 	"12345", "54321", "202020",
 	"Chassis", "chassis",
+	"Default string",
 	"EVAL",
 	"Not Applicable",
 	"None", "empty",
-- 
2.12.2

