From edacae4b2cdc41f1c0bfc93e041532ff6c49f60c Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Tue, 17 Mar 2015 22:19:17 +0100
Subject: [PATCH] Add an other forbidden string: "Serial"

---
 src/fireinfo/system.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/fireinfo/system.py b/src/fireinfo/system.py
index daf77b399d20..9d7872822b85 100644
--- a/src/fireinfo/system.py
+++ b/src/fireinfo/system.py
@@ -45,7 +45,7 @@ INVALID_ID_STRINGS = (
 	"EVAL",
 	"Not Applicable",
 	"None", "empty",
-	"System Serial Number",
+	"Serial", "System Serial Number",
 	"XXXXX",
 	"01010101-0101-0101-0101-010101010101",
 	"00020003-0004-0005-0006-000700080009",
-- 
2.1.0

