diff -up flac-1.2.1/examples/cpp/encode/file/main.cpp.gcc43 flac-1.2.1/examples/cpp/encode/file/main.cpp
--- flac-1.2.1/examples/cpp/encode/file/main.cpp.gcc43	2007-09-13 17:58:03.000000000 +0200
+++ flac-1.2.1/examples/cpp/encode/file/main.cpp	2008-01-08 10:27:39.000000000 +0100
@@ -29,6 +29,7 @@
 #endif
 
 #include <stdio.h>
+#include <string.h>
 #include <stdlib.h>
 #include "FLAC++/metadata.h"
 #include "FLAC++/encoder.h"
