diff -Naur gcc-4.1.2.org/libstdc++-v3/configure gcc-4.1.2/libstdc++-v3/configure
--- gcc-4.1.2.org/libstdc++-v3/configure	2007-01-29 11:51:01.000000000 +0100
+++ gcc-4.1.2/libstdc++-v3/configure	2011-08-29 19:25:22.973430586 +0200
@@ -8281,7 +8281,7 @@
   if test x"$with_gnu_ld" = x"yes"; then
 
     ldver=`$LD --version 2>/dev/null | head -1 | \
-           sed -e 's/GNU ld version \([0-9.][0-9.]*\).*/\1/'`
+           sed -e 's/GNU ld .*) \([0-9.][0-9.]*\).*/\1/'`
 
     glibcxx_gnu_ld_version=`echo $ldver | \
            $AWK -F. '{ if (NF<3) $3=0; print ($1*100+$2)*100+$3 }'`
