--- glibc-2.12.2/ports/sysdeps/unix/sysv/linux/arm/sigrestorer.S	2011-09-08 22:22:41.000000000 +0400
+++ glibc-2.12.2/ports/sysdeps/unix/sysv/linux/arm/sigrestorer.S	2011-09-08 21:44:49.000000000 +0400
@@ -18,16 +18,20 @@ 
 
 #include <sysdep.h>
 
+        .text
+
 /* If no SA_RESTORER function was specified by the application we use
    one of these.  This avoids the need for the kernel to synthesise a return
    instruction on the stack, which would involve expensive cache flushes. */
 
 ENTRY(__default_sa_restorer)
 	swi	SYS_ify(sigreturn)
+PSEUDO_END (__default_sa_restorer)
 
 #ifdef __NR_rt_sigreturn
 
 ENTRY(__default_rt_sa_restorer)
 	swi	SYS_ify(rt_sigreturn)
+PSEUDO_END (__default_rt_sa_restorer)
 
 #endif
