--- glibc-2.27/localedata/Makefile~	2018-02-07 22:31:00.968944449 +0000
+++ glibc-2.27/localedata/Makefile	2018-02-07 22:31:26.201764894 +0000
@@ -223,6 +223,7 @@ $(INSTALL-SUPPORTED-LOCALES): install-lo
 	echo -n '...'; \
 	input=`echo $$locale | sed 's/\([^.]*\)[^@]*\(.*\)/\1\2/'`; \
 	$(LOCALEDEF) $$flags --alias-file=../intl/locale.alias \
+		     --no-archive \
 		     -i locales/$$input -f charmaps/$$charset \
 		     $(addprefix --prefix=,$(install_root)) $$locale \
 	&& echo ' done'; \
