From 9191c04a7e19fffbea0a08523e579cd8e55142df Mon Sep 17 00:00:00 2001
From: Ulrich Drepper <drepper@gmail.com>
Date: Sat, 23 Jul 2011 15:28:31 -0400
Subject: [PATCH] Adjust test for correct installation

diff --git a/scripts/test-installation.pl b/scripts/test-installation.pl
index 90cd9d7..25a919b 100755
--- a/scripts/test-installation.pl
+++ b/scripts/test-installation.pl
@@ -1,5 +1,5 @@
 #! /usr/bin/perl -w
-# Copyright (C) 1997, 1998, 1999, 2004 Free Software Foundation, Inc.
+# Copyright (C) 1997, 1998, 1999, 2004, 2011 Free Software Foundation, Inc.
 # This file is part of the GNU C Library.
 # Contributed by Andreas Jaeger <aj@arthur.rhein-neckar.de>, 1997.
 
@@ -105,8 +105,10 @@ while (<SOVERSIONS>) {
     # - libdb1 since it conflicts with libdb
     # - libnss1_* from glibc-compat add-on
     # - libthread_db since it contains unresolved references
+    # - it's just a test NSS module
     if ($name ne "nss_ldap" && $name ne "db1"
-	&& !($name =~/^nss1_/) && $name ne "thread_db") {
+	&& !($name =~/^nss1_/) && $name ne "thread_db"
+	&& $name ne "nss_test1") {
       $link_libs .= " -l$name";
       $versions{$name} = $version;
     }
-- 
1.7.3.4

