--- grub-0.95/stage2/graphics.c.old	2005-01-11 09:34:17.125451981 -0500
+++ grub-0.95/stage2/graphics.c	2005-01-11 09:34:03.743949015 -0500
@@ -108,12 +108,8 @@
  * mode.  */
 int graphics_init()
 {
-    if (!graphics_inited) {
-        saved_videomode = set_videomode(0x12);
-    }
-
     if (!read_image(splashimage)) {
-        set_videomode(saved_videomode);
+	current_term = term_table;
         grub_printf("failed to read image\n");
         return 0;
     }
@@ -303,6 +299,8 @@
     if (!xpm_open(s))
         return 0;
 
+    saved_videomode = set_videomode(0x12);
+
     /* parse info */
     while (grub_read(&c, 1)) {
         if (c == '"')
