Only in grub: ChangeLog~
diff -ur grub-0.90/util/grub-install.in grub/util/grub-install.in
--- grub-0.90/util/grub-install.in	Wed Sep  5 15:50:06 2001
+++ grub/util/grub-install.in	Wed Sep  5 15:50:24 2001
@@ -320,6 +320,10 @@
     exit 1
 fi
 
+if ! test -e ${grubdir}/grub.conf ; then
+    test -e ${grubdir}/menu.lst && ln -s ./menu.lst ${grubdir}/grub.conf
+fi
+
 # Create a safe temporary file.
 test -x /bin/mktemp && log_file=`/bin/mktemp /tmp/grub-install.log.XXXXXX`
 
Only in grub/util: grub-install.in.orig
