diff -Naur grub-2.00.org/grub-core/kern/dl.c grub-2.00/grub-core/kern/dl.c
--- grub-2.00.org/grub-core/kern/dl.c	2012-06-22 17:26:38.000000000 +0200
+++ grub-2.00/grub-core/kern/dl.c	2017-07-05 10:45:25.501645684 +0200
@@ -354,8 +354,11 @@
     if (s->sh_type == SHT_SYMTAB)
       break;
 
+  /* Module without symbol table may still be used to pull in dependencies.
+     We verify at build time that such modules do not contain any relocations
+     that may reference symbol table. */
   if (i == e->e_shnum)
-    return grub_error (GRUB_ERR_BAD_MODULE, N_("no symbol table"));
+     return GRUB_ERR_NONE;
 
 #ifdef GRUB_MODULES_MACHINE_READONLY
   mod->symtab = grub_malloc (s->sh_size);
diff -Naur grub-2.00.org/grub-core/kern/i386/dl.c grub-2.00/grub-core/kern/i386/dl.c
--- grub-2.00.org/grub-core/kern/i386/dl.c	2012-02-26 12:14:21.000000000 +0100
+++ grub-2.00/grub-core/kern/i386/dl.c	2017-07-05 17:01:43.333955065 +0200
@@ -55,7 +55,7 @@
       break;
 
   if (i == e->e_shnum)
-    return grub_error (GRUB_ERR_BAD_MODULE, N_("no symbol table"));
+    return GRUB_ERR_NONE;
 
   entsize = s->sh_entsize;
 
diff -Naur grub-2.00.org/grub-core/kern/x86_64/dl.c grub-2.00/grub-core/kern/x86_64/dl.c
--- grub-2.00.org/grub-core/kern/x86_64/dl.c	2012-02-26 12:14:56.000000000 +0100
+++ grub-2.00/grub-core/kern/x86_64/dl.c	2017-07-05 17:02:23.374574458 +0200
@@ -55,7 +55,7 @@
       break;
 
   if (i == e->e_shnum)
-    return grub_error (GRUB_ERR_BAD_MODULE, N_("no symbol table"));
+    return GRUB_ERR_NONE;
 
   entsize = s->sh_entsize;
 
