--- grub-0.95/stage2/stage2.c.hidden	2004-09-30 16:39:38.028893664 -0400
+++ grub-0.95/stage2/stage2.c	2004-09-30 16:42:07.981097464 -0400
@@ -235,6 +235,8 @@
   char *cur_entry = 0;
   struct term_entry *prev_term = NULL;
 
+  cls();
+
   /*
    *  Main loop for menu UI.
    */
@@ -262,14 +264,16 @@
       /* Get current time.  */
       while ((time1 = getrtsecs ()) == 0xFF)
 	;
+      grub_printf("\rPress any key to enter the menu\n\n\n");
 
       while (1)
 	{
 	  /* Check if ESC is pressed.  */
-	  if (checkkey () != -1 && ASCII_CHAR (getkey ()) == '\e')
+	  if (checkkey () != -1)
 	    {
 	      grub_timeout = -1;
 	      show_menu = 1;
+	      getkey ();
 	      break;
 	    }
 
@@ -288,7 +292,8 @@
 	      grub_timeout--;
 	      
 	      /* Print a message.  */
-	      grub_printf ("\rPress `ESC' to enter the menu... %d   ",
+	      grub_printf ("\rBooting %s in %d seconds...",
+			   get_entry(menu_entries, first_entry + entryno, 0),
 			   grub_timeout);
 	    }
 	}
