--- grub-0.95/configure.ac.static	2004-06-13 13:38:17.000000000 -0400
+++ grub-0.95/configure.ac	2004-06-18 17:35:23.513374336 -0400
@@ -207,9 +207,9 @@
 
 # Unless the user specify --without-curses, check for curses.
 if test "x$with_curses" != "xno"; then
-  AC_CHECK_LIB(ncurses, wgetch, [GRUB_LIBS="$GRUB_LIBS -lncurses"
+  AC_CHECK_LIB(ncurses, wgetch, [GRUB_LIBS="$GRUB_LIBS -Wl,-Bstatic -lncurses -Wl,-Bdynamic"
   AC_DEFINE(HAVE_LIBCURSES, 1, [Define if you have a curses library])],
-    [AC_CHECK_LIB(curses, wgetch, [GRUB_LIBS="$GRUB_LIBS -lcurses"
+    [AC_CHECK_LIB(curses, wgetch, [GRUB_LIBS="$GRUB_LIBS -Wl,-Bstatic -lcurses -Wl,-Bdynamic"
        AC_DEFINE(HAVE_LIBCURSES, 1, [Define if you have a curses library])])])
 fi
 
