--- grub-0.97/util/grub-install.in.mpath	2006-08-02 17:21:53.000000000 -0400
+++ grub-0.97/util/grub-install.in	2006-08-02 17:23:23.000000000 -0400
@@ -107,6 +107,7 @@
 				  -e 's%\(fd[0-9]*\)$%\1%' \
 				  -e 's%/part[0-9]*$%/disc%' \
 				  -e 's%\(c[0-7]d[0-9]*\).*$%\1%' \
+	  -e 's%\(/mapper/[[:alpha:]]\+[[:digit:]]\+\)p[[:digit:]]\+$%\1%' \
 	  -e 's%\(/mapper/[[:alpha:]]\+_[[:alpha:]]\+\)[[:digit:]]\+$%\1%'`
 	tmp_part=`echo "$1" | grep -v '/mapper/control$' |
 		grep -v '/mapper/[[:alnum:]]\+-[[:alnum:]]\+$' | uniq |
@@ -116,6 +117,7 @@
 				  -e 's%.*/floppy/[0-9]*$%%' \
 				  -e 's%.*/\(disc\|part\([0-9]*\)\)$%\2%' \
 				  -e 's%.*c[0-7]d[0-9]*p%%' \
+	  -e 's%.*/mapper/[[:alpha:]]\+[[:digit:]]\+p\([[:digit:]]\+\)$%\1%' \
 	  -e 's%.*/mapper/[[:alpha:]]\+_[[:alpha:]]\+\([[:digit:]]\+\)$%\1%' |
 		grep -v '.*/mapper/.*'`
 	;;
