Index: icecc-0.9.6/doc/man-scheduler.1.docbook
===================================================================
--- icecc-0.9.6.orig/doc/man-scheduler.1.docbook
+++ icecc-0.9.6/doc/man-scheduler.1.docbook
@@ -23,13 +23,13 @@
 </refmeta>
 
 <refnamediv>
-	<refname>scheduler</refname>
+	<refname>icecc-scheduler</refname>
 	<refpurpose>Icecream scheduler</refpurpose>
 </refnamediv>
 
 <refsynopsisdiv>
 <cmdsynopsis>
-<command>scheduler</command>
+<command>icecc-scheduler</command>
 <group>
   <arg choice="opt">
     <option>-n</option>
@@ -151,7 +151,7 @@ need to run the scheduler with root righ
 
 <refsect1>
 <title>See Also</title>
-<para>icecream, scheduler, iceccd, icemon</para>
+<para>icecream, icecc-scheduler, iceccd, icemon</para>
 </refsect1>
 
 <refsect1>
Index: icecc-0.9.6/services/Makefile.am
===================================================================
--- icecc-0.9.6.orig/services/Makefile.am
+++ icecc-0.9.6/services/Makefile.am
@@ -11,9 +11,9 @@ ice_HEADERS = job.h comm.h
 noinst_HEADERS = bench.h exitcode.h getifaddrs.h logging.h tempfile.h platform.h
 icedir = $(includedir)/icecc
 
-sbin_PROGRAMS = scheduler
-scheduler_SOURCES = scheduler.cpp
-scheduler_LDADD = libicecc.la
+sbin_PROGRAMS = icecc-scheduler
+icecc_scheduler_SOURCES = scheduler.cpp
+icecc_scheduler_LDADD = libicecc.la
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = icecc.pc
Index: icecc-0.9.6/services/Makefile.in
===================================================================
--- icecc-0.9.6.orig/services/Makefile.in
+++ icecc-0.9.6/services/Makefile.in
@@ -37,7 +37,7 @@ PRE_UNINSTALL = :
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-sbin_PROGRAMS = scheduler$(EXEEXT)
+sbin_PROGRAMS = icecc-scheduler$(EXEEXT)
 subdir = services
 DIST_COMMON = $(ice_HEADERS) $(noinst_HEADERS) $(srcdir)/Makefile.am \
 	$(srcdir)/Makefile.in $(srcdir)/icecc.pc.in
@@ -82,9 +82,9 @@ libicecc_la_LINK = $(LIBTOOL) --tag=CXX 
 	$(LIBTOOLFLAGS) --mode=link $(CXXLD) $(libicecc_la_CXXFLAGS) \
 	$(CXXFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
 PROGRAMS = $(sbin_PROGRAMS)
-am_scheduler_OBJECTS = scheduler.$(OBJEXT)
-scheduler_OBJECTS = $(am_scheduler_OBJECTS)
-scheduler_DEPENDENCIES = libicecc.la
+am_icecc_scheduler_OBJECTS = scheduler.$(OBJEXT)
+icecc_scheduler_OBJECTS = $(am_icecc_scheduler_OBJECTS)
+icecc_scheduler_DEPENDENCIES = libicecc.la
 DEFAULT_INCLUDES = -I.@am__isrc@ -I$(top_builddir)
 depcomp = $(SHELL) $(top_srcdir)/depcomp
 am__depfiles_maybe = depfiles
@@ -107,8 +107,8 @@ CXXLD = $(CXX)
 CXXLINK = $(LIBTOOL) --tag=CXX $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) \
 	--mode=link $(CXXLD) $(AM_CXXFLAGS) $(CXXFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
-SOURCES = $(libicecc_la_SOURCES) $(scheduler_SOURCES)
-DIST_SOURCES = $(libicecc_la_SOURCES) $(scheduler_SOURCES)
+SOURCES = $(libicecc_la_SOURCES) $(icecc_scheduler_SOURCES)
+DIST_SOURCES = $(libicecc_la_SOURCES) $(icecc_scheduler_SOURCES)
 DATA = $(pkgconfig_DATA)
 HEADERS = $(ice_HEADERS) $(noinst_HEADERS)
 ETAGS = etags
@@ -242,8 +242,8 @@ libicecc_la_CXXFLAGS = -fPIC -DPIC
 ice_HEADERS = job.h comm.h
 noinst_HEADERS = bench.h exitcode.h getifaddrs.h logging.h tempfile.h platform.h
 icedir = $(includedir)/icecc
-scheduler_SOURCES = scheduler.cpp
-scheduler_LDADD = libicecc.la
+icecc_scheduler_SOURCES = scheduler.cpp
+icecc_scheduler_LDADD = libicecc.la
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = icecc.pc
 all: all-am
@@ -358,9 +358,9 @@ clean-sbinPROGRAMS:
 	list=`for p in $$list; do echo "$$p"; done | sed 's/$(EXEEXT)$$//'`; \
 	echo " rm -f" $$list; \
 	rm -f $$list
-scheduler$(EXEEXT): $(scheduler_OBJECTS) $(scheduler_DEPENDENCIES) 
-	@rm -f scheduler$(EXEEXT)
-	$(CXXLINK) $(scheduler_OBJECTS) $(scheduler_LDADD) $(LIBS)
+icecc-scheduler$(EXEEXT): $(icecc_scheduler_OBJECTS) $(icecc_scheduler_DEPENDENCIES) 
+	@rm -f icecc-scheduler$(EXEEXT)
+	$(CXXLINK) $(icecc_scheduler_OBJECTS) $(icecc_scheduler_LDADD) $(LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
