--- imspector/sslstate.h	2010-02-23 18:21:56.000000000 +0100
+++ imspector/sslstate.h	2013-12-14 15:40:04.796370688 +0100
@@ -52,7 +52,7 @@
 		int sslverify;            //!< Our VERIFY_ choice.
 		
 		bool localdebugmode;      //!< Saves us from passing this around.
-		SSL_METHOD *method;       //!< openssl method handle.
+		const SSL_METHOD *method;       //!< openssl method handle.
 		SSL_CTX *connectctx;      //!< Our IM server-sde CTX.
 		SSL_CTX *ctx;             //!< Our IM client-side CTX.

