--- net/ipv4/netfilter/ip_conntrack_standalone.c~	2006-12-12 01:50:40.000000000 +0000
+++ net/ipv4/netfilter/ip_conntrack_standalone.c	2006-12-16 16:29:28.000000000 +0000
@@ -804,7 +804,7 @@
 
 #ifdef CONFIG_PROC_FS
 	ret = -ENOMEM;
-	proc = proc_net_fops_create("ip_conntrack", 0440, &ct_file_ops);
+	proc = proc_net_fops_create("ip_conntrack", 0444, &ct_file_ops);
 	if (!proc) goto cleanup_init;
 
 	proc_exp = proc_net_fops_create("ip_conntrack_expect", 0440,
