diff -Nur iptraf-ng-1.1.4-orig/src/tcptable.c iptraf-ng-1.1.4/src/tcptable.c
--- iptraf-ng-1.1.4-orig/src/tcptable.c	2013-07-23 15:06:11.000000000 +0200
+++ iptraf-ng-1.1.4/src/tcptable.c	2013-08-28 19:10:33.000000000 +0200
@@ -444,6 +444,8 @@
 				 size_t bufsize)
 {
 	time_t interval = time(NULL) - entry->conn_starttime;
+	if (interval < 1)
+		interval = 1;
 
 	char rbuf[64];
 	rate_print(entry->bcount / interval, rbuf, sizeof(rbuf));
