*** ./icn/icnctrl.c.old	2002-11-27 14:30:43.000000000 +0000
--- ./icn/icnctrl.c	2002-11-27 14:31:10.000000000 +0000
***************
*** 223,229 ****
  	icn_cdef newcard;
  
  	cmd = strrchr(argv[0], '/');
! 	cmd = (cmd == NULL) ? argv[0] : ++cmd;
  	if (argc > 1) {
  		if (!strcmp(argv[1], "-d")) {
  			strcpy(ioctl_s.drvid, argv[2]);
--- 223,229 ----
  	icn_cdef newcard;
  
  	cmd = strrchr(argv[0], '/');
! 	cmd = (cmd == NULL) ? argv[0] : (cmd + 1);
  	if (argc > 1) {
  		if (!strcmp(argv[1], "-d")) {
  			strcpy(ioctl_s.drvid, argv[2]);
*** ./isdnlog/isdnlog/processor.c.old	2002-11-27 14:27:28.000000000 +0000
--- ./isdnlog/isdnlog/processor.c	2002-11-27 14:27:41.000000000 +0000
***************
*** 3497,3503 ****
       isdn_net_ioctl_phone netdvX_phone;
  #endif
    }	   phone;
!   auto	   int rc, chan, l1, l2, lmin, lmax, ldiv, match;
  
    if ((iflst = fopen("/proc/net/dev", "r")) == NULL)
      return(-1);
--- 3497,3503 ----
       isdn_net_ioctl_phone netdvX_phone;
  #endif
    }	   phone;
!   auto	   int rc, chan = 0, l1, l2, lmin, lmax, ldiv, match;
  
    if ((iflst = fopen("/proc/net/dev", "r")) == NULL)
      return(-1);
*** ./isdnlog/isdnrep/isdnbill.c.old	2002-11-27 14:30:11.000000000 +0000
--- ./isdnlog/isdnrep/isdnbill.c	2002-11-27 14:30:22.000000000 +0000
***************
*** 874,880 ****
  #endif
    auto     char     s[BUFSIZ], sx[BUFSIZ];
    auto     int      i, l, col, day, lday = UNKNOWN, month, lmonth = UNKNOWN;
!   auto     double   dur;
    auto     char    *version;
    auto     char    *myname = basename(argv[0]);
    auto     int      opt, go, s0, indent;
--- 874,880 ----
  #endif
    auto     char     s[BUFSIZ], sx[BUFSIZ];
    auto     int      i, l, col, day, lday = UNKNOWN, month, lmonth = UNKNOWN;
!   auto     double   dur = 0.0;
    auto     char    *version;
    auto     char    *myname = basename(argv[0]);
    auto     int      opt, go, s0, indent;
*** ./isdnlog/isdnrep/isdnrep.c.old	2002-11-27 14:29:13.000000000 +0000
--- ./isdnlog/isdnrep/isdnrep.c	2002-11-27 14:29:50.000000000 +0000
***************
*** 2593,2599 ****
  {
    static char String[256];
    one_call *tmp_call;
!   int RetCode;
  
  
    if (bill)
--- 2593,2599 ----
  {
    static char String[256];
    one_call *tmp_call;
!   int RetCode = 0;
  
  
    if (bill)
*** ./isdnlog/tools/zone.c.old	2002-11-27 14:28:01.000000000 +0000
--- ./isdnlog/tools/zone.c	2002-11-27 14:28:11.000000000 +0000
***************
*** 531,537 ****
  			char *p = value.dptr;
  			char to[10];
  			US count;
! 			int ito;
  			unsigned char z=0;
  			if (sthp->cc) /* if areacodes */
  				/* here is since 1.00 a zero-terminated strring */
--- 531,537 ----
  			char *p = value.dptr;
  			char to[10];
  			US count;
! 			int ito = 0;
  			unsigned char z=0;
  			if (sthp->cc) /* if areacodes */
  				/* here is since 1.00 a zero-terminated strring */
*** ./loop/loopctrl.c.old	2002-11-27 14:32:08.000000000 +0000
--- ./loop/loopctrl.c	2002-11-27 14:32:23.000000000 +0000
***************
*** 175,181 ****
  	isdnloop_sdef startparm;
  
  	cmd = strrchr(argv[0], '/');
! 	cmd = (cmd == NULL) ? argv[0] : ++cmd;
  	if (argc > 1) {
  		if (!strcmp(argv[1], "-d")) {
  			strcpy(ioctl_s.drvid, argv[2]);
--- 175,181 ----
  	isdnloop_sdef startparm;
  
  	cmd = strrchr(argv[0], '/');
! 	cmd = (cmd == NULL) ? argv[0] : (cmd + 1);
  	if (argc > 1) {
  		if (!strcmp(argv[1], "-d")) {
  			strcpy(ioctl_s.drvid, argv[2]);
--- eicon/eiconctrl.c.orig	2006-04-19 16:16:02.000000000 +0200
+++ eicon/eiconctrl.c	2006-04-19 16:17:25.000000000 +0200
@@ -1970,7 +1970,7 @@
 				val = strtol(p, &q, 16);
 				p = q;
 				val = strtol(p, &q, 16);
-				(unsigned short) *buffer = (unsigned short) val;
+				*buffer = (unsigned short) val;
 				pos = 2;
 				while ((p != q) && (*q != 0)) {
 					p = q;
