--- isdn4k-utils-CVS-2004-11-18/vbox/configure.in.ac25x	1998-11-23 10:18:04.000000000 +0100
+++ isdn4k-utils-CVS-2004-11-18/vbox/configure.in	2004-11-18 17:47:52.638741078 +0100
@@ -17,9 +17,9 @@
 
 AC_PREFIX_DEFAULT()
 
-AC_DEFINE_UNQUOTED(PACKAGE, "${PACKAGE}")
-AC_DEFINE_UNQUOTED(VERSION, "${VERSION}")
-AC_DEFINE_UNQUOTED(VERDATE, "${VERDATE}")
+AC_DEFINE_UNQUOTED(PACKAGE, "${PACKAGE}", [package])
+AC_DEFINE_UNQUOTED(VERSION, "${VERSION}", [version])
+AC_DEFINE_UNQUOTED(VERDATE, "${VERDATE}", [verdate])
 
 AC_SUBST(PACKAGE)
 AC_SUBST(VERSION)
@@ -87,7 +87,7 @@
 
 AC_MSG_RESULT(${VBOX_SPOOLDIR})
 
-AC_DEFINE_UNQUOTED(SPOOLDIR, "${VBOX_SPOOLDIR}")
+AC_DEFINE_UNQUOTED(SPOOLDIR, "${VBOX_SPOOLDIR}", [spool dir])
 AC_SUBST(VBOX_SPOOLDIR)
 
    dnl #===============#
@@ -105,7 +105,7 @@
 
 AC_MSG_RESULT(${VBOX_LOGDIR})
 
-AC_DEFINE_UNQUOTED(LOGFILEDIR, "${VBOX_LOGDIR}")
+AC_DEFINE_UNQUOTED(LOGFILEDIR, "${VBOX_LOGDIR}", [log dir])
 AC_SUBST(VBOX_LOGDIR)
 
    dnl #===============#
@@ -123,7 +123,7 @@
 
 AC_MSG_RESULT(${VBOX_PIDDIR})
 
-AC_DEFINE_UNQUOTED(PIDFILEDIR, "${VBOX_PIDDIR}")
+AC_DEFINE_UNQUOTED(PIDFILEDIR, "${VBOX_PIDDIR}", [pid dir])
 AC_SUBST(VBOX_PIDDIR)
 
    dnl #================#
@@ -141,7 +141,7 @@
 
 AC_MSG_RESULT(${VBOX_LCKDIR})
 
-AC_DEFINE_UNQUOTED(LCKFILEDIR, "${VBOX_LCKDIR}")
+AC_DEFINE_UNQUOTED(LCKFILEDIR, "${VBOX_LCKDIR}", [lock dir])
 AC_SUBST(VBOX_LCKDIR)
 
    dnl #===================#
@@ -159,7 +159,7 @@
 
 AC_MSG_RESULT(${ISDN_GLOBAL_CONFIG})
 
-AC_DEFINE_UNQUOTED(I4LCONFDIR, "${ISDN_GLOBAL_CONFIG}")
+AC_DEFINE_UNQUOTED(I4LCONFDIR, "${ISDN_GLOBAL_CONFIG}", [global config])
 
 dnl #------------------------------------------------------------------------#
 dnl # Check and optimize compiler flags:                                     #
--- isdn4k-utils-CVS-2004-11-18/vbox/acinclude.m4.ac25x	2002-07-06 02:11:19.000000000 +0200
+++ isdn4k-utils-CVS-2004-11-18/vbox/acinclude.m4	2004-11-18 17:47:52.639740930 +0100
@@ -49,23 +49,23 @@
          cos,
          AC_CHECK_LIB(dl,
             dlerror,
-            [AC_CHECK_LIB(${gnd_1st_tcl_lib_test},
+            AC_CHECK_LIB(${gnd_1st_tcl_lib_test},
                Tcl_CreateInterp,
                LINK_TCL_LIBS="${gnd_tcl_lib_dir} -l${gnd_1st_tcl_lib_test} -lm -ldl",
-               [AC_CHECK_LIB(${gnd_2nd_tcl_lib_test},
+               AC_CHECK_LIB(${gnd_2nd_tcl_lib_test},
                   Tcl_CreateInterp,
                   LINK_TCL_LIBS="${gnd_tcl_lib_dir} -l${gnd_2nd_tcl_lib_test} -lm -ldl",
-                  [AC_CHECK_LIB(${gnd_3rd_tcl_lib_test},
+                  AC_CHECK_LIB(${gnd_3rd_tcl_lib_test},
                      Tcl_CreateInterp,
                      LINK_TCL_LIBS="${gnd_tcl_lib_dir} -l${gnd_3rd_tcl_lib_test} -lm -ldl",
                      ,
                      ${gnd_tcl_lib_dir} -lm -ldl
-                  )],
+                  ),
                   ${gnd_tcl_lib_dir} -lm -ldl
-               )],
+               ),
                ${gnd_tcl_lib_dir} -lm -ldl
-            )],
-         )],
+            ),
+         ),
       )
 
       if (test "${LINK_TCL_LIBS}" != "")
@@ -154,7 +154,7 @@
 
          AC_CHECK_LIB(ncurses,
             resizeterm,
-            AC_DEFINE(HAVE_RESIZETERM)
+            AC_DEFINE(HAVE_RESIZETERM, 1, [have resizeterm])
 )
       fi
 
--- isdn4k-utils-CVS-2004-11-18/vbox/aclocal.m4.ac25x	2002-07-06 02:11:19.000000000 +0200
+++ isdn4k-utils-CVS-2004-11-18/vbox/aclocal.m4	2004-11-18 17:47:52.640740782 +0100
@@ -61,23 +61,23 @@
          cos,
          [AC_CHECK_LIB(dl,
             dlerror,
-            [AC_CHECK_LIB(${gnd_1st_tcl_lib_test},
+            AC_CHECK_LIB(${gnd_1st_tcl_lib_test},
                Tcl_CreateInterp,
                LINK_TCL_LIBS="${gnd_tcl_lib_dir} -l${gnd_1st_tcl_lib_test} -lm -ldl",
-               [AC_CHECK_LIB(${gnd_2nd_tcl_lib_test},
+               AC_CHECK_LIB(${gnd_2nd_tcl_lib_test},
                   Tcl_CreateInterp,
                   LINK_TCL_LIBS="${gnd_tcl_lib_dir} -l${gnd_2nd_tcl_lib_test} -lm -ldl",
-                  [AC_CHECK_LIB(${gnd_3rd_tcl_lib_test},
+                  AC_CHECK_LIB(${gnd_3rd_tcl_lib_test},
                      Tcl_CreateInterp,
                      LINK_TCL_LIBS="${gnd_tcl_lib_dir} -l${gnd_3rd_tcl_lib_test} -lm -ldl",
                      ,
                      ${gnd_tcl_lib_dir} -lm -ldl
-                  )],
+                  ),
                   ${gnd_tcl_lib_dir} -lm -ldl
-               )],
+               ),
                ${gnd_tcl_lib_dir} -lm -ldl
-            )],
-         )],
+            ),
+         ),
       )
 
       if (test "${LINK_TCL_LIBS}" != "")
@@ -103,15 +103,7 @@
 
                HAVE_TCL_INCL="y"
                LINK_TCL_INCL="${gnd_tcl_inc_dir}"
-            fi
-         else
-            AC_MSG_CHECKING("for tcl header in /usr/include/tcl8.3/tcl.h")
-            if (test -e "/usr/include/tcl8.3/tcl.h")
-            then
-               AC_MSG_RESULT("yes")
-               HAVE_TCL_INCL="y"
-               LINK_TCL_INCL="-I/usr/include/tcl8.3"
-            else
+			else
                AC_MSG_RESULT("no")
             fi
          fi
--- isdn4k-utils-CVS-2004-11-18/ipppd/configure.in.ac25x	2004-08-30 16:56:36.000000000 +0200
+++ isdn4k-utils-CVS-2004-11-18/ipppd/configure.in	2004-11-18 17:55:56.899083430 +0100
@@ -17,21 +17,21 @@
 AC_PROG_CC
 
 dnl Checks for libraries.
-AC_CHECK_LIB(bsd, daemon,HAVE_LIBBSD=1; AC_DEFINE(HAVE_LIBBSD),
+AC_CHECK_LIB(bsd, daemon,HAVE_LIBBSD=1; AC_DEFINE(HAVE_LIBBSD,1,[libbsd]),
 	[AC_MSG_WARN(Could not find libbsd, build disabled)])
 dnl Replace `main' with a function in -lcrypt:
-AC_CHECK_LIB(crypt, main,HAVE_LIBCRYPT=1; AC_DEFINE(HAVE_LIBCRYPT))
-AC_CHECK_LIB(des, des_ecb_encrypt, HAVE_LIBDES=1; AC_DEFINE(HAVE_LIBDES))
-AC_CHECK_LIB(ssl, DES_ecb_encrypt, HAVE_LIBSSL=1; AC_DEFINE(HAVE_LIBSSL))
+AC_CHECK_LIB(crypt, main,HAVE_LIBCRYPT=1; AC_DEFINE(HAVE_LIBCRYPT,1,[libcrypt]))
+AC_CHECK_LIB(des, des_ecb_encrypt, HAVE_LIBDES=1; AC_DEFINE(HAVE_LIBDES,1,[libdes]))
+AC_CHECK_LIB(ssl, DES_ecb_encrypt, HAVE_LIBSSL=1; AC_DEFINE(HAVE_LIBSSL,1,[libssl]))
 
 dnl Checks for header files.
 AC_HEADER_STDC
 AC_HEADER_SYS_WAIT
 AC_CHECK_HEADERS(fcntl.h limits.h paths.h sys/file.h sys/ioctl.h sys/time.h syslog.h unistd.h)
-AC_CHECK_HEADER(shadow.h, HAVE_SHADOW_H=1; AC_DEFINE(HAVE_SHADOW_H))
-AC_CHECK_HEADER(linux/isdn_lzscomp.h, HAVE_LZSCOMP_H=1; AC_DEFINE(HAVE_LZSCOMP_H))
-AC_CHECK_HEADER(pcap-bpf.h, [HAVE_PCAP_BPF_H=1; AC_DEFINE(HAVE_PCAP_BPF_H)],
-	[AC_CHECK_HEADER(net/bpf.h, [HAVE_NET_BPF_H=1; AC_DEFINE(HAVE_NET_BPF_H)],
+AC_CHECK_HEADER(shadow.h, HAVE_SHADOW_H=1; AC_DEFINE(HAVE_SHADOW_H,1,[have shadow.h]))
+AC_CHECK_HEADER(linux/isdn_lzscomp.h, HAVE_LZSCOMP_H=1; AC_DEFINE(HAVE_LZSCOMP_H,1,[isdn_lzscomp.h]))
+AC_CHECK_HEADER(pcap-bpf.h, [HAVE_PCAP_BPF_H=1; AC_DEFINE(HAVE_PCAP_BPF_H,1,[pcap-bpf.h])],
+	[AC_CHECK_HEADER(net/bpf.h, [HAVE_NET_BPF_H=1; AC_DEFINE(HAVE_NET_BPF_H,1,[net/bpf.h])],
 		[AC_MSG_ERROR(Could not find pcap header file)])])
 
 dnl Checks for typedefs, structures, and compiler characteristics.
@@ -51,62 +51,63 @@
 AC_ARG_WITH(sbin,
 	[  --with-sbin=DIR         Set directory where ipppd is istalled. [/sbin]],
 	CONFIG_SBINDIR="${withval}"
-	AC_DEFINE(CONFIG_SBINDIR,"${withval}"),
+	AC_DEFINE(CONFIG_SBINDIR,"${withval}",[sbin dir]),
 )
 
 dnl Optional man directory
 AC_ARG_WITH(sbin,
 	[  --with-man=DIR          Set manpage directory. [/usr/man]],
 	CONFIG_MANDIR="${withval}"
-	AC_DEFINE(CONFIG_MANDIR,"${withval}"),
+	AC_DEFINE(CONFIG_MANDIR,"${withval}",[man dir]),
 )
 
 dnl Optional /var/run directory
 AC_ARG_WITH(sbin,
 	[  --with-varrun=DIR       Set directory for .pid files. [/var/run]],
 	CONFIG_RUNDIR="${withval}"
-	AC_DEFINE(CONFIG_RUNDIR,"${withval}"),
+	AC_DEFINE(CONFIG_RUNDIR,"${withval}",[/var/run dir]),
 )
 
 dnl Optional MSCHAP
 AC_ARG_ENABLE(mschap,
 	[  --enable-mschap         Enable Microsoft chap authentication [no]],
 	CONFIG_IPPPD_MSCHAP="y"
-	AC_DEFINE(CONFIG_IPPPD_MSCHAP,"y"),
+	AC_DEFINE(CONFIG_IPPPD_MSCHAP,"y",[support M$-chap]),
 )
 
 dnl Optional RADIUS
 AC_ARG_ENABLE(radius,
 	[  --enable-radius         Enable RADIUS authentication [no]],
 	CONFIG_IPPPD_RADIUS="y"
-	AC_DEFINE(CONFIG_IPPPD_RADIUS,"y"),
+	AC_DEFINE(CONFIG_IPPPD_RADIUS,"y",[support radius]),
 )
 
 dnl Optional radiusclient config file
 AC_ARG_WITH(radiusclient_config,
 	[  --with-radiusclient=FILE Set filename for radiusclient configuration],
 	RADIUS_CLIENT_CONFIG_FILE="${withval}"
-	AC_DEFINE(RADIUS_CLIENT_CONFIG_FILE,"${withval}"),
+	AC_DEFINE(RADIUS_CLIENT_CONFIG_FILE,"${withval}",[radius config]),
 )
 
 dnl Optional RADIUS_WTMP_LOGGING
 AC_ARG_ENABLE(radius_wtmp_logging,
 	[  --enable-radius-wtmp       Enable RADIUS_WTMP_LOGGING authentication [no]],
 	CONFIG_IPPPD_RADIUS_WTMP_LOGGING="y"
-	AC_DEFINE(CONFIG_IPPPD_RADIUS_WTMP_LOGGING,"y"),
+	AC_DEFINE(CONFIG_IPPPD_RADIUS_WTMP_LOGGING,"y",[radius logging]),
 )
 
 dnl Optional DEBUGGING
 AC_ARG_ENABLE(debug,
 	[  --enable-debug          Enable debugging [no]],
 	CONFIG_IPPPD_DEBUGFLAGS="-DDEBUGALL"
-	AC_DEFINE(CONFIG_IPPPD_DEBUGFLAGS,"-DDEBUGALL"),
+	AC_DEFINE(CONFIG_IPPPD_DEBUGFLAGS,"-DDEBUGALL",[options]),
 )
 
 AC_ARG_ENABLE(ippp-filter,
 	[  --enable-ippp-filter    Enable IPPP Filters (needs kernel supports) [no]],
 	CONFIG_IPPP_FILTER="y"
 	AC_DEFINE(CONFIG_IPPP_FILTER,"y"),
+	AC_DEFINE(CONFIG_IPPP_FILTER,"y", [ipppd filter]),
 )
 
 AC_SUBST(I4LVERSION)
--- isdn4k-utils-CVS-2004-11-18/isdnlog/configure.in.ac25x	2004-10-28 03:53:29.000000000 +0200
+++ isdn4k-utils-CVS-2004-11-18/isdnlog/configure.in	2004-11-18 17:47:52.642740487 +0100
@@ -79,45 +79,46 @@
 dnl To make it possible to set variables in policy.h
 dnl they have to be defined with AC_DEFINE...
 dnl
-AC_DEFINE_UNQUOTED(OLDCONFDIR,"$OLDCONFDIR")
-AC_DEFINE_UNQUOTED(OLDCONFFILE,"$OLDCONFFILE")
-AC_DEFINE_UNQUOTED(DATADIR,"$datadir")
-AC_DEFINE_UNQUOTED(SERV_PORT,$SERV_PORT)
-AC_DEFINE_UNQUOTED(USERFILE,"$USERFILE")
-AC_DEFINE_UNQUOTED(LOGFILE,"$LOGFILE")
-AC_DEFINE_UNQUOTED(CHARGEFILE,"$CHARGEFILE")
-AC_DEFINE_UNQUOTED(RELOADCMD,"$RELOADCMD")
-AC_DEFINE_UNQUOTED(STOPCMD,"$STOPCMD")
-AC_DEFINE_UNQUOTED(REBOOTCMD,"$REBOOTCMD")
+AC_DEFINE_UNQUOTED(OLDCONFDIR,"$OLDCONFDIR", [old config dir])
+AC_DEFINE_UNQUOTED(OLDCONFFILE,"$OLDCONFFILE", [old config file])
+AC_DEFINE_UNQUOTED(DATADIR,"$datadir", [data dir])
+AC_DEFINE_UNQUOTED(SERV_PORT,$SERV_PORT, [server port])
+AC_DEFINE_UNQUOTED(USERFILE,"$USERFILE", [user file])
+AC_DEFINE_UNQUOTED(LOGFILE,"$LOGFILE", [log file])
+AC_DEFINE_UNQUOTED(CHARGEFILE,"$CHARGEFILE", [charge file])
+AC_DEFINE_UNQUOTED(RELOADCMD,"$RELOADCMD", [reload command])
+AC_DEFINE_UNQUOTED(STOPCMD,"$STOPCMD", [stop command])
+AC_DEFINE_UNQUOTED(REBOOTCMD,"$REBOOTCMD", [reboot command])
+
 if test "$CONFIG_ISDN_LOG_DE" = "y" ; then
 	NATION="de"
 	NATION_MACRO="ISDN_DE"
-	AC_DEFINE(ISDN_DE)
+	AC_DEFINE(ISDN_DE, 1, [German ISDN])
 fi
 if test "$CONFIG_ISDN_LOG_AT" = "y" ; then
 	NATION="at"
 	NATION_MACRO="ISDN_AT"
-	AC_DEFINE(ISDN_AT)
+	AC_DEFINE(ISDN_AT, 1, [Austrian ISDN])
 fi
 if test "$CONFIG_ISDN_LOG_CH" = "y" ; then
 	NATION="ch"
 	NATION_MACRO="ISDN_CH"
-	AC_DEFINE(ISDN_CH)
+	AC_DEFINE(ISDN_CH, 1, [Swiss ISDN])
 fi
 if test "$CONFIG_ISDN_LOG_NL" = "y" ; then
 	NATION="nl"
 	NATION_MACRO="ISDN_NL"
-	AC_DEFINE(ISDN_NL)
+	AC_DEFINE(ISDN_NL, 1, [Dutch ISDN])
 fi
 if test "$CONFIG_ISDN_LOG_LU" = "y" ; then
 	NATION="lu"
 	NATION_MACRO="ISDN_LU"
-	AC_DEFINE(ISDN_LU)
+	AC_DEFINE(ISDN_LU, 1, [Luxembourg ISDN])
 fi
 if test "$CONFIG_ISDN_LOG_XX" = "y" ; then
 	NATION=$CONFIG_ISDN_LOG_CC
 	NATION_MACRO="ISDN_XX"
-	AC_DEFINE(ISDN_XX)
+	AC_DEFINE(ISDN_XX, 1, [Other country])
 fi
 
 dnl Checks for programs.
@@ -130,17 +131,17 @@
 
 dnl Checks for libraries.
 if test -e tools/cdb/i4l_cdb.c  ; then
-    AC_DEFINE(USE_CDB)
+    AC_DEFINE(USE_CDB, 1, [Use CDB])
     CDBEXTRALIBS="tools/cdb/i4l_cdb.a"
     DBEXT=".cdb"
     RDBEXT=".cdb"
     AC_MSG_RESULT(Using 'cdb' as database)
 else
-AC_CHECK_LIB(gdbm, gdbm_open, DBMLIB=-lgdbm; AC_DEFINE(HAVE_LIBGDBM),
-	[AC_CHECK_LIB(dbm, dbm_open, DBMLIB=-ldbm; AC_DEFINE(HAVE_LIBDBM),
-		[AC_CHECK_LIB(db, dbm_open, DBMLIB=-ldb; AC_DEFINE(HAVE_LIBDB))])])
+AC_CHECK_LIB(gdbm, gdbm_open, DBMLIB=-lgdbm; AC_DEFINE(HAVE_LIBGDBM, 1, [Use libgdbm]),
+	AC_CHECK_LIB(dbm, dbm_open, DBMLIB=-ldbm; AC_DEFINE(HAVE_LIBDBM, 1, [Use libdbm]),
+		AC_CHECK_LIB(db, dbm_open, DBMLIB=-ldb; AC_DEFINE(HAVE_LIBDB, 1, [use libdb]))))
 fi
-AC_DEFINE_UNQUOTED(RDBEXT,"$RDBEXT")
+AC_DEFINE_UNQUOTED(RDBEXT,"$RDBEXT",[rdb extension])
 
 dnl log database support
 AC_CHECK_POSTGRES
@@ -158,7 +159,7 @@
 dnl Manual config
 AC_ARG_WITH(isdnlib,
 	[  --with-isdnlib=DIR      Set isdn library [../lib]],
-	LIBISDNDIR=$withval; AC_DEFINE_UNQUOTED(LIBISDNDIR,"$withval"))
+	LIBISDNDIR=$withval; AC_DEFINE_UNQUOTED(LIBISDNDIR,"$withval",[libisdn dir]))
 AC_ARG_WITH(area-lib,
 	[  --with-area-lib=STRING  Set area library type [area]],
 	ALIB=$withval)
@@ -167,16 +168,16 @@
 	CHARGECOUNTRY=$withval)
 AC_ARG_WITH(oldconfdir,
 	[  --with-oldconfdir=DIR   Set old config directory [/etc/isdnlog]],
-	OLDCONFDIR=$withval; AC_DEFINE_UNQUOTED(OLDCONFDIR,"$withval"))
+	OLDCONFDIR=$withval; AC_DEFINE_UNQUOTED(OLDCONFDIR,"$withval", [old config dir]))
 AC_ARG_WITH(oldconf,
 	[  --with-oldconf=NAME     Set old config file name [isdnlog.conf]],
-	OLDCONFFILE=$withval; AC_DEFINE_UNQUOTED(OLDCONFFILE,"$withval"))
+	OLDCONFFILE=$withval; AC_DEFINE_UNQUOTED(OLDCONFFILE,"$withval", [old config file]))
 AC_ARG_WITH(datadir,
         [  --with-datadir=DIR      Set data directory [/usr/lib/isdn]],
-        datadir=$withval; AC_DEFINE_UNQUOTED(DATADIR,"$withval"))
+        datadir=$withval; AC_DEFINE_UNQUOTED(DATADIR,"$withval", [data dir]))
 AC_ARG_WITH(sport,
 	[  --with-sport=INT        Set server port [20011]],
-	SERV_PORT=$withval; AC_DEFINE_UNQUOTED(SERV_PORT,$withval))
+	SERV_PORT=$withval; AC_DEFINE_UNQUOTED(SERV_PORT,$withval, [server port]))
 
 dnl Checks for typedefs, structures, and compiler characteristics.
 AC_C_CONST
@@ -196,28 +197,28 @@
 
 if test "$ALIB" = "area" ; then
 	LIBAREA=1
-	AC_DEFINE(LIBAREA)
+	AC_DEFINE(LIBAREA, 1, [libarea])
 fi
 case "$CHARGECOUNTRY" in
 	DE)
 		NATION="de"
 		NATION_MACRO="ISDN_DE"
-		AC_DEFINE(ISDN_DE)
+		AC_DEFINE(ISDN_DE, 1, [German ISDN])
 		;;
 	AT)
 		NATION="at"
 		NATION_MACRO="ISDN_AT"
-		AC_DEFINE(ISDN_AT)
+		AC_DEFINE(ISDN_AT, 1, [Austrian ISDN])
 		;;
 	CH)
 		NATION="ch"
 		NATION_MACRO="ISDN_CH"
-		AC_DEFINE(ISDN_CH)
+		AC_DEFINE(ISDN_CH, 1, [Swiss ISDN])
 		;;
 	NL)
 		NATION="nl"
 		NATION_MACRO="ISDN_NL"
-		AC_DEFINE(ISDN_NL)
+		AC_DEFINE(ISDN_NL, 1, [Dutch ISDN])
 		;;
 esac
 
--- isdn4k-utils-CVS-2004-11-18/isdnlog/aclocal.m4.ac25x	2002-07-19 21:03:55.000000000 +0200
+++ isdn4k-utils-CVS-2004-11-18/isdnlog/aclocal.m4	2004-11-18 17:47:52.643740339 +0100
@@ -45,13 +45,13 @@
 	if test "$pqdir" != "no" ; then
 		AC_MSG_RESULT("yes")
 		POSTGRES=1
-		AC_DEFINE_UNQUOTED(POSTGRES,1)
+		AC_DEFINE_UNQUOTED(POSTGRES,1,[postgres support])
 	else
 		AC_MSG_RESULT("no POSTGRES DISABLED")
 		pqdir=""
 	fi
 	POSTGRESDIR="$pqdir"
-	AC_DEFINE_UNQUOTED(POSTGRESDIR,"$pqdir")
+	AC_DEFINE_UNQUOTED(POSTGRESDIR,"$pqdir",[postgres dir])
 	AC_SUBST(POSTGRES)
 	AC_SUBST(POSTGRESDIR)
 ])
@@ -108,13 +108,13 @@
 	if test "$mydir" != "no" ; then
 		AC_MSG_RESULT("yes")
 		MYSQLDB=1
-		AC_DEFINE_UNQUOTED(MYSQLDB,1)
+		AC_DEFINE_UNQUOTED(MYSQLDB,1,[MySQL support])
 	else
 		AC_MSG_RESULT("no MYSQL DISABLED")
 		mydir=""
 	fi
 	MYSQLDIR="$mydir"
-	AC_DEFINE_UNQUOTED(MYSQLDIR,"$mydir")
+	AC_DEFINE_UNQUOTED(MYSQLDIR,"$mydir",[MySQL dir])
 	AC_SUBST(MYSQLDB)
 	AC_SUBST(MYSQLDIR)
 ])
@@ -134,7 +134,7 @@
 	if test "$oradir" != "no" ; then
 		AC_MSG_RESULT("yes")
 		ORACLE=1
-		AC_DEFINE_UNQUOTED(ORACLE,1)
+		AC_DEFINE_UNQUOTED(ORACLE,1,[oracle support, unsupported])
 	else
 		AC_MSG_RESULT("no ORACLE DISABLED")
 	fi
--- isdn4k-utils-CVS-2004-11-18/vbox3/configure.in.ac25x	1998-11-10 19:36:19.000000000 +0100
+++ isdn4k-utils-CVS-2004-11-18/vbox3/configure.in	2004-11-18 17:47:52.644740191 +0100
@@ -17,11 +17,9 @@
 ## programs ###############################################################
 
 AC_PROG_CC
-AC_PROG_CC_WORKS
-AC_PROG_CC_GNU
 
 AC_PROG_RANLIB
-AM_PROG_INSTALL
+AC_PROG_INSTALL
 
 ## libraries ##############################################################
 
--- isdn4k-utils-CVS-2004-11-18/vbox3/vboxgetty/Makefile.am.ac25x	1998-11-10 19:36:25.000000000 +0100
+++ isdn4k-utils-CVS-2004-11-18/vbox3/vboxgetty/Makefile.am	2004-11-18 17:47:52.644740191 +0100
@@ -12,7 +12,7 @@
 
 ## vboxgetty ##############################################################
 
-DEFS		       += -DSYSCONFDIR='"$(sysconfdir)"' -DPKGDATADIR='"$(pkgdatadir)"' -DLOGDIR='"$(packagelogdir)"' -DLOCKDIR='"$(packagelockdir)"' -DPIDDIR='"$(packagepiddir)"'
+DEFS		        = -DSYSCONFDIR='"$(sysconfdir)"' -DPKGDATADIR='"$(pkgdatadir)"' -DLOGDIR='"$(packagelogdir)"' -DLOCKDIR='"$(packagelockdir)"' -DPIDDIR='"$(packagepiddir)"'
 
 INCLUDES		= $(all_includes) $(tcl_include)
 
--- isdn4k-utils-CVS-2004-11-18/isdnctrl/configure.in.ac25x	2002-07-19 23:23:54.000000000 +0200
+++ isdn4k-utils-CVS-2004-11-18/isdnctrl/configure.in	2004-11-18 17:47:52.645740043 +0100
@@ -30,21 +30,21 @@
 AC_ARG_WITH(sbin,
 	[  --with-sbin=DIR         Set dir where binary is istalled. [/sbin]],
 	CONFIG_SBINDIR="${withval}"
-	AC_DEFINE(CONFIG_SBINDIR,"${withval}"),
+	AC_DEFINE(CONFIG_SBINDIR,"${withval}",[sbin dir]),
 )
 
 dnl Optional man directory
 AC_ARG_WITH(man,
 	[  --with-man=DIR          Set manpage dir. [/usr/man]],
 	CONFIG_MANDIR="${withval}"
-	AC_DEFINE(CONFIG_MANDIR,"${withval}"),
+	AC_DEFINE(CONFIG_MANDIR,"${withval}",[man dir]),
 )
 
 dnl Optional config-file-option
 AC_ARG_ENABLE(cfile,
 	[  --enable-cfile          Enable configfile option [no]],
 	CONFIG_ISDNCTRL_CONF=y
-	AC_DEFINE(CONFIG_ISDNCTRL_CONF,y),
+	AC_DEFINE(CONFIG_ISDNCTRL_CONF,y,[enable config file]),
 )
 
 AC_SUBST(INSTALL)
--- isdn4k-utils-CVS-2004-11-18/imon/configure.in.ac25x	2002-07-19 21:03:53.000000000 +0200
+++ isdn4k-utils-CVS-2004-11-18/imon/configure.in	2004-11-18 17:47:52.646739895 +0100
@@ -38,14 +38,14 @@
 AC_ARG_WITH(sbin,
 	[  --with-sbin=DIR         Set dir where binary is istalled. [/sbin]],
 	CONFIG_SBINDIR="${withval}"
-	AC_DEFINE(CONFIG_SBINDIR,"${withval}"),
+	AC_DEFINE(CONFIG_SBINDIR,"${withval}",[kernel directory]),
 )
 
 dnl Optional man directory
 AC_ARG_WITH(man,
-	[  --with-man=DIR          Set manpage dir. [/usr/man]],
+	[  --with-man=DIR          Set manpage dir. [/usr/share/man]],
 	CONFIG_MANDIR="${withval}"
-	AC_DEFINE(CONFIG_MANDIR,"${withval}"),
+	AC_DEFINE(CONFIG_MANDIR,"${withval}",[man directory]),
 )
 
 AC_SUBST(INSTALL)
--- isdn4k-utils-CVS-2004-11-18/eicon/configure.in.ac25x	2002-07-19 20:42:01.000000000 +0200
+++ isdn4k-utils-CVS-2004-11-18/eicon/configure.in	2004-11-18 17:47:52.646739895 +0100
@@ -35,21 +35,21 @@
 AC_ARG_WITH(firmware,
 	[  --with-firmware=DIR     Set dir where firmware istalled. [/usr/lib/isdn]],
 	CONFIG_DATADIR="${withval}"
-	AC_DEFINE(CONFIG_DATADIR,"${withval}"),
+	AC_DEFINE(CONFIG_DATADIR,"${withval}",[data dir]),
 )
 
 dnl Optional sbin directory
 AC_ARG_WITH(sbin,
 	[  --with-sbin=DIR         Set dir where binary is istalled. [/sbin]],
 	CONFIG_SBINDIR="${withval}"
-	AC_DEFINE(CONFIG_SBINDIR,"${withval}"),
+	AC_DEFINE(CONFIG_SBINDIR,"${withval}",[sbin dir]),
 )
 
 dnl Optional man directory
 AC_ARG_WITH(man,
 	[  --with-man=DIR          Set manpage dir. [/usr/man]],
 	CONFIG_MANDIR="${withval}"
-	AC_DEFINE(CONFIG_MANDIR,"${withval}"),
+	AC_DEFINE(CONFIG_MANDIR,"${withval}",[man dir]),
 )
 
 dnl Check, if user wants dump option for debugging
@@ -58,10 +58,10 @@
 	CONFIG_EICONCTRL_DEBUG="y"
 )
 
-AC_DEFINE(HAVE_XLOG)
+AC_DEFINE(HAVE_XLOG,1,[define if you have xlog])
 AC_SUBST(HAVE_XLOG)
 
-AC_DEFINE(HAVE_TRACE)
+AC_DEFINE(HAVE_TRACE,1,[define if you have trace])
 AC_SUBST(HAVE_TRACE)
 
 AC_SUBST(HAVE_NPCI)
