diff -up jwhois-4.0/src/utils.c.orig jwhois-4.0/src/utils.c
--- jwhois-4.0/src/utils.c.orig	2010-09-29 16:19:24.453608330 +0200
+++ jwhois-4.0/src/utils.c	2010-09-29 16:20:10.686608189 +0200
@@ -292,7 +292,10 @@ make_connect(const char *host, int port)
       if (error < 0 || retval)
 	{
 	  close (sockfd);
-	  return -1;
+	  if (retval == ENETUNREACH)
+	    continue;
+	  else
+	    return -1;
 	}
 
       break;
