--- tmp/kudzu-1.2.60/Makefile	2006-09-05 15:56:21.000000000 +0200
+++ build/usr/src/kudzu-1.2.60/Makefile	2006-11-20 15:08:32.000000000 +0100
@@ -63,7 +63,7 @@
 	$(RANLIB) libkudzu.a
 
 kudzu: libkudzu.a $(KUDOBJS) po
-	$(CC) $(CFLAGS) $(LDFLAGS) $(KUDOBJS) -o kudzu -L. -lkudzu -L. -lpci -Wl,-Bstatic -lpopt -Wl,-Bdynamic
+	$(CC) $(CFLAGS) $(LDFLAGS) $(KUDOBJS) -o kudzu -L. -lkudzu -L. -lpci -L. -lintl -Wl,-Bstatic -lpopt -Wl,-Bdynamic
 
 module_upgrade: libkudzu.a module_upgrade.c
 	$(CC) $(CFLAGS) $(LDFLAGS) module_upgrade.c -o module_upgrade -L. -lkudzu -lpci
