diff -ruN kudzu-0.99.50.old/updfstab.c kudzu-0.99.50/updfstab.c
--- kudzu-0.99.50.old/updfstab.c	Mon Apr  8 05:37:01 2002
+++ kudzu-0.99.50/updfstab.c	Tue Dec 16 15:22:26 2003
@@ -688,33 +688,6 @@
 	    return 1;
 	}
 
-	/* do this before moving onto /etc/fstab; this keeps magicdev
-	   from noticing the change before the permissions are correct */
-
-	pid = fork();
-	if (!pid) {
-	    execl("/sbin/pam_console_apply", "/sbin/pam_console_apply",
-		  "-f", revokeFileName, "-r", NULL);
-	    printf("failed\n");
-	    exit(-1);
-	}
-	waitpid(pid, &status, 0);
-
-	if (!WIFEXITED(status) || WEXITSTATUS(status))
-	    fprintf(stderr, _("/sbin/pam_console_apply -r failed\n"));
-
-	pid = fork();
-	if (!pid) {
-	    execl("/sbin/pam_console_apply", "/sbin/pam_console_apply",
-		  "-f", outputFileName, NULL);
-	    printf("failed\n");
-	    exit(-1);
-	}
-
-	waitpid(pid, &status, 0);
-	if (!WIFEXITED(status) || WEXITSTATUS(status))
-	    fprintf(stderr, _("/sbin/pam_console_apply failed\n"));
-
 	if (rename(outputFileName, fileName)) {
 	    fprintf(stderr, _("failed to rename %s to %s: %s\n"),
 		    outputFileName, fileName, strerror(errno));
