diff -Naur kudzu.org/usb.c kudzu-1.2.64/usb.c
--- kudzu-1.2.64/usb.c	2006-11-28 17:58:58.000000000 +0100
+++ kudzu-1.2.64/usb.c	2007-11-29 09:24:35.000000000 +0100
@@ -268,6 +268,7 @@
 			    !strcmp(t->driver, "ax8817x") ||
 			    !strcmp(t->driver, "zd1201") ||
 			    !strcmp(t->driver, "asix") ||
+			    !strcmp(t->driver, "mcs7830") ||
 			    !strcmp(t->driver, "usbnet"))) {
 				if (t->type == CLASS_OTHER)
 					t->type = CLASS_NETWORK;
