diff -ruN libcdada-0.3.4.orig/examples/Makefile.am libcdada-0.3.4/examples/Makefile.am
--- libcdada-0.3.4.orig/examples/Makefile.am	2021-01-15 00:01:59.000000000 +0100
+++ libcdada-0.3.4/examples/Makefile.am	2021-02-28 15:41:13.940082099 +0100
@@ -11,32 +11,32 @@
 
 list_u32_SOURCES = list_u32.c
 list_u32_CPPFLAGS = -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-list_u32_LDADD = ../src/.libs/libcdada.a -lstdc++
+list_u32_LDADD = ../src/.libs/libcdada.so -lstdc++
 
 map_struct_SOURCES = map_struct.c
 map_struct_CPPFLAGS = -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-map_struct_LDADD = ../src/.libs/libcdada.a -lstdc++
+map_struct_LDADD = ../src/.libs/libcdada.so -lstdc++
 
 custom_gen_SOURCES = custom-gen/my_program_gen.c custom-gen/cdada.cc
 custom_gen_CPPFLAGS = -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-custom_gen_LDADD = ../src/.libs/libcdada.a -lstdc++
+custom_gen_LDADD = ../src/.libs/libcdada.so -lstdc++
 
 custom_macro_SOURCES = custom-macro/my_program.c custom-macro/cdada.cc
 custom_macro_CPPFLAGS = -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-custom_macro_LDADD = ../src/.libs/libcdada.a -lstdc++
+custom_macro_LDADD = ../src/.libs/libcdada.so -lstdc++
 
 str_SOURCES = str.c
 str_CPPFLAGS = -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-str_LDADD = ../src/.libs/libcdada.a -lstdc++
+str_LDADD = ../src/.libs/libcdada.so -lstdc++
 
 bbitmap_SOURCES = bbitmap.c
 bbitmap_CPPFLAGS = -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-bbitmap_LDADD = ../src/.libs/libcdada.a -lstdc++
+bbitmap_LDADD = ../src/.libs/libcdada.so -lstdc++
 
 queue_u32_SOURCES = queue_u32.c
 queue_u32_CPPFLAGS = -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-queue_u32_LDADD = ../src/.libs/libcdada.a -lstdc++
+queue_u32_LDADD = ../src/.libs/libcdada.so -lstdc++
 
 stack_u32_SOURCES = stack_u32.c
 stack_u32_CPPFLAGS = -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-stack_u32_LDADD = ../src/.libs/libcdada.a -lstdc++
+stack_u32_LDADD = ../src/.libs/libcdada.so -lstdc++
diff -ruN libcdada-0.3.4.orig/test/Makefile.am libcdada-0.3.4/test/Makefile.am
--- libcdada-0.3.4.orig/test/Makefile.am	2021-01-15 00:01:59.000000000 +0100
+++ libcdada-0.3.4/test/Makefile.am	2021-02-28 15:40:37.458117227 +0100
@@ -63,7 +63,7 @@
 benchmark_SOURCES= benchmark.cc
 benchmark_CXXFLAGS= -O3
 benchmark_CPPFLAGS= -I$(top_builddir)/include/ -I$(top_srcdir)/include/
-benchmark_LDADD= -lpthread ../src/.libs/libcdada.a
+benchmark_LDADD= -lpthread ../src/.libs/libcdada.so
 
 #Valgrind
 if VALGRIND_ENABLED
