diff -Naur libdlna-0.2.3.org/configure libdlna-0.2.3/configure
--- libdlna-0.2.3.org/configure	2007-11-26 21:47:43.000000000 +0100
+++ libdlna-0.2.3/configure	2008-12-27 22:22:18.000000000 +0100
@@ -610,9 +610,9 @@
 fi
 
 echolog "Checking for libavformat ..."
-check_lib ffmpeg/avformat.h av_register_all -lavformat || die "Error, can't find libavformat !"
+check_lib libavformat/avformat.h av_register_all -lavformat || die "Error, can't find libavformat !"
 echolog "Checking for libavcodec ..."
-check_lib ffmpeg/avcodec.h avcodec_register_all -lavcodec || die "Error, can't find libavcodec !"
+check_lib libavcodec/avcodec.h avcodec_register_all -lavcodec || die "Error, can't find libavcodec !"
 
 #################################################
 #   version
diff -Naur libdlna-0.2.3.org/src/av_mpeg4_part10.c libdlna-0.2.3/src/av_mpeg4_part10.c
--- libdlna-0.2.3.org/src/av_mpeg4_part10.c	2007-11-26 21:47:43.000000000 +0100
+++ libdlna-0.2.3/src/av_mpeg4_part10.c	2008-12-27 22:23:12.000000000 +0100
@@ -26,7 +26,7 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 
-#include <ffmpeg/avcodec.h>
+#include <libavcodec/avcodec.h>
 
 #include "dlna_internals.h"
 #include "profiles.h"
diff -Naur libdlna-0.2.3.org/src/av_mpeg4_part2.c libdlna-0.2.3/src/av_mpeg4_part2.c
--- libdlna-0.2.3.org/src/av_mpeg4_part2.c	2007-11-26 21:47:43.000000000 +0100
+++ libdlna-0.2.3/src/av_mpeg4_part2.c	2008-12-27 22:23:25.000000000 +0100
@@ -26,7 +26,7 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 
-#include <ffmpeg/avcodec.h>
+#include <libavcodec/avcodec.h>
 
 #include "dlna_internals.h"
 #include "profiles.h"
diff -Naur libdlna-0.2.3.org/src/containers.c libdlna-0.2.3/src/containers.c
--- libdlna-0.2.3.org/src/containers.c	2007-11-26 21:47:43.000000000 +0100
+++ libdlna-0.2.3/src/containers.c	2008-12-27 22:23:47.000000000 +0100
@@ -4,7 +4,7 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 
-#include <ffmpeg/avformat.h>
+#include <libavformat/avformat.h>
 
 #include "containers.h"
 #include "profiles.h"
diff -Naur libdlna-0.2.3.org/src/profiles.h libdlna-0.2.3/src/profiles.h
--- libdlna-0.2.3.org/src/profiles.h	2007-11-26 21:47:43.000000000 +0100
+++ libdlna-0.2.3/src/profiles.h	2008-12-27 22:24:47.000000000 +0100
@@ -22,8 +22,8 @@
 #ifndef _PROFILES_H_
 #define _PROFILES_H_
 
-#include <ffmpeg/avcodec.h>
-#include <ffmpeg/avformat.h>
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
 
 #include "dlna_internals.h"
 #include "containers.h"
