From b178117bac33b4b1e7ce341a6f2eec493cca13f8 Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Mon, 21 Dec 2020 16:25:46 +0000
Subject: [PATCH] location: Fix list-networks-by-as

Fixes: #12554
Signed-off-by: Michael Tremer <michael.tremer@ipfire.org>
---
 src/python/location.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/python/location.in b/src/python/location.in
index b30beae..ad2ccf5 100644
--- a/src/python/location.in
+++ b/src/python/location.in
@@ -493,7 +493,7 @@ class CLI(object):
 			f = writer(sys.stdout, prefix="AS%s" % asn)
 
 			# Print all matching networks
-			for n in db.search_networks(asn=asn, family=ns.family):
+			for n in db.search_networks(asns=[asn], family=ns.family):
 				f.write(n)
 
 			f.finish()
-- 
2.26.2

