diff --git a/src/perl/Makefile.PL b/src/perl/Makefile.PL
index 38d5390..62016ca 100644
--- a/src/perl/Makefile.PL
+++ b/src/perl/Makefile.PL
@@ -11,6 +11,7 @@ WriteMakefile(
     LIBS              => ['-lloc'],
     DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
     INC               => '-I. -I../../',
+    CCFLAGS           => '-fno-stack-protector',
 	# Un-comment this if you add C files to link with later:
     # OBJECT            => '$(O_FILES)', # link all the C files too
 );
