diff -Naur linux-3.10.30.org/sound/pci/cs5535audio/cs5535audio.c linux-3.10.30/sound/pci/cs5535audio/cs5535audio.c
--- linux-3.10.30.org/sound/pci/cs5535audio/cs5535audio.c	2014-02-13 22:48:15.000000000 +0100
+++ linux-3.10.30/sound/pci/cs5535audio/cs5535audio.c	2014-02-15 12:45:49.545688703 +0100
@@ -83,8 +83,8 @@
 			break;
 		udelay(1);
 	} while (--timeout);
-	if (!timeout)
-		snd_printk(KERN_ERR "Failure writing to cs5535 codec\n");
+//	if (!timeout)
+//		snd_printk(KERN_ERR "Failure writing to cs5535 codec\n");
 }
 
 static unsigned short snd_cs5535audio_codec_read(struct cs5535audio *cs5535au,
@@ -108,9 +108,9 @@
 			break;
 		udelay(1);
 	} while (--timeout);
-	if (!timeout)
-		snd_printk(KERN_ERR "Failure reading codec reg 0x%x,"
-					"Last value=0x%x\n", reg, val);
+//	if (!timeout)
+//		snd_printk(KERN_ERR "Failure reading codec reg 0x%x,"
+//					"Last value=0x%x\n", reg, val);
 
 	return (unsigned short) val;
 }
