diff -Naur linux-3.10.55.org/drivers/net/wireless/rt2x00/rt2800usb.c linux-3.10.55/drivers/net/wireless/rt2x00/rt2800usb.c
--- linux-3.10.55.org/drivers/net/wireless/rt2x00/rt2800usb.c	2014-09-17 18:04:18.000000000 +0200
+++ linux-3.10.55/drivers/net/wireless/rt2x00/rt2800usb.c	2014-10-05 19:25:25.070972490 +0200
@@ -128,7 +128,7 @@
 
 	tout = time_after(jiffies, entry->last_action + msecs_to_jiffies(100));
 	if (unlikely(tout))
-		rt2x00_warn(entry->queue->rt2x00dev,
+		rt2x00_dbg(entry->queue->rt2x00dev,
 			    "TX status timeout for entry %d in queue %d\n",
 			    entry->entry_idx, entry->queue->qid);
 	return tout;
@@ -569,7 +569,7 @@
 		queue = rt2x00queue_get_tx_queue(rt2x00dev, qid);
 
 		if (unlikely(rt2x00queue_empty(queue))) {
-			rt2x00_warn(rt2x00dev, "Got TX status for an empty queue %u, dropping\n",
+			rt2x00_dbg(rt2x00dev, "Got TX status for an empty queue %u, dropping\n",
 				    qid);
 			break;
 		}
@@ -578,7 +578,7 @@
 
 		if (unlikely(test_bit(ENTRY_OWNER_DEVICE_DATA, &entry->flags) ||
 			     !test_bit(ENTRY_DATA_STATUS_PENDING, &entry->flags))) {
-			rt2x00_warn(rt2x00dev, "Data pending for entry %u in queue %u\n",
+			rt2x00_dbg(rt2x00dev, "Data pending for entry %u in queue %u\n",
 				    entry->entry_idx, qid);
 			break;
 		}
