diff -Naur linux-2.6.27.25.org/drivers/usb/serial/option.c linux-2.6.27.25/drivers/usb/serial/option.c
--- linux-2.6.27.25.org/drivers/usb/serial/option.c	2009-06-12 05:04:50.000000000 +0200
+++ linux-2.6.27.25/drivers/usb/serial/option.c	2009-07-30 10:19:36.000000000 +0200
@@ -286,6 +286,12 @@
 #define ZTE_PRODUCT_MF622			0x0001
 #define ZTE_PRODUCT_MF628			0x0015
 #define ZTE_PRODUCT_MF626			0x0031
+
+/* Vodafone K3565-Z - pid depends on firmware version */
+#define ZTE_PRODUCT_VFK3565_1			0x0049
+#define ZTE_PRODUCT_VFK3565_2			0x0052
+#define ZTE_PRODUCT_VFK3565_3			0x0063
+
 #define ZTE_PRODUCT_CDMA_TECH			0xfffe
 
 #define BENQ_VENDOR_ID				0x04a5
@@ -500,6 +506,9 @@
 	{ USB_DEVICE(ZTE_VENDOR_ID, ZTE_PRODUCT_MF622) },
 	{ USB_DEVICE(ZTE_VENDOR_ID, ZTE_PRODUCT_MF626) },
 	{ USB_DEVICE(ZTE_VENDOR_ID, ZTE_PRODUCT_MF628) },
+	{ USB_DEVICE(ZTE_VENDOR_ID, ZTE_PRODUCT_VFK3565_1) },
+	{ USB_DEVICE(ZTE_VENDOR_ID, ZTE_PRODUCT_VFK3565_2) },
+	{ USB_DEVICE(ZTE_VENDOR_ID, ZTE_PRODUCT_VFK3565_3) },
 	{ USB_DEVICE(ZTE_VENDOR_ID, ZTE_PRODUCT_CDMA_TECH) },
 	{ USB_DEVICE(BENQ_VENDOR_ID, BENQ_PRODUCT_H10) },
 	{ USB_DEVICE(0x1da5, 0x4515) }, /* BenQ H20 */
