diff -Naur linux-2.6.20.21.org/drivers/net/wireless/zd1211rw/zd_usb.c linux-2.6.20.21/drivers/net/wireless/zd1211rw/zd_usb.c
--- linux-2.6.20.21.org/drivers/net/wireless/zd1211rw/zd_usb.c	2007-10-17 21:31:14.000000000 +0200
+++ linux-2.6.20.21/drivers/net/wireless/zd1211rw/zd_usb.c	2008-06-16 19:31:03.000000000 +0200
@@ -54,6 +54,7 @@
 	{ USB_DEVICE(0x14ea, 0xab13), .driver_info = DEVICE_ZD1211 },
 	/* ZD1211B */
 	{ USB_DEVICE(0x0ace, 0x1215), .driver_info = DEVICE_ZD1211B },
+	{ USB_DEVICE(0x0baf, 0x0121), .driver_info = DEVICE_ZD1211B },
 	{ USB_DEVICE(0x157e, 0x300d), .driver_info = DEVICE_ZD1211B },
 	{ USB_DEVICE(0x079b, 0x0062), .driver_info = DEVICE_ZD1211B },
 	{ USB_DEVICE(0x1582, 0x6003), .driver_info = DEVICE_ZD1211B },
