*** old/pci_ids.h	Mon Aug 25 13:44:44 2003
--- include/linux/pci_ids.h	Fri Oct 10 17:55:38 2003
***************
*** 862,865 ****
--- 862,866 ----
  #define PCI_DEVICE_ID_PLX_DJINN_ITOO	0x1151
  #define PCI_DEVICE_ID_PLX_R753		0x1152
+ #define PCI_DEVICE_ID_PLX_R753_2	0x1187
  #define PCI_DEVICE_ID_PLX_9050		0x9050
  #define PCI_DEVICE_ID_PLX_9060		0x9060
*** old/config.c	Mon Aug 25 13:44:42 2003
--- drivers/isdn/hisax/config.c	Fri Oct 10 17:56:02 2003
***************
*** 2130,2133 ****
--- 2130,2134 ----
  	{PCI_VENDOR_ID_PLX,      PCI_DEVICE_ID_PLX_R753,         PCI_ANY_ID, PCI_ANY_ID},
  	{PCI_VENDOR_ID_PLX,      PCI_DEVICE_ID_PLX_DJINN_ITOO,   PCI_ANY_ID, PCI_ANY_ID},
+ 	{PCI_VENDOR_ID_PLX,      PCI_DEVICE_ID_PLX_R753_2,       PCI_ANY_ID, PCI_ANY_ID},
  #endif
  #ifdef CONFIG_HISAX_QUADRO
*** old/gazel.c	Fri Dec 21 18:41:54 2001
--- drivers/isdn/hisax/gazel.c	Fri Oct 10 18:04:52 2003
***************
*** 559,563 ****
  	}
  	seekcard = PCI_DEVICE_ID_PLX_R685;
! 	for (nbseek = 0; nbseek < 3; nbseek++) {
  		if ((dev_tel = pci_find_device(PCI_VENDOR_ID_PLX, seekcard, dev_tel))) {
  			if (pci_enable_device(dev_tel))
--- 559,563 ----
  	}
  	seekcard = PCI_DEVICE_ID_PLX_R685;
! 	for (nbseek = 0; nbseek < 4; nbseek++) {
  		if ((dev_tel = pci_find_device(PCI_VENDOR_ID_PLX, seekcard, dev_tel))) {
  			if (pci_enable_device(dev_tel))
***************
*** 578,581 ****
--- 578,584 ----
  					seekcard = PCI_DEVICE_ID_PLX_DJINN_ITOO;
  					break;
+ 				case PCI_DEVICE_ID_PLX_DJINN_ITOO:
+ 					seekcard = PCI_DEVICE_ID_PLX_R753_2;
+ 					break;
  			}
  		}
***************
*** 621,624 ****
--- 624,635 ----
  			cs->subtyp = R753;
  			test_and_set_bit(HW_IPAC, &cs->HW_Flags);
+ 			printk(KERN_INFO
+ 			    "Gazel: config irq:%d ipac:0x%X  cfg:0x%X\n",
+ 			cs->irq, cs->hw.gazel.ipac, cs->hw.gazel.cfg_reg);
+ 			break;
+ 		case PCI_DEVICE_ID_PLX_R753_2:
+ 			printk(KERN_INFO "Gazel: Card PCI R753 0x1187 found\n");
+ 			cs->subtyp = R753;
+ 			test_and_set_bit(HW_IPAC, &cs->HW_Flags);
  			printk(KERN_INFO
  			    "Gazel: config irq:%d ipac:0x%X  cfg:0x%X\n",
