diff -Naur linux-3.10.37.org/drivers/net/wireless/rt2x00/rt2800usb.c linux-3.10.37/drivers/net/wireless/rt2x00/rt2800usb.c
--- linux-3.10.37.org/drivers/net/wireless/rt2x00/rt2800usb.c	2014-04-14 15:42:31.000000000 +0200
+++ linux-3.10.37/drivers/net/wireless/rt2x00/rt2800usb.c	2014-04-25 10:25:12.796630526 +0200
@@ -1175,6 +1175,7 @@
 	/* Belkin */
 	{ USB_DEVICE(0x050d, 0x945b) },
 	/* D-Link */
+	{ USB_DEVICE(0x2001, 0x3317) },
 	{ USB_DEVICE(0x2001, 0x3c17) },
 	/* Panasonic */
 	{ USB_DEVICE(0x083a, 0xb511) },
