diff -Naur linux-2.6.25.18.org/kernel/sysctl.c linux-2.6.25.18/kernel/sysctl.c
--- linux-2.6.25.18.org/kernel/sysctl.c	2008-10-09 04:58:32.000000000 +0200
+++ linux-2.6.25.18/kernel/sysctl.c	2008-10-10 18:03:33.000000000 +0200
@@ -2807,6 +2807,10 @@
 	if ((args->nlen == 2) && (name[0] == CTL_KERN) && (name[1] == KERN_VERSION))
 		return 0;
 
+	/* Ignore kudzu syscall 1.23 */
+	if ((args->nlen == 2) && (name[0] == 1) && (name[1] == 23))
+		return 0;
+
 	if (msg_count < 5) {
 		msg_count++;
 		printk(KERN_INFO
