diff -Naur linux-3.2.19.org/sound/arm/bcm2835-ctl.c linux-3.2.19/sound/arm/bcm2835-ctl.c
--- linux-3.2.19.org/sound/arm/bcm2835-ctl.c	2012-06-04 22:38:54.000000000 +0200
+++ linux-3.2.19/sound/arm/bcm2835-ctl.c	2012-06-05 22:37:46.343725845 +0200
@@ -53,7 +53,7 @@
 		uinfo->type = SNDRV_CTL_ELEM_TYPE_INTEGER;
 		uinfo->count = 1;
 		uinfo->value.integer.min = -10240;
-		uinfo->value.integer.max = 2303;
+		uinfo->value.integer.max = 400;      /* 2303 */
 	} else if (kcontrol->private_value == PCM_PLAYBACK_MUTE) {
 		uinfo->type = SNDRV_CTL_ELEM_TYPE_INTEGER;
 		uinfo->count = 1;
