diff -Nur lz4-1.8.1.2.orig/lib/Makefile lz4-1.8.1.2/lib/Makefile
--- lz4-1.8.1.2.orig/lib/Makefile	2018-01-21 07:24:41.881357925 +0000
+++ lz4-1.8.1.2/lib/Makefile	2018-01-21 07:25:44.823444081 +0000
@@ -120,7 +120,7 @@
 # directory variables : GNU conventions prefer lowercase
 # see https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html
 # support both lower and uppercase (BSD), use uppercase in script
-prefix      ?= /usr/local
+prefix      ?= /usr
 PREFIX      ?= $(prefix)
 exec_prefix ?= $(PREFIX)
 libdir      ?= $(exec_prefix)/lib
diff -Nur lz4-1.8.1.2.orig/NEWS lz4-1.8.1.2/NEWS
--- lz4-1.8.1.2.orig/NEWS	2018-01-21 07:24:41.809360114 +0000
+++ lz4-1.8.1.2/NEWS	2018-01-21 07:26:00.838957106 +0000
@@ -96,7 +96,7 @@
 New    : lz4 cli supports long commands (suggested by Takayuki Matsuoka)
 New    : lz4frame & lz4cli frame content size support
 New    : lz4frame supports skippable frames, as requested by Sergey Cherepanov
-Changed: Default "make install" directory is /usr/local, as notified by Ron Johnson
+Changed: Default "make install" directory is /usr, as notified by Ron Johnson
 New    : lz4 cli supports "pass-through" mode, requested by Neil Wilson
 New    : datagen can generate sparse files
 New    : scan-build tests, thanks to kind help by Takayuki Matsuoka
diff -Nur lz4-1.8.1.2.orig/programs/Makefile lz4-1.8.1.2/programs/Makefile
--- lz4-1.8.1.2.orig/programs/Makefile	2018-01-21 07:24:41.881357925 +0000
+++ lz4-1.8.1.2/programs/Makefile	2018-01-21 07:26:24.562235766 +0000
@@ -121,7 +121,7 @@
 # directory variables : GNU conventions prefer lowercase
 # see https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html
 # support both lower and uppercase (BSD), use uppercase in script
-prefix      ?= /usr/local
+prefix      ?= /usr
 PREFIX      ?= $(prefix)
 exec_prefix ?= $(PREFIX)
 bindir      ?= $(exec_prefix)/bin
