diff -au lz4-1.9.3/lib/Makefile.orig lz4-1.9.3/lib/Makefile
--- lz4-1.9.3/lib/Makefile.orig	2020-11-16 05:59:35.000000000 +0100
+++ lz4-1.9.3/lib/Makefile	2021-02-08 13:05:10.478443593 +0100
@@ -144,7 +144,7 @@
 # directory variables : GNU conventions prefer lowercase
 # see https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html
 # support both lower and uppercase (BSD), use lower in script
-PREFIX      ?= /usr/local
+PREFIX      ?= /usr
 prefix      ?= $(PREFIX)
 EXEC_PREFIX ?= $(prefix)
 exec_prefix ?= $(EXEC_PREFIX)
diff -au lz4-1.9.3/programs/Makefile.orig lz4-1.9.3/programs/Makefile
--- lz4-1.9.3/programs/Makefile.orig	2020-11-16 05:59:35.000000000 +0100
+++ lz4-1.9.3/programs/Makefile	2021-02-08 13:05:55.475676260 +0100
@@ -146,7 +146,7 @@
 # directory variables : GNU conventions prefer lowercase
 # see https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html
 # support both lower and uppercase (BSD), use lowercase in script
-PREFIX      ?= /usr/local
+PREFIX      ?= /usr
 prefix      ?= $(PREFIX)
 EXEC_PREFIX ?= $(prefix)
 exec_prefix ?= $(EXEC_PREFIX)
