diff -Nur lz4-1.9.2.orig/lib/Makefile lz4-1.9.2/lib/Makefile
--- lz4-1.9.2.orig/lib/Makefile	2019-10-29 08:14:03.167415441 +0000
+++ lz4-1.9.2/lib/Makefile	2019-10-29 08:14:43.243864994 +0000
@@ -144,7 +144,7 @@
 # directory variables : GNU conventions prefer lowercase
 # see https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html
 # support both lower and uppercase (BSD), use lower in script
-PREFIX      ?= /usr/local
+PREFIX      ?= /usr
 prefix      ?= $(PREFIX)
 EXEC_PREFIX ?= $(prefix)
 exec_prefix ?= $(EXEC_PREFIX)
diff -Nur lz4-1.9.2.orig/programs/Makefile lz4-1.9.2/programs/Makefile
--- lz4-1.9.2.orig/programs/Makefile	2019-10-29 08:14:03.177415386 +0000
+++ lz4-1.9.2/programs/Makefile	2019-10-29 08:14:30.267268628 +0000
@@ -132,7 +132,7 @@
 # directory variables : GNU conventions prefer lowercase
 # see https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html
 # support both lower and uppercase (BSD), use lowercase in script
-PREFIX      ?= /usr/local
+PREFIX      ?= /usr
 prefix      ?= $(PREFIX)
 EXEC_PREFIX ?= $(prefix)
 exec_prefix ?= $(EXEC_PREFIX)
