--- netfilter/iptcrdr.c.old	2010-02-24 14:22:23.000000000 +0100
+++ netfilter/iptcrdr.c	2010-02-24 14:48:00.000000000 +0100
@@ -20,6 +20,13 @@
 
 #if IPTABLES_143
 /* IPTABLES API version >= 1.4.3 */
+
+#define BUILD_BUG_ON_ZERO(e) (sizeof(struct { int:-!!(e); }))
+#define __must_be_array(a) \
+	BUILD_BUG_ON_ZERO(__builtin_types_compatible_p(typeof(a), typeof(&a[0])))
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]) + __must_be_array(arr))
+#define LIST_POISON2  ((void *) 0x00200200 )
+
 #include <net/netfilter/nf_nat.h>
 #define ip_nat_multi_range	nf_nat_multi_range
 #define ip_nat_range		nf_nat_range
