diff -Naur mISDN.org/drivers/isdn/hardware/mISDN/hfcpci.c mISDN/drivers/isdn/hardware/mISDN/hfcpci.c
--- mISDN.org/drivers/isdn/hardware/mISDN/hfcpci.c	2010-05-25 10:41:29.000000000 +0200
+++ mISDN/drivers/isdn/hardware/mISDN/hfcpci.c	2010-10-31 11:19:05.000000000 +0100
@@ -88,6 +88,7 @@
 	HFC_DIGI_DF_M_A,
 	HFC_ABOCOM_2BD1,
 	HFC_SITECOM_DC105V2,
+	HFC_NONAME_1AE7_0520,
 };
 
 struct hfcPCI_hw {
@@ -2182,6 +2183,7 @@
 	{HFC_DIGI_DF_M_A, 0,
 	    "Digi International DataFire Micro V (North America)"},
 	{HFC_SITECOM_DC105V2, 0, "Sitecom Connectivity DC-105 ISDN TA"},
+	{HFC_NONAME_1AE7_0520, 0, "NoName HFC-S (1AE7:0520) ISDN TA"},
 	{},
 };
 
@@ -2233,6 +2235,8 @@
 		PCI_ANY_ID, PCI_ANY_ID, 0, 0, (unsigned long) &hfc_map[21]},
 	{PCI_VENDOR_ID_SITECOM, PCI_DEVICE_ID_SITECOM_DC105V2,
 		PCI_ANY_ID, PCI_ANY_ID, 0, 0, (unsigned long) &hfc_map[22]},
+	{0x1ae7, 0x0520,
+		PCI_ANY_ID, PCI_ANY_ID, 0, 0, (unsigned long) &hfc_map[23]},
 	{},
 };
 
