diff -Naur mISDN.org/drivers/isdn/hardware/mISDN/hfcpci.c mISDN/drivers/isdn/hardware/mISDN/hfcpci.c
--- mISDN.org/drivers/isdn/hardware/mISDN/hfcpci.c	2012-10-08 14:48:40.000000000 +0200
+++ mISDN/drivers/isdn/hardware/mISDN/hfcpci.c	2012-10-08 17:09:29.527792226 +0200
@@ -53,6 +53,8 @@
 
 #include "hfc_pci.h"
 
+#define	PCI_VENDOR_ID_0x1AE7	0x1AE7
+
 static const char *hfcpci_revision = "2.0";
 
 static int HFC_cnt;
@@ -90,6 +92,7 @@
 	HFC_DIGI_DF_M_A,
 	HFC_ABOCOM_2BD1,
 	HFC_SITECOM_DC105V2,
+	HFC_NONAME_1AE7_0520,
 };
 
 struct hfcPCI_hw {
@@ -2161,6 +2164,7 @@
 	{HFC_DIGI_DF_M_A, 0,
 	 "Digi International DataFire Micro V (North America)"},
 	{HFC_SITECOM_DC105V2, 0, "Sitecom Connectivity DC-105 ISDN TA"},
+	{HFC_NONAME_1AE7_0520, 0, "NoName HFC-S (1AE7:0520) ISDN TA"},
 	{},
 };
 
@@ -2212,6 +2216,8 @@
 	  (unsigned long) &hfc_map[21] },
 	{ PCI_VDEVICE(SITECOM, PCI_DEVICE_ID_SITECOM_DC105V2),
 	  (unsigned long) &hfc_map[22] },
+	{ PCI_VDEVICE(0x1AE7, 0x0520),
+	  (unsigned long) &hfc_map[23] },
 	{},
 };
 
