diff -Naur mISDN.org/drivers/isdn/hardware/mISDN/hfcsusb.c mISDN/drivers/isdn/hardware/mISDN/hfcsusb.c
--- mISDN.org/drivers/isdn/hardware/mISDN/hfcsusb.c	2009-09-06 17:02:14.000000000 +0200
+++ mISDN/drivers/isdn/hardware/mISDN/hfcsusb.c	2010-02-06 17:44:22.000000000 +0100
@@ -654,6 +654,8 @@
 	case (1):
 		test_and_clear_bit(FLG_ACTIVE, &dch->Flags);
 		test_and_clear_bit(FLG_L2_ACTIVATED, &dch->Flags);
+		_queue_data(&dch->dev.D, PH_DEACTIVATE_IND,
+			MISDN_ID_ANY, 0, NULL, GFP_ATOMIC);
 		hw->nt_timer = 0;
 		hw->timers &= ~NT_ACTIVATION_TIMER;
 		handle_led(hw, LED_S0_OFF);
