--- avm_fritz.c.old	Tue Feb 13 11:43:45 2007
+++ avm_fritz.c	Wed Apr 11 15:26:32 2007
@@ -781,6 +781,8 @@
 			spin_lock_irqsave(inst->hwlock, flags);
 			ret = modehdlc(bch, bch->channel,
 				bch->inst.pid.protocol[1]);
+			if (bch->inst.pid.protocol[2] == ISDN_PID_L2_B_TRANS)
+				test_and_set_bit(FLG_L2DATA, &bch->Flags);
 			spin_unlock_irqrestore(inst->hwlock, flags);
 		}
 		skb_trim(skb, 0);
