diff -Naur net-tools-1.60.org/iptunnel.c net-tools-1.60/iptunnel.c
--- net-tools-1.60.org/iptunnel.c	2001-04-08 19:04:23.000000000 +0200
+++ net-tools-1.60/iptunnel.c	2017-02-07 17:37:17.956405434 +0100
@@ -26,7 +26,7 @@
 #include <sys/socket.h>
 #include <sys/ioctl.h>
 #include <netinet/in.h>
-#include <netinet/ip.h>
+#include <linux/ip.h>
 #include <arpa/inet.h>
 #if defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 1))
 #include <net/if.h>
