--- net-tools-2.10/Makefile.orig	2021-01-07 00:22:35.000000000 +0100
+++ net-tools-2.10/Makefile	2021-02-06 19:30:07.104393812 +0100
@@ -39,9 +39,6 @@
 ifeq ($(HAVE_ARP_TOOLS),1)
 PROGS	+= arp rarp
 endif
-ifeq ($(HAVE_HOSTNAME_TOOLS),1)
-PROGS	+= hostname
-endif
 ifeq ($(HAVE_IP_TOOLS),1)
 PROGS   += iptunnel ipmaddr
 endif
@@ -173,9 +170,6 @@
 nameif:		$(NET_LIB) nameif.o
 		$(CC) $(CFLAGS) $(LDFLAGS) -o $@ nameif.o $(NLIB) $(LDLIBS)
 
-hostname:	hostname.o
-		$(CC) $(CFLAGS) $(LDFLAGS) -o $@ hostname.o $(DNLIB) $(LDLIBS)
-
 route:		$(NET_LIB) route.o
 		$(CC) $(CFLAGS) $(LDFLAGS) -o $@ route.o $(NLIB) $(LDLIBS)
 
@@ -219,18 +213,6 @@
 	install -m 0755 arp        ${BASEDIR}${SBINDIR}
 	install -m 0755 rarp       ${BASEDIR}${SBINDIR}
 endif
-ifeq ($(HAVE_HOSTNAME_TOOLS),1)
-	install -m 0755 hostname   ${BASEDIR}${BINDIR}
-	ln -fs hostname $(BASEDIR)${BINDIR}/dnsdomainname
-ifeq ($(HAVE_HOSTNAME_SYMLINKS),1)
-	ln -fs hostname $(BASEDIR)${BINDIR}/ypdomainname
-	ln -fs hostname $(BASEDIR)${BINDIR}/nisdomainname
-	ln -fs hostname $(BASEDIR)${BINDIR}/domainname
-endif
-ifeq ($(HAVE_AFDECnet),1)
-	ln -fs hostname $(BASEDIR)${BINDIR}/nodename
-endif
-endif
 ifeq ($(HAVE_IP_TOOLS),1)
 	install -m 0755 ipmaddr    $(BASEDIR)${SBINDIR}
 	install -m 0755 iptunnel   $(BASEDIR)${SBINDIR}
