diff -Naur org/oinkmaster.conf new/oinkmaster.conf
--- org/oinkmaster.conf	2008-10-25 18:53:30.000000000 +0200
+++ new/oinkmaster.conf	2009-12-07 20:21:15.000000000 +0100
@@ -117,7 +117,7 @@
 # TMPDIR or TEMPDIR, or otherwise use "/tmp" if none of them was set.
 
 # Example for UNIX.
-# tmpdir = /home/oinkmaster/tmp/
+tmpdir = /var/tmp/
 
 # Example if running native Win32 or Cygwin.
 # tmpdir = c:\tmp
