diff -Naur oinkmaster-2.0.org/oinkmaster.pl oinkmaster-2.0/oinkmaster.pl
--- oinkmaster-2.0.org/oinkmaster.pl	2007-04-20 05:20:32.000000000 +0200
+++ oinkmaster-2.0/oinkmaster.pl	2013-07-15 16:46:40.000000000 +0200
@@ -1159,6 +1159,10 @@
         }
     }
 
+  # hack for community-ruleset.
+  if (-d "$dir/community-rules") {
+	move("$dir/community-rules","$dir/$rules_dir");
+  }
   # Make sure that non-empty rules directory existed in archive.
   # We permit empty rules directory if min_files is set to 0 though.
     clean_exit("$url: no \"$rules_dir\" directory found in tar file.")
