diff -ruN openswan-1.0.1.bak/pluto/server.c openswan-1.0.1/pluto/server.c
--- openswan-1.0.1.bak/pluto/server.c	2003-11-17 00:32:11.000000000 +0100
+++ openswan-1.0.1/pluto/server.c	2004-03-12 14:58:00.000000000 +0100
@@ -98,12 +98,13 @@
     else
     {
 	/* to keep control socket secure, use umask */
-	mode_t ou = umask(~S_IRWXU);
+	mode_t ou = umask(~(S_IRWXU | S_IRWXG));
 
 	if (bind(ctl_fd, (struct sockaddr *)&ctl_addr
 	, offsetof(struct sockaddr_un, sun_path) + strlen(ctl_addr.sun_path)) < 0)
 	    failed = "bind";
 	umask(ou);
+	chown(ctl_addr.sun_path, 0, 99);
     }
 
     /* 5 is a haphazardly chosen limit for the backlog.
