--- openssl-1.0.1m/ssl/ssl_lib.c.old	2015-03-19 15:56:40.966287977 +0100
+++ openssl-1.0.1m/ssl/ssl_lib.c	2015-03-19 15:57:07.976160846 +0100
@@ -1892,6 +1892,9 @@
      */
     ret->options |= SSL_OP_LEGACY_SERVER_CONNECT;
 
+    /* Disable SSLv2 and SSLv3 by default (affects the SSLv23_method() only) */
+    ret->options |= SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3;
+
     return (ret);
  err:
     SSLerr(SSL_F_SSL_CTX_NEW, ERR_R_MALLOC_FAILURE);
