diff -Naur openswan-2.6.24.org/programs/_confread/Makefile openswan-2.6.24/programs/_confread/Makefile
--- openswan-2.6.24.org/programs/_confread/Makefile	2010-01-10 02:34:38.000000000 +0100
+++ openswan-2.6.24/programs/_confread/Makefile	2010-01-10 22:44:33.000000000 +0100
@@ -16,7 +16,13 @@
 OPENSWANSRCDIR?=$(shell cd ../..; pwd)
 include ${OPENSWANSRCDIR}/Makefile.inc
 
-EXTRA5MAN=ipsec.conf.5
+SUBDIRS=doc lib programs testing
+ifeq ($(findstring $(SUBDIRS),doc),doc)
+  EXTRA5MAN=ipsec.conf.5
+else
+  EXTRA5MAN=
+endif
+
 CONFFILES=ipsec.conf
 EXCONFILES=ipsec.win2k.conf
 SOURCEDIR=_confread
diff -Naur openswan-2.6.24.org/programs/pluto/Makefile openswan-2.6.24/programs/pluto/Makefile
--- openswan-2.6.24.org/programs/pluto/Makefile	2010-01-10 02:34:38.000000000 +0100
+++ openswan-2.6.24/programs/pluto/Makefile	2010-01-10 22:51:15.000000000 +0100
@@ -172,7 +172,7 @@
 all: $(BINNAMEPLUTO) $(BINNAMEADNSIFNEEDED) $(BINNAMEWHACK) #$(BINNAMEWHACKINIT) 
 programs: $(BINNAMEPLUTO) $(BINNAMEADNSIFNEEDED) $(BINNAMEWHACK) #$(BINNAMEWHACKINIT)
 clean:	cleanall
-install: doinstall
+install: doinstall doinstalldoc
 endif
 
 doinstall: all
@@ -188,6 +188,9 @@
 	$(INSTALL) $(INSTBINFLAGS) $(BINNAMEPLUTO) $(BINNAMEWHACK) $(LIBEXECDIR)
 	#$(INSTALL) $(INSTSUIDFLAGS) $(BINNAMEWHACKINIT) $(LIBEXECDIR)
 	if $(USE_ADNS) ; then $(INSTALL) $(INSTBINFLAGS) $(BINNAMEADNS)  $(LIBEXECDIR) ; fi
+
+ifeq ($(findstring $(SUBDIRS),doc),doc)
+doinstalldoc:
 	$(INSTALL) $(INSTMANFLAGS) ${srcdir}pluto.8 $(PMANDIR)/ipsec_pluto.8
 	sh ${OPENSWANSRCDIR}/packaging/utils/manlink ${srcdir}pluto.8 | \
 		while read from to ; \
@@ -201,10 +204,7 @@
 		 ln -s -f $$from $(FMANDIR)/$$to; \
 		done
 
-install_file_list:
-	@echo $(LIBEXECDIR)/$(BINNAMEPLUTO)
-	@if $(USE_ADNS) ; then echo $(LIBDIR)/$(BINNAMEADNS) ; fi
-	@echo $(LIBEXECDIR)/$(BINNAMEWHACK)
+install_file_listdoc:
 	@echo $(PMANDIR)/ipsec_pluto.8
 	@sh ${OPENSWANSRCDIR}/packaging/utils/manlink pluto.8 | \
 		while read from to; \
@@ -217,6 +217,15 @@
 		do \
 			echo $(FMANDIR)/$$to; \
 		done
+else
+doinstalldoc:
+install_file_listdoc:
+endif
+
+install_file_list: install_file_listdoc
+	@echo $(LIBEXECDIR)/$(BINNAMEPLUTO)
+	@if $(USE_ADNS) ; then echo $(LIBDIR)/$(BINNAMEADNS) ; fi
+	@echo $(LIBEXECDIR)/$(BINNAMEWHACK)
 
 alg_info_test: alg_info_test.o ike_alg.o constants.o defs.o log.o db_ops.o crypto.o $(LIBDESLITE) $(OPENSWANLIB)
 	$(CC) -o $@ $^ $(LDFLAGS) $(LIBSPLUTO)
