--- openswan-1.0.6.old/klips/net/ipsec/ipsec_tunnel.c	2003-12-04 03:03:24.000000000 +0000
+++ openswan-1.0.6/klips/net/ipsec/ipsec_tunnel.c	2004-09-01 21:20:46.870711465 +0100
@@ -3186,6 +3186,174 @@
 	return 0;
 }
 
+static struct device dev_ipsec15 = 
+{
+	"ipsec15\0  ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec14 = 
+{
+	"ipsec14\0  ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec13 = 
+{
+	"ipsec13\0  ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec12 = 
+{
+	"ipsec12\0  ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec11 = 
+{
+	"ipsec11\0  ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec10 = 
+{
+	"ipsec10\0  ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec9 = 
+{
+	"ipsec9\0   ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec8 = 
+{
+	"ipsec8\0   ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec7 = 
+{
+	"ipsec7\0   ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec6 = 
+{
+	"ipsec6\0   ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec5 = 
+{
+	"ipsec5\0   ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
+static struct device dev_ipsec4 = 
+{
+	"ipsec4\0   ",		/* name */
+	0,			/* recv memory end */
+	0,			/* recv memory start */
+	0,			/* memory end */
+	0,			/* memory start */
+ 	0x0,			/* base I/O address */
+	0,			/* IRQ */
+	0, 0, 0,		/* flags */
+	NULL,			/* next device */
+	ipsec_tunnel_probe	/* setup */
+};
+
 static struct device dev_ipsec3 = 
 {
 	"ipsec3\0   ",		/* name */
@@ -3277,6 +3445,102 @@
 #endif
 	if (register_netdev(&dev_ipsec3) != 0)
 		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec4.name);
+#endif
+	if (register_netdev(&dev_ipsec4) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec5.name);
+#endif
+	if (register_netdev(&dev_ipsec5) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec6.name);
+#endif
+	if (register_netdev(&dev_ipsec6) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec7.name);
+#endif
+	if (register_netdev(&dev_ipsec7) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec8.name);
+#endif
+	if (register_netdev(&dev_ipsec8) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec9.name);
+#endif
+	if (register_netdev(&dev_ipsec9) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec10.name);
+#endif
+	if (register_netdev(&dev_ipsec10) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec11.name);
+#endif
+	if (register_netdev(&dev_ipsec11) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec12.name);
+#endif
+	if (register_netdev(&dev_ipsec12) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec13.name);
+#endif
+	if (register_netdev(&dev_ipsec13) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec14.name);
+#endif
+	if (register_netdev(&dev_ipsec14) != 0)
+		return -EIO;
+#if 0
+	KLIPS_PRINT(debug_tunnel & DB_TN_INIT,
+		    "klips_debug:ipsec_tunnel_init_devices: "
+		    "registering device %s\n",
+		    dev_ipsec15.name);
+#endif
+	if (register_netdev(&dev_ipsec15) != 0)
+		return -EIO;
 	return 0;
 }
 
@@ -3290,14 +3554,50 @@
 	unregister_netdev(&dev_ipsec1);
 	unregister_netdev(&dev_ipsec2);
 	unregister_netdev(&dev_ipsec3);
+	unregister_netdev(&dev_ipsec4);
+	unregister_netdev(&dev_ipsec5);
+	unregister_netdev(&dev_ipsec6);
+	unregister_netdev(&dev_ipsec7);
+	unregister_netdev(&dev_ipsec8);
+	unregister_netdev(&dev_ipsec9);
+	unregister_netdev(&dev_ipsec10);
+	unregister_netdev(&dev_ipsec11);
+	unregister_netdev(&dev_ipsec12);
+	unregister_netdev(&dev_ipsec13);
+	unregister_netdev(&dev_ipsec14);
+	unregister_netdev(&dev_ipsec15);
 	kfree(dev_ipsec0.priv);
 	kfree(dev_ipsec1.priv);
 	kfree(dev_ipsec2.priv);
 	kfree(dev_ipsec3.priv);
+	kfree(dev_ipsec4.priv);
+	kfree(dev_ipsec5.priv);
+	kfree(dev_ipsec6.priv);
+	kfree(dev_ipsec7.priv);
+	kfree(dev_ipsec8.priv);
+	kfree(dev_ipsec9.priv);
+	kfree(dev_ipsec10.priv);
+	kfree(dev_ipsec11.priv);
+	kfree(dev_ipsec12.priv);
+	kfree(dev_ipsec13.priv);
+	kfree(dev_ipsec14.priv);
+	kfree(dev_ipsec15.priv);
 	dev_ipsec0.priv=NULL;
 	dev_ipsec1.priv=NULL;
 	dev_ipsec2.priv=NULL;
 	dev_ipsec3.priv=NULL;
+	dev_ipsec4.priv=NULL;
+	dev_ipsec5.priv=NULL;
+	dev_ipsec6.priv=NULL;
+	dev_ipsec7.priv=NULL;
+	dev_ipsec8.priv=NULL;
+	dev_ipsec9.priv=NULL;
+	dev_ipsec10.priv=NULL;
+	dev_ipsec11.priv=NULL;
+	dev_ipsec12.priv=NULL;
+	dev_ipsec13.priv=NULL;
+	dev_ipsec14.priv=NULL;
+	dev_ipsec15.priv=NULL;
 
 	return error;
 }
--- openswan-1.0.6.old/klips/net/ipsec/ipsec_tunnel.h	2002-09-05 05:13:17.000000000 +0200
+++ openswan-1.0.6/klips/net/ipsec/ipsec_tunnel.h	2004-08-25 12:30:01.000000000 +0200
@@ -98,7 +98,7 @@
 	int mtu;	/* What is the desired MTU? */
 };
 
-#define IPSEC_NUM_IF	4
+#define IPSEC_NUM_IF	16
 
 extern char ipsec_tunnel_c_version[];
 
