--- /usr/lib/ipsec/_updown	2007-10-14 00:56:15.000000000 +0200
+++ /usr/lib/ipsec/_updown	2007-10-27 00:00:26.000000000 +0200
@@ -376,8 +376,8 @@
 		# opportunistic encryption work around
 		# need to provide route that eclipses default, without 
 		# replacing it.
-		it="ip route $1 0.0.0.0/1 $parms2 $parms3 &&
-			ip route $1 128.0.0.0/1 $parms2 $parms3"
+		#it="ip route $1 0.0.0.0/1 $parms2 $parms3 &&
+	#		ip route $1 128.0.0.0/1 $parms2 $parms3"
 		;;
 	*)	it="ip route $1 $parms $parms2 $parms3"
 		;;
@@ -401,13 +401,13 @@
 prepare-host:*|prepare-client:*)
 	# delete possibly-existing route (preliminary to adding a route)
 	case "$PLUTO_PEER_CLIENT" in
-	"0.0.0.0/0")
+  "0.0.0.0/0")
 		# need to provide route that eclipses default, without 
 		# replacing it.
 		parms1="0.0.0.0/1"
 		parms2="128.0.0.0/1"
-		it="ip route delete $parms1 $IPROUTEARGS 2>&1 ; ip route delete $parms2 $IPROUTEARGS 2>&1"
-		oops="`ip route delete $parms1 $IPROUTEARGS 2>&1 ; ip route delete $parms2 $IPROUTEARGS 2>&1`"
+	#	it="ip route delete $parms1 $IPROUTEARGS 2>&1 ; ip route delete $parms2 $IPROUTEARGS 2>&1"
+	#	oops="`ip route delete $parms1 $IPROUTEARGS 2>&1 ; ip route delete $parms2 $IPROUTEARGS 2>&1`"
 		;;
 	*)
 		parms="$PLUTO_PEER_CLIENT $IPROUTEARGS"
