diff -Naur openswan-2.6.16.org/linux/net/ipsec/pfkey_v2_build.c openswan-2.6.16/linux/net/ipsec/pfkey_v2_build.c
--- openswan-2.6.16.org/linux/net/ipsec/pfkey_v2_build.c	2008-08-18 19:07:48.000000000 +0200
+++ openswan-2.6.16/linux/net/ipsec/pfkey_v2_build.c	2008-09-13 15:38:43.000000000 +0200
@@ -442,6 +442,7 @@
 	}
 
 	switch(address->sa_family) {
+	case AF_UNSPEC:
 	case AF_INET:
 		DEBUGGING(PF_KEY_DEBUG_BUILD,
 			"pfkey_address_build: "
