From 0d6d6fd2b89c609a6f8daff225c6c98b54a041cc Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Mon, 4 Mar 2013 17:18:52 +0100
Subject: [PATCH] database: Add inst_size column when creating new databases.

---
 python/pakfire/repository/database.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/python/pakfire/repository/database.py b/python/pakfire/repository/database.py
index fc8a1a9..d8751e9 100644
--- a/python/pakfire/repository/database.py
+++ b/python/pakfire/repository/database.py
@@ -190,6 +190,7 @@ class DatabaseLocal(Database):
 				groups		TEXT,
 				filename	TEXT,
 				size		INTEGER,
+				inst_size	INTEGER,
 				hash1		TEXT,
 				provides	TEXT,
 				requires	TEXT,
-- 
1.7.11.7

