diff --git a/Makefile b/Makefile
index 0361af6..0bcb1bf 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 
 include Makeconfig
 
-SUBDIRS = po python tools tools/fake-environ
+SUBDIRS = python tools tools/fake-environ
 
 all: build
 
diff --git a/tools/Makefile b/tools/Makefile
index d7cbf70..499beb4 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -12,7 +12,7 @@ LIBS_POPT = -lpopt
 
 SCRIPTS = $(SCRIPTS_BIN) $(SCRIPTS_SHELL)
 SCRIPTS_BIN   = \
-	debugedit
+	#debugedit
 
 SCRIPTS_SHELL = \
 	chroot-shell \
@@ -50,7 +50,7 @@ install: $(SCRIPTS)
 
 	# Create script aliases.
 	-mkdir -pv $(DESTDIR)/usr/bin
-	ln -svf ../..$(SCRIPT_DIR)/pakfire-multicall.py $(DESTDIR)/usr/bin/pakfire
+	ln -svf ../..$(SCRIPT_DIR)/pakfire-multicall.py $(DESTDIR)/usr/bin/pakfire3
 	ln -svf ../..$(SCRIPT_DIR)/pakfire-multicall.py $(DESTDIR)/usr/bin/pakfire-builder
 	ln -svf ../..$(SCRIPT_DIR)/pakfire-multicall.py $(DESTDIR)/usr/bin/pakfire-client
 	ln -svf ../..$(SCRIPT_DIR)/pakfire-multicall.py $(DESTDIR)/usr/bin/pakfire-daemon
