--- pciutils-2.2.1/Makefile.idpath	2006-02-23 12:24:12.000000000 +0100
+++ pciutils-2.2.1/Makefile	2006-02-23 12:24:18.000000000 +0100
@@ -10,7 +10,7 @@
 PREFIX=/usr/local
 SBINDIR=$(PREFIX)/sbin
 SHAREDIR=$(PREFIX)/share
-IDSDIR=$(SHAREDIR)
+IDSDIR=$(SHAREDIR)/hwdata
 MANDIR:=$(shell if [ -d $(PREFIX)/share/man ] ; then echo $(PREFIX)/share/man ; else echo $(PREFIX)/man ; fi)
 INSTALL=install
 DIRINSTALL=install -d
