--- pciutils-2.1.99-test3/Makefile.strip	2004-02-25 01:46:14.315787866 -0500
+++ pciutils-2.1.99-test3/Makefile	2004-02-25 01:47:45.478046260 -0500
@@ -32,7 +32,7 @@
 all: $(PCILIB) lspci setpci lspci.8 setpci.8 update-pciids update-pciids.8 pci.ids
 
 $(PCILIB): $(PCIINC) force
-	$(MAKE) -C lib all
+	CFLAGS="$(CFLAGS) -fPIC" $(MAKE) -C lib all
 
 force:
 
--- pciutils-2.1.99-test8/Makefile.foo	2005-05-10 15:24:45.000000000 -0400
+++ pciutils-2.1.99-test8/Makefile	2005-05-10 15:24:50.000000000 -0400
@@ -65,7 +65,7 @@
 install: all
 # -c is ignored on Linux, but required on FreeBSD
 	$(DIRINSTALL) -m 755 $(SBINDIR) $(IDSDIR) $(MANDIR)/man8
-	$(INSTALL) -c -m 755 -s lspci setpci $(SBINDIR)
+	$(INSTALL) -c -m 755 lspci setpci $(SBINDIR)
 	$(INSTALL) -c -m 755 update-pciids $(SBINDIR)
 	$(INSTALL) -c -m 644 pci.ids $(IDSDIR)
 	$(INSTALL) -c -m 644 lspci.8 setpci.8 update-pciids.8 $(MANDIR)/man8
