--- perl-5.8.4/lib/CGI.pm.old	2004-03-11 20:55:39.000000000 +0000
+++ perl-5.8.4/lib/CGI.pm	2003-09-30 18:10:42.000000000 +0100
@@ -328,20 +328,9 @@
   return $self;
 }
 
-# We provide a DESTROY method so that we can ensure that
-# temporary files are closed (via Fh->DESTROY) before they
-# are unlinked (via CGITempFile->DESTROY) because it is not
-# possible to unlink an open file on Win32. We explicitly
-# call DESTROY on each, rather than just undefing them and
-# letting Perl DESTROY them by garbage collection, in case the
-# user is still holding any reference to them as well.
-sub DESTROY {
-  my $self = shift;
-  foreach my $href (values %{$self->{'.tmpfiles'}}) {
-    $href->{hndl}->DESTROY if defined $href->{hndl};
-    $href->{name}->DESTROY if defined $href->{name};
-  }
-}
+# We provide a DESTROY method so that the autoloader
+# doesn't bother trying to find it.
+sub DESTROY { }
 
 sub r {
   my $self = shift;
