diff -ur perl-5.8.5.orig/lib/ExtUtils/instmodsh perl-5.8.5/lib/ExtUtils/instmodsh
--- perl-5.8.5.orig/lib/ExtUtils/instmodsh	Tue Sep 30 17:10:47 2003
+++ perl-5.8.5/lib/ExtUtils/instmodsh	Mon Jan 19 20:14:11 2004
@@ -2,6 +2,7 @@
 
 use strict;
 use IO::File;
+use File::Temp;
 use ExtUtils::Packlist;
 use ExtUtils::Installed;
 
@@ -58,16 +59,12 @@
       $reply =~ /^t\s*/ and do
          {
          my $file = (split(' ', $reply))[1];
-         my $tmp = "/tmp/inst.$$";
-         if (my $fh = IO::File->new($tmp, "w"))
-            {
-            $fh->print(join("\n", $Inst->files($module)));
-            $fh->close();
-            system("tar cvf $file -I $tmp");
-            unlink($tmp);
-            last CASE;
-            }
-         else { print("Can't open $file: $!\n"); }
+         my ($fh, $tmp) = File::Temp::tempfile(UNLINK => 1);
+         $fh->print(join("\n", $Inst->files($module)));
+         $fh->close();
+         # This used to use -I which is wrong for GNU tar.
+         system("tar cvf $file -T $tmp");
+         unlink($tmp);
          last CASE;
          };
       $reply eq 'v' and do

diff -ur perl-5.8.5.orig/lib/perl5db.pl perl-5.8.5/lib/perl5db.pl
--- perl-5.8.5.orig/lib/perl5db.pl	Mon Jan 19 18:46:25 2004
+++ perl-5.8.5/lib/perl5db.pl	Mon Jan 19 20:14:11 2004
@@ -206,7 +206,7 @@
 =item * noTTY 
 
 if set, goes in NonStop mode.  On interrupt, if TTY is not set,
-uses the value of noTTY or "/tmp/perldbtty$$" to find TTY using
+uses the value of noTTY or "/var/run/perldbtty$$" to find TTY using
 Term::Rendezvous.  Current variant is to have the name of TTY in this
 file.
 
@@ -5689,8 +5689,8 @@
         else {
             eval "require Term::Rendezvous;" or die;
             # See if we have anything to pass to Term::Rendezvous.
-            # Use /tmp/perldbtty$$ if not.
-            my $rv = $ENV{PERLDB_NOTTY} || "/tmp/perldbtty$$";
+            # Use /var/run/perldbtty$$ if not.
+            my $rv = $ENV{PERLDB_NOTTY} || "/var/run/perldbtty$$";
 
             # Rendezvous and get the filehandles.
             my $term_rv = new Term::Rendezvous $rv;


