diff -ur ppp-2.4.2.old/pppd/main.c ppp-2.4.2/pppd/main.c
--- ppp-2.4.2.old/pppd/main.c	2004-01-13 04:00:34.000000000 +0000
+++ ppp-2.4.2/pppd/main.c	2004-05-24 19:05:34.000000000 +0100
@@ -1662,8 +1648,11 @@
 
     /* Ensure that nothing of our device environment is inherited. */
     closelog();
+    /* Some plugins dont have a close function, so just close the devfd */
     if (the_channel->close)
 	(*the_channel->close)();
+    else
+	close(devfd);
 
     /* Don't pass handles to the PPP device, even by accident. */
     dup2(fd_devnull, 0);
