diff -Nur src/bindings/pool.i sat-solver-bindings/bindings/pool.i
--- src/bindings/pool.i        2011-04-25 12:03:31.000000000 +0000
+++ sat-solver-bindings/bindings/pool.i        2011-04-29 18:40:59.645929695 +0000
@@ -474,6 +474,7 @@
     return repo;
   }
 
+#if defined(FEDORA)
   /*
    * Add RPM database to Pool.
    *
@@ -493,6 +494,7 @@
     repo_add_rpmdb( repo, NULL, rootdir, 0 );
     return repo;
   }
+#endif
 
   %newobject create_repo;
   /*
diff -Nur src/bindings/repo.i sat-solver-bindings/bindings/repo.i
--- src/bindings/repo.i        2011-04-25 12:03:31.000000000 +0000
+++ sat-solver-bindings/bindings/repo.i        2011-04-29 18:37:13.856570319 +0000
@@ -139,11 +139,13 @@
     }
   }
 
+#if defined(FEDORA)
   /*
    * Add RPM database, optionally passing a _root_ directory
    */
   void add_rpmdb( const char *rootdir )
   { repo_add_rpmdb( $self, NULL, rootdir, 0); }
+#endif
 
   /*
    * Create solvable with +name+ and +evr+ in the Repo

