diff -Naur qemu-kvm-0.15.0.org/hw/9pfs/virtio-9p-local.c qemu-kvm-0.15.0/hw/9pfs/virtio-9p-local.c
--- qemu-kvm-0.15.0.org/hw/9pfs/virtio-9p-local.c	2011-08-09 14:40:29.000000000 +0200
+++ qemu-kvm-0.15.0/hw/9pfs/virtio-9p-local.c	2011-08-14 10:31:22.711480316 +0200
@@ -21,6 +21,16 @@
 #include <sys/un.h>
 #include <attr/xattr.h>
 
+#ifndef AT_FDCWD
+/* Copied from linux/include/linux/fcntl.h * because direct include fails */
+#define AT_FDCWD		-100    /* Special value used to indicate
+                                           openat should use the current
+                                           working directory. */
+#define AT_SYMLINK_NOFOLLOW	0x100   /* Do not follow symbolic links.  */
+#define AT_REMOVEDIR		0x200   /* Remove directory instead of
+                                           unlinking file.  */
+#define AT_SYMLINK_FOLLOW	0x400   /* Follow symbolic links.  */
+#endif
 
 static int local_lstat(FsContext *fs_ctx, const char *path, struct stat *stbuf)
 {
diff -Naur qemu-kvm-0.15.0.org/linux-user/syscall.c qemu-kvm-0.15.0/linux-user/syscall.c
--- qemu-kvm-0.15.0.org/linux-user/syscall.c	2011-08-09 14:40:29.000000000 +0200
+++ qemu-kvm-0.15.0/linux-user/syscall.c	2011-08-14 12:43:43.190231600 +0200
@@ -971,6 +971,16 @@
     return result;
 }
 
+/* Copied from linux/include/asm/resource.h * because direct include fails */
+
+#ifndef RLIMIT_NICE
+#define RLIMIT_NICE		13	/* max nice prio allowed to raise to
+					   0-39 for nice level 19 .. -20 */
+#endif
+#ifndef RLIMIT_RTPRIO
+#define RLIMIT_RTPRIO		14	/* maximum realtime priority */
+#endif
+
 static inline int target_to_host_resource(int code)
 {
     switch (code) {
