diff -Naur qemu-0.11.1.org/hw/pc.c qemu-0.11.1/hw/pc.c
--- qemu-0.11.1.org/hw/pc.c	2009-12-02 21:27:02.000000000 +0100
+++ qemu-0.11.1/hw/pc.c	2010-12-27 12:09:41.845206649 +0100
@@ -43,7 +43,7 @@
 /* Show multiboot debug output */
 //#define DEBUG_MULTIBOOT
 
-#define BIOS_FILENAME "bios.bin"
+#define BIOS_FILENAME "bios-kqemu.bin"
 #define VGABIOS_FILENAME "vgabios.bin"
 #define VGABIOS_CIRRUS_FILENAME "vgabios-cirrus.bin"
 
