diff -Naur qemu-0.14.1.org/hw/virtio-9p-local.c qemu-0.14.1/hw/virtio-9p-local.c
--- qemu-0.14.1.org/hw/virtio-9p-local.c	2011-05-06 21:01:43.000000000 +0200
+++ qemu-0.14.1/hw/virtio-9p-local.c	2011-05-18 14:04:32.432444320 +0200
@@ -10,6 +10,7 @@
  * the COPYING file in the top-level directory.
  *
  */
+
 #include "virtio.h"
 #include "virtio-9p.h"
 #include "virtio-9p-xattr.h"
@@ -20,6 +21,16 @@
 #include <sys/un.h>
 #include <attr/xattr.h>
 
+#ifndef AT_FDCWD
+/* Copied from linux/include/linux/fcntl.h * because direct include fails */
+#define AT_FDCWD		-100    /* Special value used to indicate
+                                           openat should use the current
+                                           working directory. */
+#define AT_SYMLINK_NOFOLLOW	0x100   /* Do not follow symbolic links.  */
+#define AT_REMOVEDIR		0x200   /* Remove directory instead of
+                                           unlinking file.  */
+#define AT_SYMLINK_FOLLOW	0x400   /* Follow symbolic links.  */
+#endif
 
 static int local_lstat(FsContext *fs_ctx, const char *path, struct stat *stbuf)
 {
