diff -Naur r8169-6.011.00.org/src/r8169_n.c r8169-6.011.00/src/r8169_n.c
--- r8169-6.011.00.org/src/r8169_n.c	2009-07-14 08:39:25.000000000 +0200
+++ r8169-6.011.00/src/r8169_n.c	2009-10-21 23:24:26.000000000 +0200
@@ -114,6 +114,11 @@
 static struct pci_device_id rtl8169_pci_tbl[] = {
 	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8167), 0, 0, RTL_CFG_0 },
 	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8169), 0, 0, RTL_CFG_0 },
+	{ PCI_DEVICE(PCI_VENDOR_ID_REALTEK,	0x8129), 0, 0, RTL_CFG_0 },
+	{ 0x1186,0x4300, 0, 0, RTL_CFG_0 },
+	{ 0x1737,0x1032, 0, 0, RTL_CFG_0 },
+	{ 0x1259,0xC107, 0, 0, RTL_CFG_0 },
+	{ 0x16EC,0x8129, 0, 0, RTL_CFG_0 },
 	{0,},
 };
 
