--- getrandom-0.2.4/Cargo.toml.orig	2022-01-27 18:43:09.339254197 +0000
+++ getrandom-0.2.4/Cargo.toml	2022-01-27 18:44:42.583701271 +0000
@@ -20,33 +20,13 @@
 [target.'cfg(unix)'.dependencies]
 libc = { version = "0.2.64", default-features = false }
 
-[target.'cfg(target_os = "wasi")'.dependencies]
-wasi = "0.10"
-
-[target.'cfg(all(target_arch = "wasm32", target_os = "unknown"))'.dependencies]
-wasm-bindgen = { version = "0.2.62", default-features = false, optional = true }
-js-sys = { version = "0.3", optional = true }
-[target.'cfg(all(target_arch = "wasm32", target_os = "unknown"))'.dev-dependencies]
-wasm-bindgen-test = "0.3.18"
-
 [features]
 # Implement std-only traits for getrandom::Error
 std = []
 # Feature to enable fallback RDRAND-based implementation on x86/x86_64
 rdrand = []
-# Feature to enable JavaScript bindings on wasm32-unknown-unknown
-js = ["wasm-bindgen", "js-sys"]
 # Feature to enable custom RNG implementations
 custom = []
-# Unstable feature to support being a libstd dependency
-rustc-dep-of-std = [
-  "compiler_builtins",
-  "core",
-  "libc/rustc-dep-of-std",
-  "wasi/rustc-dep-of-std",
-]
-# Unstable/test-only feature to run wasm-bindgen tests in a browser
-test-in-browser = []
 
 [package.metadata.docs.rs]
 features = ["std", "custom"]
