diff -up screen-4.0.3/tty.sh.stropts screen-4.0.3/tty.sh
--- screen-4.0.3/tty.sh.stropts	2003-09-08 16:24:25.000000000 +0200
+++ screen-4.0.3/tty.sh	2008-04-07 11:28:34.000000000 +0200
@@ -76,7 +76,7 @@ exit 0
 #endif
 
 #include "config.h"
-#ifdef SVR4
+#if 0
 #include <sys/stropts.h>	/* for I_POP */
 #endif
 
diff -up screen-4.0.3/screen.c.stropts screen-4.0.3/screen.c
--- screen-4.0.3/screen.c.stropts	2008-04-07 11:25:21.000000000 +0200
+++ screen-4.0.3/screen.c	2008-04-07 11:29:14.000000000 +0200
@@ -50,7 +50,7 @@
 
 #include "config.h"
 
-#ifdef SVR4
+#if 0
 # include <sys/stropts.h>
 #endif
 
diff -up screen-4.0.3/process.c.stropts screen-4.0.3/process.c
--- screen-4.0.3/process.c.stropts	2003-09-18 14:53:54.000000000 +0200
+++ screen-4.0.3/process.c	2008-04-07 11:29:47.000000000 +0200
@@ -37,7 +37,7 @@
 #include "config.h"
 
 /* for solaris 2.1, Unixware (SVR4.2) and possibly others: */
-#ifdef SVR4
+#if 0 
 # include <sys/stropts.h>
 #endif
 
diff -up screen-4.0.3/pty.c.stropts screen-4.0.3/pty.c
--- screen-4.0.3/pty.c.stropts	2003-09-08 16:26:18.000000000 +0200
+++ screen-4.0.3/pty.c	2008-04-07 11:30:07.000000000 +0200
@@ -34,7 +34,7 @@
 #endif
 
 /* for solaris 2.1, Unixware (SVR4.2) and possibly others */
-#ifdef HAVE_SVR4_PTYS
+#if 0 
 # include <sys/stropts.h>
 #endif
 
