diff -Naur shadow-4.14.2.orig/man/cs/Makefile.in shadow-4.14.2/man/cs/Makefile.in
--- shadow-4.14.2.orig/man/cs/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/cs/Makefile.in	2023-11-09 14:39:12.397603428 +0100
@@ -334,8 +334,8 @@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 man_MANS = man1/expiry.1 man5/faillog.5 man8/faillog.8 man1/gpasswd.1 \
-	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/groups.1 \
-	man8/grpck.8 man5/gshadow.5 man8/nologin.8 man5/passwd.5 \
+	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/ \
+	man8/grpck.8 man5/gshadow.5 man8/nologin.8 man5/ \
 	man5/shadow.5 man1/su.1 man8/vipw.8 $(am__append_1)
 EXTRA_DIST = $(man_MANS) \
 	man1/id.1 \
diff -Naur shadow-4.14.2.orig/man/da/Makefile.in shadow-4.14.2/man/da/Makefile.in
--- shadow-4.14.2.orig/man/da/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/da/Makefile.in	2023-11-09 14:40:35.104051500 +0100
@@ -337,7 +337,7 @@
 top_srcdir = @top_srcdir@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
-man_MANS = man1/chfn.1 man8/groupdel.8 man1/groups.1 man5/gshadow.5 \
+man_MANS = man1/chfn.1 man8/groupdel.8 man1/ man5/gshadow.5 \
 	man8/logoutd.8 man1/newgrp.1 man8/nologin.8 man1/sg.1 \
 	man8/vigr.8 man8/vipw.8 $(am__append_1)
 man_nopam = 
diff -Naur shadow-4.14.2.orig/man/de/Makefile.in shadow-4.14.2/man/de/Makefile.in
--- shadow-4.14.2.orig/man/de/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/de/Makefile.in	2023-11-09 14:58:08.953606477 +0100
@@ -339,11 +339,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man1/login.1 \
 	man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 man8/newusers.8 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man3/shadow.3 \
 	man5/shadow.5 man1/su.1 man5/suauth.5 man8/useradd.8 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
diff -Naur shadow-4.14.2.orig/man/fr/Makefile.in shadow-4.14.2/man/fr/Makefile.in
--- shadow-4.14.2.orig/man/fr/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/fr/Makefile.in	2023-11-09 14:57:48.282831779 +0100
@@ -341,11 +341,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man1/login.1 \
 	man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 man8/newusers.8 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man3/shadow.3 \
 	man5/shadow.5 man1/su.1 man5/suauth.5 man8/useradd.8 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
diff -Naur shadow-4.14.2.orig/man/hu/Makefile.in shadow-4.14.2/man/hu/Makefile.in
--- shadow-4.14.2.orig/man/hu/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/hu/Makefile.in	2023-11-09 14:44:53.850447743 +0100
@@ -333,8 +333,8 @@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-man_MANS = man1/chsh.1 man1/gpasswd.1 man1/groups.1 man1/login.1 \
-	man1/newgrp.1 man1/passwd.1 man5/passwd.5 man1/sg.1 man1/su.1 \
+man_MANS = man1/chsh.1 man1/gpasswd.1 man1/ man1/login.1 \
+	man1/newgrp.1 man1/passwd.1 man5/ man1/sg.1 man1/su.1 \
 	$(am__append_1)
 EXTRA_DIST = $(man_MANS)
 all: all-am
diff -Naur shadow-4.14.2.orig/man/it/Makefile.in shadow-4.14.2/man/it/Makefile.in
--- shadow-4.14.2.orig/man/it/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/it/Makefile.in	2023-11-09 14:57:24.455272085 +0100
@@ -339,11 +339,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man1/login.1 \
 	man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 man8/newusers.8 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man3/shadow.3 \
 	man5/shadow.5 man1/su.1 man5/suauth.5 man8/useradd.8 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
diff -Naur shadow-4.14.2.orig/man/ja/Makefile.in shadow-4.14.2/man/ja/Makefile.in
--- shadow-4.14.2.orig/man/ja/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/ja/Makefile.in	2023-11-09 14:56:59.957687244 +0100
@@ -337,10 +337,10 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chpasswd.8 man1/chsh.1 \
 	man1/expiry.1 man5/faillog.5 man8/faillog.8 man1/gpasswd.1 \
-	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/groups.1 \
+	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/ \
 	man8/grpck.8 man8/grpconv.8 man8/grpunconv.8 man1/login.1 \
 	man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 man8/newusers.8 \
-	man1/passwd.1 man5/passwd.5 man8/pwck.8 man8/pwconv.8 \
+	man1/passwd.1 man5/ man8/pwck.8 man8/pwconv.8 \
 	man8/pwunconv.8 man1/sg.1 man5/shadow.5 man1/su.1 \
 	man5/suauth.5 man8/useradd.8 man8/userdel.8 man8/usermod.8 \
 	man8/vigr.8 man8/vipw.8 $(am__append_1) $(am__append_2)
diff -Naur shadow-4.14.2.orig/man/ko/Makefile.in shadow-4.14.2/man/ko/Makefile.in
--- shadow-4.14.2.orig/man/ko/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/ko/Makefile.in	2023-11-09 14:47:00.355198614 +0100
@@ -335,9 +335,9 @@
 man_MANS = \
 	man1/chfn.1 \
 	man1/chsh.1 \
-	man1/groups.1 \
+	man1/ \
 	man1/login.1 \
-	man5/passwd.5 \
+	man5/ \
 	man1/su.1 \
 	man8/vigr.8 \
 	man8/vipw.8
diff -Naur shadow-4.14.2.orig/man/Makefile.in shadow-4.14.2/man/Makefile.in
--- shadow-4.14.2.orig/man/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/Makefile.in	2023-11-09 14:56:20.699549105 +0100
@@ -406,11 +406,11 @@
 @USE_NLS_TRUE@SUBDIRS = po cs da de es fi fr hu id it ja ko pl pt_BR ru sv tr uk zh_CN zh_TW
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man1/login.1 \
 	man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 man8/newusers.8 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man3/shadow.3 \
 	man5/shadow.5 man5/suauth.5 man8/useradd.8 man8/userdel.8 \
 	man8/usermod.8 man8/vigr.8 man8/vipw.8 $(am__append_1) \
diff -Naur shadow-4.14.2.orig/man/pl/Makefile.in shadow-4.14.2/man/pl/Makefile.in
--- shadow-4.14.2.orig/man/pl/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/pl/Makefile.in	2023-11-09 14:49:03.766497242 +0100
@@ -340,8 +340,8 @@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
 man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
-	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/logoutd.8 man1/newgrp.1 man1/sg.1 man3/shadow.3 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
 	$(am__append_1) $(am__append_2)
diff -Naur shadow-4.14.2.orig/man/pt_BR/Makefile.in shadow-4.14.2/man/pt_BR/Makefile.in
--- shadow-4.14.2.orig/man/pt_BR/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/pt_BR/Makefile.in	2023-11-09 14:49:55.465103649 +0100
@@ -337,7 +337,7 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5
 
 EXTRA_DIST = $(man_MANS)
diff -Naur shadow-4.14.2.orig/man/ru/Makefile.in shadow-4.14.2/man/ru/Makefile.in
--- shadow-4.14.2.orig/man/ru/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/ru/Makefile.in	2023-11-09 14:51:11.867969686 +0100
@@ -339,11 +339,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man1/login.1 \
 	man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 man8/newusers.8 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man3/shadow.3 \
 	man5/shadow.5 man1/su.1 man5/suauth.5 man8/useradd.8 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
diff -Naur shadow-4.14.2.orig/man/sv/Makefile.in shadow-4.14.2/man/sv/Makefile.in
--- shadow-4.14.2.orig/man/sv/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/sv/Makefile.in	2023-11-09 14:52:06.113337499 +0100
@@ -340,10 +340,10 @@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
 man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
-	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man5/gshadow.5 man8/logoutd.8 man1/newgrp.1 man8/nologin.8 \
-	man1/passwd.1 man5/passwd.5 man8/pwck.8 man1/sg.1 \
+	man1/passwd.1 man5/ man8/pwck.8 man1/sg.1 \
 	man3/shadow.3 man5/suauth.5 man8/userdel.8 man8/vigr.8 \
 	man8/vipw.8 $(am__append_1) $(am__append_2)
 man_nopam = \
diff -Naur shadow-4.14.2.orig/man/tr/Makefile.in shadow-4.14.2/man/tr/Makefile.in
--- shadow-4.14.2.orig/man/tr/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/tr/Makefile.in	2023-11-09 14:52:37.971198864 +0100
@@ -340,7 +340,7 @@
 	man8/groupmod.8 \
 	man1/login.1 \
 	man1/passwd.1 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5 \
 	man1/su.1 \
 	man8/useradd.8 \
diff -Naur shadow-4.14.2.orig/man/zh_CN/Makefile.in shadow-4.14.2/man/zh_CN/Makefile.in
--- shadow-4.14.2.orig/man/zh_CN/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/zh_CN/Makefile.in	2023-11-09 14:53:19.266080649 +0100
@@ -339,11 +339,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man1/login.1 \
 	man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 man8/newusers.8 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man3/shadow.3 \
 	man5/shadow.5 man1/su.1 man5/suauth.5 man8/useradd.8 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
diff -Naur shadow-4.14.2.orig/man/zh_TW/Makefile.in shadow-4.14.2/man/zh_TW/Makefile.in
--- shadow-4.14.2.orig/man/zh_TW/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/man/zh_TW/Makefile.in	2023-11-09 14:53:45.617068672 +0100
@@ -342,7 +342,7 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man5/passwd.5 \
+	man5/ \
 	man1/su.1 \
 	man8/useradd.8 \
 	man8/userdel.8 \
diff -Naur shadow-4.14.2.orig/src/Makefile.in shadow-4.14.2/src/Makefile.in
--- shadow-4.14.2.orig/src/Makefile.in	2023-10-30 02:10:52.000000000 +0100
+++ shadow-4.14.2/src/Makefile.in	2023-11-09 14:55:13.373692050 +0100
@@ -88,8 +88,7 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-bin_PROGRAMS = groups$(EXEEXT) login$(EXEEXT) $(am__EXEEXT_1) \
-	$(am__EXEEXT_2)
+bin_PROGRAMS = login$(EXEEXT) $(am__EXEEXT_1)
 sbin_PROGRAMS = nologin$(EXEEXT)
 ubin_PROGRAMS = faillog$(EXEEXT) chage$(EXEEXT) chfn$(EXEEXT) \
 	chsh$(EXEEXT) expiry$(EXEEXT) gpasswd$(EXEEXT) newgrp$(EXEEXT) \
