diff -Naur shadow-4.13.orig/man/cs/Makefile.in shadow-4.13/man/cs/Makefile.in
--- shadow-4.13.orig/man/cs/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/cs/Makefile.in	2023-01-07 23:37:45.110080801 +0100
@@ -332,12 +332,12 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man1/groups.1 \
+	man1/ \
 	man8/grpck.8 \
 	man5/gshadow.5 \
 	man8/lastlog.8 \
 	man8/nologin.8 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5 \
 	man1/su.1 \
 	man8/vipw.8
diff -Naur shadow-4.13.orig/man/da/Makefile.in shadow-4.13/man/da/Makefile.in
--- shadow-4.13.orig/man/da/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/da/Makefile.in	2023-01-07 23:39:23.687584267 +0100
@@ -329,7 +329,7 @@
 top_srcdir = @top_srcdir@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
-man_MANS = man1/chfn.1 man8/groupdel.8 man1/groups.1 man5/gshadow.5 \
+man_MANS = man1/chfn.1 man8/groupdel.8 man1/ man5/gshadow.5 \
 	man8/logoutd.8 man1/newgrp.1 man8/nologin.8 man1/sg.1 \
 	man8/vigr.8 man8/vipw.8 $(am__append_1)
 man_nopam = 
diff -Naur shadow-4.13.orig/man/de/Makefile.in shadow-4.13/man/de/Makefile.in
--- shadow-4.13.orig/man/de/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/de/Makefile.in	2023-01-07 23:41:28.644419661 +0100
@@ -330,11 +330,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.13.orig/man/fr/Makefile.in shadow-4.13/man/fr/Makefile.in
--- shadow-4.13.orig/man/fr/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/fr/Makefile.in	2023-01-08 12:12:06.805231749 +0100
@@ -332,11 +332,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+  man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.13.orig/man/hu/Makefile.in shadow-4.13/man/hu/Makefile.in
--- shadow-4.13.orig/man/hu/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/hu/Makefile.in	2023-01-08 12:12:58.220574975 +0100
@@ -327,12 +327,12 @@
 man_MANS = \
 	man1/chsh.1 \
 	man1/gpasswd.1 \
-	man1/groups.1 \
+	man1/ \
 	man8/lastlog.8 \
 	man1/login.1 \
 	man1/newgrp.1 \
 	man1/passwd.1 \
-	man5/passwd.5 \
+	man5/ \
 	man1/sg.1 \
 	man1/su.1
 
diff -Naur shadow-4.13.orig/man/it/Makefile.in shadow-4.13/man/it/Makefile.in
--- shadow-4.13.orig/man/it/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/it/Makefile.in	2023-01-08 12:14:02.598918966 +0100
@@ -330,11 +330,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.13.orig/man/ja/Makefile.in shadow-4.13/man/ja/Makefile.in
--- shadow-4.13.orig/man/ja/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/ja/Makefile.in	2023-01-08 12:14:59.754116344 +0100
@@ -328,10 +328,10 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chpasswd.8 man1/chsh.1 \
 	man1/expiry.1 man5/faillog.5 man8/faillog.8 man1/gpasswd.1 \
-	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/groups.1 \
+	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/ \
 	man8/grpck.8 man8/grpconv.8 man8/grpunconv.8 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/newusers.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man5/shadow.5 \
 	man1/su.1 man5/suauth.5 man8/useradd.8 man8/userdel.8 \
 	man8/usermod.8 man8/vigr.8 man8/vipw.8 $(am__append_1)
diff -Naur shadow-4.13.orig/man/ko/Makefile.in shadow-4.13/man/ko/Makefile.in
--- shadow-4.13.orig/man/ko/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/ko/Makefile.in	2023-01-08 12:15:43.959647149 +0100
@@ -327,9 +327,9 @@
 man_MANS = \
 	man1/chfn.1 \
 	man1/chsh.1 \
-	man1/groups.1 \
+	man1/ \
 	man1/login.1 \
-	man5/passwd.5 \
+	man5/ \
 	man1/su.1 \
 	man8/vigr.8 \
 	man8/vipw.8
diff -Naur shadow-4.13.orig/man/Makefile.in shadow-4.13/man/Makefile.in
--- shadow-4.13.orig/man/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/Makefile.in	2023-01-08 12:17:08.914132063 +0100
@@ -396,11 +396,11 @@
 @USE_NLS_TRUE@SUBDIRS = po cs da de es fi fr hu id it ja ko pl pt_BR ru sv tr uk zh_CN zh_TW
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man5/suauth.5 man8/useradd.8 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
diff -Naur shadow-4.13.orig/man/pl/Makefile.in shadow-4.13/man/pl/Makefile.in
--- shadow-4.13.orig/man/pl/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/pl/Makefile.in	2023-01-08 12:18:16.339068110 +0100
@@ -331,8 +331,8 @@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
 man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
-	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 man1/sg.1 \
 	man3/shadow.3 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
 	man8/vipw.8 $(am__append_1)
diff -Naur shadow-4.13.orig/man/pt_BR/Makefile.in shadow-4.13/man/pt_BR/Makefile.in
--- shadow-4.13.orig/man/pt_BR/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/pt_BR/Makefile.in	2023-01-08 12:18:46.844952016 +0100
@@ -329,7 +329,7 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5
 
 EXTRA_DIST = $(man_MANS)
diff -Naur shadow-4.13.orig/man/ru/Makefile.in shadow-4.13/man/ru/Makefile.in
--- shadow-4.13.orig/man/ru/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/ru/Makefile.in	2023-01-08 12:20:05.019613396 +0100
@@ -330,11 +330,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.13.orig/man/sv/Makefile.in shadow-4.13/man/sv/Makefile.in
--- shadow-4.13.orig/man/sv/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/sv/Makefile.in	2023-01-08 12:21:14.481165613 +0100
@@ -331,10 +331,10 @@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
 man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
-	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man5/gshadow.5 man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man1/sg.1 man3/shadow.3 man5/suauth.5 man8/userdel.8 \
 	man8/vigr.8 man8/vipw.8 $(am__append_1)
 man_nopam = \
diff -Naur shadow-4.13.orig/man/tr/Makefile.in shadow-4.13/man/tr/Makefile.in
--- shadow-4.13.orig/man/tr/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/tr/Makefile.in	2023-01-08 12:21:46.353681866 +0100
@@ -332,7 +332,7 @@
 	man8/groupmod.8 \
 	man1/login.1 \
 	man1/passwd.1 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5 \
 	man1/su.1 \
 	man8/useradd.8 \
diff -Naur shadow-4.13.orig/man/zh_CN/Makefile.in shadow-4.13/man/zh_CN/Makefile.in
--- shadow-4.13.orig/man/zh_CN/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/zh_CN/Makefile.in	2023-01-08 12:22:45.888822579 +0100
@@ -330,11 +330,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.13.orig/man/zh_TW/Makefile.in shadow-4.13/man/zh_TW/Makefile.in
--- shadow-4.13.orig/man/zh_TW/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/man/zh_TW/Makefile.in	2023-01-08 12:23:20.354606005 +0100
@@ -334,7 +334,7 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man5/passwd.5 \
+	man5/ \
 	man1/su.1 \
 	man8/useradd.8 \
 	man8/userdel.8 \
diff -Naur shadow-4.13.orig/src/Makefile.in shadow-4.13/src/Makefile.in
--- shadow-4.13.orig/src/Makefile.in	2022-11-08 17:37:10.000000000 +0100
+++ shadow-4.13/src/Makefile.in	2023-01-08 12:24:35.952669040 +0100
@@ -88,8 +88,7 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-bin_PROGRAMS = groups$(EXEEXT) login$(EXEEXT) $(am__EXEEXT_1) \
-	$(am__EXEEXT_2)
+bin_PROGRAMS = login$(EXEEXT) $(am__EXEEXT_1)
 sbin_PROGRAMS = nologin$(EXEEXT)
 ubin_PROGRAMS = faillog$(EXEEXT) lastlog$(EXEEXT) chage$(EXEEXT) \
 	chfn$(EXEEXT) chsh$(EXEEXT) expiry$(EXEEXT) gpasswd$(EXEEXT) \
