diff -Naur shadow-4.11.1-orig/man/cs/Makefile.in shadow-4.11.1/man/cs/Makefile.in
--- shadow-4.11.1-orig/man/cs/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/cs/Makefile.in	2022-01-17 13:04:13.758312001 +0100
@@ -326,12 +326,12 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man1/groups.1 \
+	man1/ \
 	man8/grpck.8 \
 	man5/gshadow.5 \
 	man8/lastlog.8 \
 	man8/nologin.8 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5 \
 	man1/su.1 \
 	man8/vipw.8
diff -Naur shadow-4.11.1-orig/man/da/Makefile.in shadow-4.11.1/man/da/Makefile.in
--- shadow-4.11.1-orig/man/da/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/da/Makefile.in	2022-01-17 13:05:29.513200259 +0100
@@ -323,7 +323,7 @@
 top_srcdir = @top_srcdir@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
-man_MANS = man1/chfn.1 man8/groupdel.8 man1/groups.1 man5/gshadow.5 \
+man_MANS = man1/chfn.1 man8/groupdel.8 man1/ man5/gshadow.5 \
 	man8/logoutd.8 man1/newgrp.1 man8/nologin.8 man1/sg.1 \
 	man8/vigr.8 man8/vipw.8 $(am__append_1)
 man_nopam = 
diff -Naur shadow-4.11.1-orig/man/de/Makefile.in shadow-4.11.1/man/de/Makefile.in
--- shadow-4.11.1-orig/man/de/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/de/Makefile.in	2022-01-17 13:06:43.188118562 +0100
@@ -324,11 +324,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.11.1-orig/man/es/Makefile.in shadow-4.11.1/man/es/Makefile.in
--- shadow-4.11.1-orig/man/es/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/es/Makefile.in	2022-01-17 13:07:24.967505172 +0100
@@ -292,7 +292,7 @@
 #	man1/login.1 \
 #	man1/newgrp.1 \
 #	man1/passwd.1 \
-#	man5/passwd.5 \
+#	man5/ \
 #	man1/su.1 \
 #	man8/vigr.8 \
 #	man8/vipw.8
diff -Naur shadow-4.11.1-orig/man/fr/Makefile.in shadow-4.11.1/man/fr/Makefile.in
--- shadow-4.11.1-orig/man/fr/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/fr/Makefile.in	2022-01-17 13:08:52.850214886 +0100
@@ -326,11 +326,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+    man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.11.1-orig/man/hu/Makefile.in shadow-4.11.1/man/hu/Makefile.in
--- shadow-4.11.1-orig/man/hu/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/hu/Makefile.in	2022-01-17 13:09:43.917465155 +0100
@@ -321,12 +321,12 @@
 man_MANS = \
 	man1/chsh.1 \
 	man1/gpasswd.1 \
-	man1/groups.1 \
+	man1/ \
 	man8/lastlog.8 \
 	man1/login.1 \
 	man1/newgrp.1 \
 	man1/passwd.1 \
-	man5/passwd.5 \
+	man5/ \
 	man1/sg.1 \
 	man1/su.1
 
diff -Naur shadow-4.11.1-orig/man/it/Makefile.in shadow-4.11.1/man/it/Makefile.in
--- shadow-4.11.1-orig/man/it/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/it/Makefile.in	2022-01-17 13:10:43.664587974 +0100
@@ -324,11 +324,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.11.1-orig/man/ja/Makefile.in shadow-4.11.1/man/ja/Makefile.in
--- shadow-4.11.1-orig/man/ja/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/ja/Makefile.in	2022-01-17 13:11:27.991937191 +0100
@@ -322,10 +322,10 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chpasswd.8 man1/chsh.1 \
 	man1/expiry.1 man5/faillog.5 man8/faillog.8 man1/gpasswd.1 \
-	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/groups.1 \
+	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/ \
 	man8/grpck.8 man8/grpconv.8 man8/grpunconv.8 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/newusers.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man5/shadow.5 \
 	man1/su.1 man5/suauth.5 man8/useradd.8 man8/userdel.8 \
 	man8/usermod.8 man8/vigr.8 man8/vipw.8 $(am__append_1)
diff -Naur shadow-4.11.1-orig/man/ko/Makefile.in shadow-4.11.1/man/ko/Makefile.in
--- shadow-4.11.1-orig/man/ko/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/ko/Makefile.in	2022-01-17 13:12:18.603194118 +0100
@@ -321,9 +321,9 @@
 man_MANS = \
 	man1/chfn.1 \
 	man1/chsh.1 \
-	man1/groups.1 \
+	man1/ \
 	man1/login.1 \
-	man5/passwd.5 \
+	man5/ \
 	man1/su.1 \
 	man8/vigr.8 \
 	man8/vipw.8
diff -Naur shadow-4.11.1-orig/man/Makefile.in shadow-4.11.1/man/Makefile.in
--- shadow-4.11.1-orig/man/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/Makefile.in	2022-01-17 13:13:43.341950016 +0100
@@ -392,11 +392,11 @@
 @USE_NLS_TRUE@SUBDIRS = po cs da de es fi fr hu id it ja ko pl pt_BR ru sv tr zh_CN zh_TW
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man5/suauth.5 man8/useradd.8 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
diff -Naur shadow-4.11.1-orig/man/pl/Makefile.in shadow-4.11.1/man/pl/Makefile.in
--- shadow-4.11.1-orig/man/pl/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/pl/Makefile.in	2022-01-17 13:14:42.781077363 +0100
@@ -325,8 +325,8 @@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
 man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
-	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 man1/sg.1 \
 	man3/shadow.3 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
 	man8/vipw.8 $(am__append_1)
diff -Naur shadow-4.11.1-orig/man/pt_BR/Makefile.in shadow-4.11.1/man/pt_BR/Makefile.in
--- shadow-4.11.1-orig/man/pt_BR/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/pt_BR/Makefile.in	2022-01-17 13:15:09.668682617 +0100
@@ -323,7 +323,7 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5
 
 EXTRA_DIST = $(man_MANS)
diff -Naur shadow-4.11.1-orig/man/ru/Makefile.in shadow-4.11.1/man/ru/Makefile.in
--- shadow-4.11.1-orig/man/ru/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/ru/Makefile.in	2022-01-17 13:16:07.647831387 +0100
@@ -324,11 +324,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.11.1-orig/man/sv/Makefile.in shadow-4.11.1/man/sv/Makefile.in
--- shadow-4.11.1-orig/man/sv/Makefile.in	2022-01-03 04:09:23.000000000 +0100
+++ shadow-4.11.1/man/sv/Makefile.in	2022-01-17 13:16:58.503084748 +0100
@@ -325,10 +325,10 @@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
 man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
-	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man5/gshadow.5 man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man1/sg.1 man3/shadow.3 man5/suauth.5 man8/userdel.8 \
 	man8/vigr.8 man8/vipw.8 $(am__append_1)
 man_nopam = \
diff -Naur shadow-4.11.1-orig/man/tr/Makefile.in shadow-4.11.1/man/tr/Makefile.in
--- shadow-4.11.1-orig/man/tr/Makefile.in	2022-01-03 04:09:24.000000000 +0100
+++ shadow-4.11.1/man/tr/Makefile.in	2022-01-17 13:17:28.478644677 +0100
@@ -326,7 +326,7 @@
 	man8/groupmod.8 \
 	man1/login.1 \
 	man1/passwd.1 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5 \
 	man1/su.1 \
 	man8/useradd.8 \
diff -Naur shadow-4.11.1-orig/man/zh_CN/Makefile.in shadow-4.11.1/man/zh_CN/Makefile.in
--- shadow-4.11.1-orig/man/zh_CN/Makefile.in	2022-01-03 04:09:24.000000000 +0100
+++ shadow-4.11.1/man/zh_CN/Makefile.in	2022-01-17 13:18:26.633790862 +0100
@@ -324,11 +324,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.11.1-orig/man/zh_TW/Makefile.in shadow-4.11.1/man/zh_TW/Makefile.in
--- shadow-4.11.1-orig/man/zh_TW/Makefile.in	2022-01-03 04:09:24.000000000 +0100
+++ shadow-4.11.1/man/zh_TW/Makefile.in	2022-01-17 13:18:55.225371102 +0100
@@ -328,7 +328,7 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man5/passwd.5 \
+	man5/ \
 	man1/su.1 \
 	man8/useradd.8 \
 	man8/userdel.8 \
diff -Naur shadow-4.11.1-orig/src/Makefile.in shadow-4.11.1/src/Makefile.in
--- shadow-4.11.1-orig/src/Makefile.in	2022-01-03 04:09:24.000000000 +0100
+++ shadow-4.11.1/src/Makefile.in	2022-01-17 13:20:19.360135889 +0100
@@ -88,8 +88,7 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-bin_PROGRAMS = groups$(EXEEXT) login$(EXEEXT) $(am__EXEEXT_1) \
-	$(am__EXEEXT_2)
+bin_PROGRAMS = login$(EXEEXT) $(am__EXEEXT_1)
 sbin_PROGRAMS = nologin$(EXEEXT)
 ubin_PROGRAMS = faillog$(EXEEXT) lastlog$(EXEEXT) chage$(EXEEXT) \
 	chfn$(EXEEXT) chsh$(EXEEXT) expiry$(EXEEXT) gpasswd$(EXEEXT) \
