diff -Naur shadow-4.9-orig/man/cs/Makefile.in shadow-4.9/man/cs/Makefile.in
--- shadow-4.9-orig/man/cs/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/cs/Makefile.in	2021-09-28 10:02:20.088234227 +0200
@@ -326,12 +326,12 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man1/groups.1 \
+	man1/ \
 	man8/grpck.8 \
 	man5/gshadow.5 \
 	man8/lastlog.8 \
 	man8/nologin.8 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5 \
 	man1/su.1 \
 	man8/vipw.8
diff -Naur shadow-4.9-orig/man/da/Makefile.in shadow-4.9/man/da/Makefile.in
--- shadow-4.9-orig/man/da/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/da/Makefile.in	2021-09-28 10:03:35.455124215 +0200
@@ -323,7 +323,7 @@
 top_srcdir = @top_srcdir@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
-man_MANS = man1/chfn.1 man8/groupdel.8 man1/groups.1 man5/gshadow.5 \
+man_MANS = man1/chfn.1 man8/groupdel.8 man1/ man5/gshadow.5 \
 	man8/logoutd.8 man1/newgrp.1 man8/nologin.8 man1/sg.1 \
 	man8/vigr.8 man8/vipw.8 $(am__append_1)
 man_nopam = 
diff -Naur shadow-4.9-orig/man/de/Makefile.in shadow-4.9/man/de/Makefile.in
--- shadow-4.9-orig/man/de/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/de/Makefile.in	2021-09-28 10:06:00.828982852 +0200
@@ -324,11 +324,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+ 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+ 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.9-orig/man/es/Makefile.in shadow-4.9/man/es/Makefile.in
--- shadow-4.9-orig/man/es/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/es/Makefile.in	2021-09-28 10:06:40.528417341 +0200
@@ -292,7 +292,7 @@
 #	man1/login.1 \
 #	man1/newgrp.1 \
 #	man1/passwd.1 \
-#	man5/passwd.5 \
+#	man5/ \
 #	man1/su.1 \
 #	man8/vigr.8 \
 #	man8/vipw.8
diff -Naur shadow-4.9-orig/man/fr/Makefile.in shadow-4.9/man/fr/Makefile.in
--- shadow-4.9-orig/man/fr/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/fr/Makefile.in	2021-09-28 12:11:23.391355328 +0200
@@ -326,11 +326,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+ 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+ 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+ 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.9-orig/man/hu/Makefile.in shadow-4.9/man/hu/Makefile.in
--- shadow-4.9-orig/man/hu/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/hu/Makefile.in	2021-09-28 12:12:10.834664816 +0200
@@ -321,12 +321,12 @@
 man_MANS = \
 	man1/chsh.1 \
 	man1/gpasswd.1 \
-	man1/groups.1 \
+	man1/ \
 	man8/lastlog.8 \
 	man1/login.1 \
 	man1/newgrp.1 \
 	man1/passwd.1 \
-	man5/passwd.5 \
+	man5/ \
 	man1/sg.1 \
 	man1/su.1
 
diff -Naur shadow-4.9-orig/man/it/Makefile.in shadow-4.9/man/it/Makefile.in
--- shadow-4.9-orig/man/it/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/it/Makefile.in	2021-09-28 12:13:14.321740885 +0200
@@ -324,11 +324,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+ 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+ 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+ 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.9-orig/man/ja/Makefile.in shadow-4.9/man/ja/Makefile.in
--- shadow-4.9-orig/man/ja/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/ja/Makefile.in	2021-09-28 12:14:06.920974945 +0200
@@ -322,10 +322,10 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chpasswd.8 man1/chsh.1 \
 	man1/expiry.1 man5/faillog.5 man8/faillog.8 man1/gpasswd.1 \
-	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/groups.1 \
+ 	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/ \
 	man8/grpck.8 man8/grpconv.8 man8/grpunconv.8 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+ 	man8/newusers.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man5/shadow.5 \
 	man1/su.1 man5/suauth.5 man8/useradd.8 man8/userdel.8 \
 	man8/usermod.8 man8/vigr.8 man8/vipw.8 $(am__append_1)
diff -Naur shadow-4.9-orig/man/ko/Makefile.in shadow-4.9/man/ko/Makefile.in
--- shadow-4.9-orig/man/ko/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/ko/Makefile.in	2021-09-28 12:14:40.432487050 +0200
@@ -321,9 +321,9 @@
 man_MANS = \
 	man1/chfn.1 \
 	man1/chsh.1 \
-	man1/groups.1 \
+	man1/ \
 	man1/login.1 \
-	man5/passwd.5 \
+	man5/ \
 	man1/su.1 \
 	man8/vigr.8 \
 	man8/vipw.8
diff -Naur shadow-4.9-orig/man/Makefile.in shadow-4.9/man/Makefile.in
--- shadow-4.9-orig/man/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/Makefile.in	2021-09-28 12:15:46.395526512 +0200
@@ -392,11 +392,11 @@
 @USE_NLS_TRUE@SUBDIRS = po cs da de es fi fr hu id it ja ko pl pt_BR ru sv tr zh_CN zh_TW
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+ 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+ 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+ 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man5/suauth.5 man8/useradd.8 \
 	man8/userdel.8 man8/usermod.8 man8/vigr.8 man8/vipw.8 \
diff -Naur shadow-4.9-orig/man/pl/Makefile.in shadow-4.9/man/pl/Makefile.in
--- shadow-4.9-orig/man/pl/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/pl/Makefile.in	2021-09-28 12:16:46.974644527 +0200
@@ -325,8 +325,8 @@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
 man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
-	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+ 	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
+ 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 man1/sg.1 \
 	man3/shadow.3 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
 	man8/vipw.8 $(am__append_1)
diff -Naur shadow-4.9-orig/man/pt_BR/Makefile.in shadow-4.9/man/pt_BR/Makefile.in
--- shadow-4.9-orig/man/pt_BR/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/pt_BR/Makefile.in	2021-09-28 12:17:15.610227402 +0200
@@ -323,7 +323,7 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5
 
 EXTRA_DIST = $(man_MANS)
diff -Naur shadow-4.9-orig/man/ru/Makefile.in shadow-4.9/man/ru/Makefile.in
--- shadow-4.9-orig/man/ru/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/ru/Makefile.in	2021-09-28 12:18:14.697366956 +0200
@@ -324,11 +324,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+ 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+ 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+ 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.9-orig/man/sv/Makefile.in shadow-4.9/man/sv/Makefile.in
--- shadow-4.9-orig/man/sv/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/sv/Makefile.in	2021-09-28 12:19:19.328425616 +0200
@@ -325,10 +325,10 @@
 
 # 2012.01.28 - activate manpages with more than 50% translated messages
 man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
-	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+ 	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
+ 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man5/gshadow.5 man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 \
-	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
+ 	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
 	man1/sg.1 man3/shadow.3 man5/suauth.5 man8/userdel.8 \
 	man8/vigr.8 man8/vipw.8 $(am__append_1)
 man_nopam = \
diff -Naur shadow-4.9-orig/man/tr/Makefile.in shadow-4.9/man/tr/Makefile.in
--- shadow-4.9-orig/man/tr/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/tr/Makefile.in	2021-09-28 12:19:52.315944994 +0200
@@ -326,7 +326,7 @@
 	man8/groupmod.8 \
 	man1/login.1 \
 	man1/passwd.1 \
-	man5/passwd.5 \
+	man5/ \
 	man5/shadow.5 \
 	man1/su.1 \
 	man8/useradd.8 \
diff -Naur shadow-4.9-orig/man/zh_CN/Makefile.in shadow-4.9/man/zh_CN/Makefile.in
--- shadow-4.9-orig/man/zh_CN/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/zh_CN/Makefile.in	2021-09-28 12:20:58.358982917 +0200
@@ -324,11 +324,11 @@
 top_srcdir = @top_srcdir@
 man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
 	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
-	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
-	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
+ 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
+ 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
 	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
 	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
-	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
+ 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
 	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
 	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
 	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -Naur shadow-4.9-orig/man/zh_TW/Makefile.in shadow-4.9/man/zh_TW/Makefile.in
--- shadow-4.9-orig/man/zh_TW/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/man/zh_TW/Makefile.in	2021-09-28 12:21:26.770569070 +0200
@@ -328,7 +328,7 @@
 	man8/groupadd.8 \
 	man8/groupdel.8 \
 	man8/groupmod.8 \
-	man5/passwd.5 \
+	man5/ \
 	man1/su.1 \
 	man8/useradd.8 \
 	man8/userdel.8 \
diff -Naur shadow-4.9-orig/src/Makefile.in shadow-4.9/src/Makefile.in
--- shadow-4.9-orig/src/Makefile.in	2021-07-22 23:56:58.000000000 +0200
+++ shadow-4.9/src/Makefile.in	2021-09-28 12:25:18.743217348 +0200
@@ -88,7 +88,7 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-bin_PROGRAMS = groups$(EXEEXT) login$(EXEEXT) $(am__EXEEXT_1)
+bin_PROGRAMS = login$(EXEEXT) $(am__EXEEXT_1)
 sbin_PROGRAMS = nologin$(EXEEXT)
 ubin_PROGRAMS = faillog$(EXEEXT) lastlog$(EXEEXT) chage$(EXEEXT) \
 	chfn$(EXEEXT) chsh$(EXEEXT) expiry$(EXEEXT) gpasswd$(EXEEXT) \
