diff -crB shadow-4.2.1-a/etc/login.defs shadow-4.2.1-b/etc/login.defs
*** shadow-4.2.1-a/etc/login.defs	2014-05-09 10:20:28.000000000 +0000
--- shadow-4.2.1-b/etc/login.defs	2016-03-13 10:51:09.680171239 +0000
***************
*** 118,124 ****
  #   Directory where mailboxes reside, _or_ name of file, relative to the
  #   home directory.  If you _do_ define both, MAIL_DIR takes precedence.
  #
! MAIL_DIR	/var/spool/mail
  #MAIL_FILE	.mail
  
  #
--- 118,124 ----
  #   Directory where mailboxes reside, _or_ name of file, relative to the
  #   home directory.  If you _do_ define both, MAIL_DIR takes precedence.
  #
! MAIL_DIR	/var/mail
  #MAIL_FILE	.mail
  
  #
***************
*** 317,323 ****
  # Note: If you use PAM, it is recommended to use a value consistent with
  # the PAM modules configuration.
  #
! #ENCRYPT_METHOD DES
  
  #
  # Only works if ENCRYPT_METHOD is set to SHA256 or SHA512.
--- 317,323 ----
  # Note: If you use PAM, it is recommended to use a value consistent with
  # the PAM modules configuration.
  #
! ENCRYPT_METHOD SHA512
  
  #
  # Only works if ENCRYPT_METHOD is set to SHA256 or SHA512.
Only in shadow-4.2.1-b: SHA512_password_hashing.patch
