diff -crB shadow-4.2.1-a/man/cs/Makefile.in shadow-4.2.1-b/man/cs/Makefile.in
*** shadow-4.2.1-a/man/cs/Makefile.in	2014-05-09 16:49:46.000000000 +0000
--- shadow-4.2.1-b/man/cs/Makefile.in	2016-03-13 10:47:55.924166976 +0000
***************
*** 300,311 ****
  	man8/groupadd.8 \
  	man8/groupdel.8 \
  	man8/groupmod.8 \
! 	man1/groups.1 \
  	man8/grpck.8 \
  	man5/gshadow.5 \
  	man8/lastlog.8 \
  	man8/nologin.8 \
! 	man5/passwd.5 \
  	man5/shadow.5 \
  	man1/su.1 \
  	man8/vipw.8
--- 300,311 ----
  	man8/groupadd.8 \
  	man8/groupdel.8 \
  	man8/groupmod.8 \
! 	man1/ \
  	man8/grpck.8 \
  	man5/gshadow.5 \
  	man8/lastlog.8 \
  	man8/nologin.8 \
! 	man5/ \
  	man5/shadow.5 \
  	man1/su.1 \
  	man8/vipw.8
diff -crB shadow-4.2.1-a/man/da/Makefile.in shadow-4.2.1-b/man/da/Makefile.in
*** shadow-4.2.1-a/man/da/Makefile.in	2014-05-09 16:49:46.000000000 +0000
--- shadow-4.2.1-b/man/da/Makefile.in	2016-03-13 10:47:55.928166977 +0000
***************
*** 298,304 ****
  top_srcdir = @top_srcdir@
  
  # 2012.01.28 - activate manpages with more than 50% translated messages
! man_MANS = man1/chfn.1 man8/groupdel.8 man1/groups.1 man5/gshadow.5 \
  	man8/logoutd.8 man1/newgrp.1 man8/nologin.8 man1/sg.1 \
  	man8/vigr.8 man8/vipw.8 $(am__append_1)
  man_nopam = 
--- 298,304 ----
  top_srcdir = @top_srcdir@
  
  # 2012.01.28 - activate manpages with more than 50% translated messages
! man_MANS = man1/chfn.1 man8/groupdel.8 man1/ man5/gshadow.5 \
  	man8/logoutd.8 man1/newgrp.1 man8/nologin.8 man1/sg.1 \
  	man8/vigr.8 man8/vipw.8 $(am__append_1)
  man_nopam = 
diff -crB shadow-4.2.1-a/man/de/Makefile.in shadow-4.2.1-b/man/de/Makefile.in
*** shadow-4.2.1-a/man/de/Makefile.in	2014-05-09 16:49:46.000000000 +0000
--- shadow-4.2.1-b/man/de/Makefile.in	2016-03-13 10:47:55.916166976 +0000
***************
*** 299,309 ****
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
--- 299,309 ----
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -crB shadow-4.2.1-a/man/es/Makefile.in shadow-4.2.1-b/man/es/Makefile.in
*** shadow-4.2.1-a/man/es/Makefile.in	2014-05-09 16:49:46.000000000 +0000
--- shadow-4.2.1-b/man/es/Makefile.in	2016-03-13 10:47:55.872166975 +0000
***************
*** 266,272 ****
  #	man1/login.1 \
  #	man1/newgrp.1 \
  #	man1/passwd.1 \
! #	man5/passwd.5 \
  #	man1/su.1 \
  #	man8/vigr.8 \
  #	man8/vipw.8
--- 266,272 ----
  #	man1/login.1 \
  #	man1/newgrp.1 \
  #	man1/passwd.1 \
! #	man5/ \
  #	man1/su.1 \
  #	man8/vigr.8 \
  #	man8/vipw.8
diff -crB shadow-4.2.1-a/man/fr/Makefile.in shadow-4.2.1-b/man/fr/Makefile.in
*** shadow-4.2.1-a/man/fr/Makefile.in	2014-05-09 16:49:46.000000000 +0000
--- shadow-4.2.1-b/man/fr/Makefile.in	2016-03-13 10:47:55.984166978 +0000
***************
*** 301,311 ****
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
--- 301,311 ----
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -crB shadow-4.2.1-a/man/hu/Makefile.in shadow-4.2.1-b/man/hu/Makefile.in
*** shadow-4.2.1-a/man/hu/Makefile.in	2014-05-09 16:49:46.000000000 +0000
--- shadow-4.2.1-b/man/hu/Makefile.in	2016-03-13 10:47:55.972166978 +0000
***************
*** 295,306 ****
  man_MANS = \
  	man1/chsh.1 \
  	man1/gpasswd.1 \
! 	man1/groups.1 \
  	man8/lastlog.8 \
  	man1/login.1 \
  	man1/newgrp.1 \
  	man1/passwd.1 \
! 	man5/passwd.5 \
  	man1/sg.1 \
  	man1/su.1
  
--- 295,306 ----
  man_MANS = \
  	man1/chsh.1 \
  	man1/gpasswd.1 \
! 	man1/ \
  	man8/lastlog.8 \
  	man1/login.1 \
  	man1/newgrp.1 \
  	man1/passwd.1 \
! 	man5/ \
  	man1/sg.1 \
  	man1/su.1
  
diff -crB shadow-4.2.1-a/man/it/Makefile.in shadow-4.2.1-b/man/it/Makefile.in
*** shadow-4.2.1-a/man/it/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/it/Makefile.in	2016-03-13 10:47:55.896166976 +0000
***************
*** 299,309 ****
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
--- 299,309 ----
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -crB shadow-4.2.1-a/man/ja/Makefile.in shadow-4.2.1-b/man/ja/Makefile.in
*** shadow-4.2.1-a/man/ja/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/ja/Makefile.in	2016-03-13 10:47:55.932166977 +0000
***************
*** 296,305 ****
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chpasswd.8 man1/chsh.1 \
  	man1/expiry.1 man5/faillog.5 man8/faillog.8 man1/gpasswd.1 \
! 	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/groups.1 \
  	man8/grpck.8 man8/grpconv.8 man8/grpunconv.8 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
  	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man5/shadow.5 \
  	man1/su.1 man5/suauth.5 man8/useradd.8 man8/userdel.8 \
  	man8/usermod.8 man8/vigr.8 man8/vipw.8 $(am__append_1)
--- 296,305 ----
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chpasswd.8 man1/chsh.1 \
  	man1/expiry.1 man5/faillog.5 man8/faillog.8 man1/gpasswd.1 \
! 	man8/groupadd.8 man8/groupdel.8 man8/groupmod.8 man1/ \
  	man8/grpck.8 man8/grpconv.8 man8/grpunconv.8 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man1/passwd.1 man5/ man8/pwck.8 \
  	man8/pwconv.8 man8/pwunconv.8 man1/sg.1 man5/shadow.5 \
  	man1/su.1 man5/suauth.5 man8/useradd.8 man8/userdel.8 \
  	man8/usermod.8 man8/vigr.8 man8/vipw.8 $(am__append_1)
diff -crB shadow-4.2.1-a/man/ko/Makefile.in shadow-4.2.1-b/man/ko/Makefile.in
*** shadow-4.2.1-a/man/ko/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/ko/Makefile.in	2016-03-13 10:47:55.940166977 +0000
***************
*** 295,303 ****
  man_MANS = \
  	man1/chfn.1 \
  	man1/chsh.1 \
! 	man1/groups.1 \
  	man1/login.1 \
! 	man5/passwd.5 \
  	man1/su.1 \
  	man8/vigr.8 \
  	man8/vipw.8
--- 295,303 ----
  man_MANS = \
  	man1/chfn.1 \
  	man1/chsh.1 \
! 	man1/ \
  	man1/login.1 \
! 	man5/ \
  	man1/su.1 \
  	man8/vigr.8 \
  	man8/vipw.8
diff -crB shadow-4.2.1-a/man/Makefile.in shadow-4.2.1-b/man/Makefile.in
*** shadow-4.2.1-a/man/Makefile.in	2014-05-09 16:49:46.000000000 +0000
--- shadow-4.2.1-b/man/Makefile.in	2016-03-13 10:47:55.880166976 +0000
***************
*** 365,375 ****
  @USE_NLS_TRUE@SUBDIRS = po cs da de es fi fr hu id it ja ko pl pt_BR ru sv tr zh_CN zh_TW
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
--- 365,375 ----
  @USE_NLS_TRUE@SUBDIRS = po cs da de es fi fr hu id it ja ko pl pt_BR ru sv tr zh_CN zh_TW
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -crB shadow-4.2.1-a/man/pl/Makefile.in shadow-4.2.1-b/man/pl/Makefile.in
*** shadow-4.2.1-a/man/pl/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/pl/Makefile.in	2016-03-13 10:47:55.912166976 +0000
***************
*** 300,307 ****
  
  # 2012.01.28 - activate manpages with more than 50% translated messages
  man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
! 	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
  	man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 man1/sg.1 \
  	man3/shadow.3 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
  	man8/vipw.8 $(am__append_1)
--- 300,307 ----
  
  # 2012.01.28 - activate manpages with more than 50% translated messages
  man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
! 	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
  	man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 man1/sg.1 \
  	man3/shadow.3 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
  	man8/vipw.8 $(am__append_1)
diff -crB shadow-4.2.1-a/man/pt_BR/Makefile.in shadow-4.2.1-b/man/pt_BR/Makefile.in
*** shadow-4.2.1-a/man/pt_BR/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/pt_BR/Makefile.in	2016-03-13 10:47:55.904166976 +0000
***************
*** 297,303 ****
  	man8/groupadd.8 \
  	man8/groupdel.8 \
  	man8/groupmod.8 \
! 	man5/passwd.5 \
  	man5/shadow.5
  
  EXTRA_DIST = $(man_MANS)
--- 297,303 ----
  	man8/groupadd.8 \
  	man8/groupdel.8 \
  	man8/groupmod.8 \
! 	man5/ \
  	man5/shadow.5
  
  EXTRA_DIST = $(man_MANS)
diff -crB shadow-4.2.1-a/man/ru/Makefile.in shadow-4.2.1-b/man/ru/Makefile.in
*** shadow-4.2.1-a/man/ru/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/ru/Makefile.in	2016-03-13 10:47:55.944166977 +0000
***************
*** 299,309 ****
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
--- 299,309 ----
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -crB shadow-4.2.1-a/man/sv/Makefile.in shadow-4.2.1-b/man/sv/Makefile.in
*** shadow-4.2.1-a/man/sv/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/sv/Makefile.in	2016-03-13 10:47:55.988166978 +0000
***************
*** 300,309 ****
  
  # 2012.01.28 - activate manpages with more than 50% translated messages
  man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
! 	man8/faillog.8 man3/getspnam.3 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
  	man5/gshadow.5 man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 \
! 	man8/nologin.8 man1/passwd.1 man5/passwd.5 man8/pwck.8 \
  	man1/sg.1 man3/shadow.3 man5/suauth.5 man8/userdel.8 \
  	man8/vigr.8 man8/vipw.8 $(am__append_1)
  man_nopam = \
--- 300,309 ----
  
  # 2012.01.28 - activate manpages with more than 50% translated messages
  man_MANS = man1/chage.1 man1/chsh.1 man1/expiry.1 man5/faillog.5 \
! 	man8/faillog.8 man3/ man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
  	man5/gshadow.5 man8/lastlog.8 man8/logoutd.8 man1/newgrp.1 \
! 	man8/nologin.8 man1/passwd.1 man5/ man8/pwck.8 \
  	man1/sg.1 man3/shadow.3 man5/suauth.5 man8/userdel.8 \
  	man8/vigr.8 man8/vipw.8 $(am__append_1)
  man_nopam = \
diff -crB shadow-4.2.1-a/man/tr/Makefile.in shadow-4.2.1-b/man/tr/Makefile.in
*** shadow-4.2.1-a/man/tr/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/tr/Makefile.in	2016-03-13 10:47:55.964166977 +0000
***************
*** 300,306 ****
  	man8/groupmod.8 \
  	man1/login.1 \
  	man1/passwd.1 \
! 	man5/passwd.5 \
  	man5/shadow.5 \
  	man1/su.1 \
  	man8/useradd.8 \
--- 300,306 ----
  	man8/groupmod.8 \
  	man1/login.1 \
  	man1/passwd.1 \
! 	man5/ \
  	man5/shadow.5 \
  	man1/su.1 \
  	man8/useradd.8 \
diff -crB shadow-4.2.1-a/man/zh_CN/Makefile.in shadow-4.2.1-b/man/zh_CN/Makefile.in
*** shadow-4.2.1-a/man/zh_CN/Makefile.in	2014-05-09 16:49:47.000000000 +0000
--- shadow-4.2.1-b/man/zh_CN/Makefile.in	2016-03-13 10:47:55.952166977 +0000
***************
*** 299,309 ****
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/getspnam.3 man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/groups.1 man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/passwd.5 \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
--- 299,309 ----
  top_srcdir = @top_srcdir@
  man_MANS = man1/chage.1 man1/chfn.1 man8/chgpasswd.8 man8/chpasswd.8 \
  	man1/chsh.1 man1/expiry.1 man5/faillog.5 man8/faillog.8 \
! 	man3/ man1/gpasswd.1 man8/groupadd.8 man8/groupdel.8 \
! 	man8/groupmems.8 man8/groupmod.8 man1/ man8/grpck.8 \
  	man8/grpconv.8 man8/grpunconv.8 man5/gshadow.5 man8/lastlog.8 \
  	man1/login.1 man5/login.defs.5 man8/logoutd.8 man1/newgrp.1 \
! 	man8/newusers.8 man8/nologin.8 man1/passwd.1 man5/ \
  	man8/pwck.8 man8/pwconv.8 man8/pwunconv.8 man1/sg.1 \
  	man3/shadow.3 man5/shadow.5 man1/su.1 man5/suauth.5 \
  	man8/useradd.8 man8/userdel.8 man8/usermod.8 man8/vigr.8 \
diff -crB shadow-4.2.1-a/man/zh_TW/Makefile.in shadow-4.2.1-b/man/zh_TW/Makefile.in
*** shadow-4.2.1-a/man/zh_TW/Makefile.in	2014-05-09 16:49:48.000000000 +0000
--- shadow-4.2.1-b/man/zh_TW/Makefile.in	2016-03-13 10:47:55.956166977 +0000
***************
*** 302,308 ****
  	man8/groupadd.8 \
  	man8/groupdel.8 \
  	man8/groupmod.8 \
! 	man5/passwd.5 \
  	man1/su.1 \
  	man8/useradd.8 \
  	man8/userdel.8 \
--- 302,308 ----
  	man8/groupadd.8 \
  	man8/groupdel.8 \
  	man8/groupmod.8 \
! 	man5/ \
  	man1/su.1 \
  	man8/useradd.8 \
  	man8/userdel.8 \
diff -crB shadow-4.2.1-a/src/Makefile.in shadow-4.2.1-b/src/Makefile.in
*** shadow-4.2.1-a/src/Makefile.in	2014-05-09 16:49:48.000000000 +0000
--- shadow-4.2.1-b/src/Makefile.in	2016-03-13 10:47:38.824166600 +0000
***************
*** 78,84 ****
  POST_UNINSTALL = :
  build_triplet = @build@
  host_triplet = @host@
! bin_PROGRAMS = groups$(EXEEXT) login$(EXEEXT) su$(EXEEXT)
  sbin_PROGRAMS = nologin$(EXEEXT)
  ubin_PROGRAMS = faillog$(EXEEXT) lastlog$(EXEEXT) chage$(EXEEXT) \
  	chfn$(EXEEXT) chsh$(EXEEXT) expiry$(EXEEXT) gpasswd$(EXEEXT) \
--- 78,84 ----
  POST_UNINSTALL = :
  build_triplet = @build@
  host_triplet = @host@
! bin_PROGRAMS = login$(EXEEXT) su$(EXEEXT)
  sbin_PROGRAMS = nologin$(EXEEXT)
  ubin_PROGRAMS = faillog$(EXEEXT) lastlog$(EXEEXT) chage$(EXEEXT) \
  	chfn$(EXEEXT) chsh$(EXEEXT) expiry$(EXEEXT) gpasswd$(EXEEXT) \
