--- linux-2.6.27.21/include/linux/skbuff.h	2009-04-03 10:39:10.000000000 +0000
+++ linux-2.6.27.21/include/linux/skbuff.h	2009-04-03 10:41:59.000000000 +0000
@@ -330,11 +330,17 @@
 #if defined(CONFIG_MAC80211) || defined(CONFIG_MAC80211_MODULE)
 	__u8			do_not_encrypt:1;
 #endif
-	/* 0/13/14 bit hole */
+	
 #if defined(CONFIG_IMQ) || defined(CONFIG_IMQ_MODULE)
 	__u8			imq_flags:IMQ_F_BITS;
 #endif
 
+#ifdef CONFIG_XEN
+       __u8                    proto_data_valid:1,
+                               proto_csum_blank:1;
+#endif
+       /* 10-16 bit hole */
+
 #ifdef CONFIG_NET_DMA
 	dma_cookie_t		dma_cookie;
 #endif
