--- a/utility.h	Sun Apr 24 16:59:15 2016
+++ b/utility.h	Sat Aug 20 22:40:33 2016
@@ -97,7 +97,11 @@
 
 // Replacement for exit(status)
 // (exit is not compatible with C++ destructors)
-#define EXIT(status) { throw (int)(status); }
+
+//This doesn't work on IPFire.
+//#define EXIT(status) { throw (int)(status); }
+//tried to use exit and found no problems yet
+#define EXIT(status) { exit ((int)(status)); }
 
 // Compile time check of byte ordering
 // (inline const function allows compiler to remove dead code)

