diff -Naur smartmontools-5.39-svn_r2877/utility.h smartmontools-5.39-svn_r2877.new/utility.h
--- smartmontools-5.39-svn_r2877/utility.h	2009-08-24 12:48:50.000000000 +0200
+++ smartmontools-5.39-svn_r2877.new/utility.h	2009-08-29 09:11:07.000000000 +0200
@@ -102,7 +102,11 @@
 
 // Replacement for exit(status)
 // (exit is not compatible with C++ destructors)
-#define EXIT(status) { throw (int)(status); }
+
+//This doesn't work on IPFire.
+//#define EXIT(status) { throw (int)(status); }
+//tried to use exit and found no problems yet
+#define EXIT(status) { exit ((int)(status)); }
 
 
 #ifdef OLD_INTERFACE
