--- squidclamav-5.11/src/pattern.c~	2014-10-29 13:08:05.658143495 +0000
+++ squidclamav-5.11/src/pattern.c	2014-10-29 13:08:20.964642365 +0000
@@ -151,10 +151,6 @@
      return 1;
   }
   
-  /* extract source ipaddress and source fqdn */
-  if (parseSourceAddress(in_buff.src_address, "/") != 0) {
-  }
-  
   if (debug != 0)
 	logit(log_file, "DEBUG Parsed request: %s %s/%s %s %s\n", in_buff.url, in_buff.ipaddress, in_buff.fqdn, in_buff.ident, in_buff.method);
 
