diff -urN squidGuard-1.2.0_orig/configure.in squidGuard-1.2.0/configure.in
--- squidGuard-1.2.0_orig/configure.in	2001-12-18 10:16:36.000000000 +0000
+++ squidGuard-1.2.0/configure.in	2006-06-10 12:29:26.000000000 +0000
@@ -71,7 +71,12 @@
 
 dnl Checks for libraries.
 
-AC_CHECK_LIB(db,db_version,,[
+AC_CHECK_LIB(db,db_version,,
+AC_CHECK_LIB(db,db_version_4000,,
+AC_CHECK_LIB(db,db_version_4001,,
+AC_CHECK_LIB(db,db_version_4002,,
+AC_CHECK_LIB(db,db_version_4003,,
+AC_CHECK_LIB(db,db_version_4004,,[
 	echo
 	echo "** The Berkley DB library is required for squidGuard"
 	echo "   to compile. Get it from http://www.sleepycat.com"
@@ -79,7 +84,7 @@
 	echo "   its location. (default is $dbprefix/BerkeleyDB)"
 	echo
 	exit 1
-	])
+	]))))))
 
 dnl Checks for header files.
 AC_HEADER_STDC
diff -urN squidGuard-1.2.0_orig/src/sgDb.c squidGuard-1.2.0/src/sgDb.c
--- squidGuard-1.2.0_orig/src/sgDb.c	2001-05-14 13:40:12.000000000 +0000
+++ squidGuard-1.2.0/src/sgDb.c	2006-06-10 12:27:04.000000000 +0000
@@ -98,13 +98,21 @@
     if(createdb)
       flag = flag | DB_TRUNCATE;
     if ((ret = 
+#if DB_VERSION_MINOR == 1 || DB_VERSION_MINOR == 2 || DB_VERSION_MINOR == 3 || DB_VERSION_MINOR == 4
+	 Db->dbp->open(Db->dbp, NULL, dbfile, NULL, DB_BTREE, flag, 0664)) != 0) {
+#else
 	 Db->dbp->open(Db->dbp, dbfile, NULL, DB_BTREE, flag, 0664)) != 0) {
+#endif
       (void) Db->dbp->close(Db->dbp, 0);
       sgLogFatalError("Error db_open: %s", strerror(ret));
     }
   } else {
     if ((ret = 
+#if DB_VERSION_MINOR == 1 || DB_VERSION_MINOR == 2 || DB_VERSION_MINOR == 3 || DB_VERSION_MINOR == 4
+	 Db->dbp->open(Db->dbp, NULL, dbfile, NULL, DB_BTREE, DB_CREATE, 0664)) != 0) {
+#else
 	 Db->dbp->open(Db->dbp, dbfile, NULL, DB_BTREE, DB_CREATE, 0664)) != 0) {
+#endif
       sgLogFatalError("Error db_open: %s", strerror(ret));
     }
   }
