diff -Nur a/src/squidclamav.c b/src/squidclamav.c
--- a/src/squidclamav.c	2012-10-29 09:46:06.000000000 +0100
+++ b/src/squidclamav.c	2013-07-06 19:10:56.375292374 +0200
@@ -413,6 +413,9 @@
 	  /* Suppress error: SSL certificate problem, verify that the CA cert is OK */
 	  curl_easy_setopt (eh, CURLOPT_SSL_VERIFYHOST, 0);
 	  curl_easy_setopt (eh, CURLOPT_SSL_VERIFYPEER, 0);
+
+	  /* Prevent squidclamav from using IPv6 - fix by Nico Prenzel */
+	  curl_easy_setopt (eh, CURLOPT_IPRESOLVE, CURL_IPRESOLVE_V4);
 	}
    }
   /* create a squidguard child process and setup pipes */
