--- strongswan-5.3.1/scripts/Makefile.am.old	2015-06-04 17:20:43.539244145 +0000
+++ strongswan-5.3.1/scripts/Makefile.am	2015-06-04 17:20:51.760510631 +0000
@@ -42,7 +42,7 @@
 dnssec_LDADD = $(top_builddir)/src/libstrongswan/libstrongswan.la
 aes_test_LDADD = $(top_builddir)/src/libstrongswan/libstrongswan.la
 settings_test_LDADD = $(top_builddir)/src/libstrongswan/libstrongswan.la
-timeattack_LDADD = $(top_builddir)/src/libstrongswan/libstrongswan.la
+timeattack_LDADD = $(top_builddir)/src/libstrongswan/libstrongswan.la $(RTLIB)
 
 key2keyid.o :	$(top_builddir)/config.status
 
