diff --git a/src/libcharon/sa/ikev2/task_manager_v2.c b/src/libcharon/sa/ikev2/t
index e9a677a655e4..69118abe78df 100644
--- a/src/libcharon/sa/ikev2/task_manager_v2.c
+++ b/src/libcharon/sa/ikev2/task_manager_v2.c
@@ -1339,7 +1339,7 @@ METHOD(task_manager_t, process_message, status_t,
        {
            DBG1(DBG_IKE, "received message ID %d, expected %d. Ignored",
                 mid, this->responding.mid);
-           if (msg->get_exchange_type(msg) == IKE_SA_INIT)
+           if (mid != 0 && msg->get_exchange_type(msg) == IKE_SA_INIT)
            {   /* clean up IKE_SA state if IKE_SA_INIT has invalid msg ID */
                return DESTROY_ME;
            }
