borrowed from gentoo
--- syslinux/com32/libutil/Makefile
+++ syslinux/com32/libutil/Makefile
@@ -40,7 +40,7 @@
 AR	   = ar
 NASM	   = nasm
 RANLIB	   = ranlib
-CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -D__COM32__ -W -Wall -march=i386 -Os -fomit-frame-pointer -I./include -I../include $(call gcc_ok,-fno-stack-protector,)
+CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -D__COM32__ -W -Wall -march=i386 -Os -fomit-frame-pointer -I./include -I../include $(call gcc_ok,-fno-stack-protector,) $(call gcc_ok,-nopie,)
 SFLAGS     = $(M32) -D__COM32__ -march=i386
 LDFLAGS    = -T ../lib/com32.ld
 LNXCFLAGS  = -I./include -W -Wall -fno-stack-protector -O -g
--- syslinux/dos/Makefile
+++ syslinux/dos/Makefile
@@ -3,7 +3,7 @@
 OBJCOPY  = objcopy
 OPTFLAGS = -g -Os -march=i386 -falign-functions=0 -falign-jumps=0 -falign-loops=0 -fomit-frame-pointer
 INCLUDES = -include code16.h -I. -I.. -I../libfat
-CFLAGS	 = -W -Wall -fno-stack-protector -ffreestanding -msoft-float $(OPTFLAGS) $(INCLUDES)
+CFLAGS	 = -W -Wall -nopie -fno-stack-protector -ffreestanding -msoft-float $(OPTFLAGS) $(INCLUDES)
 LDFLAGS	 = -T com16.ld
 AR       = ar
 RANLIB   = ranlib
