borrowed from gentoo
--- syslinux/Makefile
+++ syslinux/Makefile
@@ -18,7 +18,7 @@
 OSTYPE   = $(shell uname -msr)
 CC	 = gcc
 INCLUDE  =
-CFLAGS   = -W -Wall -Os -fomit-frame-pointer -D_FILE_OFFSET_BITS=64
+CFLAGS   = -W -Wall -fno-stack-protector -Os -fomit-frame-pointer -D_FILE_OFFSET_BITS=64
 PIC      = -fPIC
 LDFLAGS  = -O2 -s
 AR	 = ar
@@ -194,9 +194,8 @@ libsyslinux.a: bootsect_bin.o ldlinux_bi
 $(LIB_SO): bootsect_bin.o ldlinux_bin.o syslxmod.o
 	$(CC) $(LDFLAGS) -shared -Wl,-soname,$(LIB_SONAME) -o $@ $^
 
-gethostip.o: gethostip.c
-
-gethostip: gethostip.o
+gethostip: gethostip.c
+	$(CC) $(LDFLAGS) gethostip.c -o gethostip -fno-stack-protector
 
 mkdiskimage: mkdiskimage.in mbr.bin bin2hex.pl
 	$(PERL) bin2hex.pl < mbr.bin | cat mkdiskimage.in - > $@
--- syslinux/com32/lib/MCONFIG
+++ syslinux/com32/lib/MCONFIG
@@ -18,7 +18,7 @@ LIBFLAGS = -DDYNAMIC_CRC_TABLE -DPNG_NO_
 
 REQFLAGS  = -g -m32 -mregparm=3 -DREGPARM=3 -D__COM32__ -I. -I./sys -I../include
 OPTFLAGS  = -Os -march=i386 -falign-functions=0 -falign-jumps=0 \
-	    -falign-labels=0
+	    -falign-labels=0 -fno-stack-protector
 WARNFLAGS = -W -Wall -Wpointer-arith -Wwrite-strings -Wstrict-prototypes -Winline
 
 CFLAGS  = -Wp,-MT,$@,-MD,$(dir $@).$(notdir $@).d $(OPTFLAGS) \
--- syslinux/com32/libutil/Makefile
+++ syslinux/com32/libutil/Makefile
@@ -40,10 +40,10 @@ LD         = ld -m elf_i386
 AR	   = ar
 NASM	   = nasm
 RANLIB	   = ranlib
-CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -D__COM32__ -W -Wall -march=i386 -Os -fomit-frame-pointer -I./include -I../include
+CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -D__COM32__ -W -Wall -march=i386 -Os -fomit-frame-pointer -I./include -I../include $(call gcc_ok,-fno-stack-protector,)
 SFLAGS     = $(M32) -D__COM32__ -march=i386
 LDFLAGS    = -T ../lib/com32.ld
-LNXCFLAGS  = -I./include -W -Wall -O -g
+LNXCFLAGS  = -I./include -W -Wall -O -g $(call gcc_ok,-fno-stack-protector,)
 LNXSFLAGS  = -g
 LNXLDFLAGS = -g
 OBJCOPY    = objcopy
--- syslinux/com32/modules/Makefile
+++ syslinux/com32/modules/Makefile
@@ -25,8 +25,8 @@ LD         = ld -m elf_i386
 AR	   = ar
 NASM	   = nasm
 RANLIB	   = ranlib
-CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -W -Wall -march=i386 -Os -fomit-frame-pointer -I../libutil/include -I../include  -D__COM32__
-LNXCFLAGS  = -W -Wall -O -g -I../libutil/include
+CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -W -Wall -march=i386 -Os -fomit-frame-pointer -I../libutil/include -I../include  -D__COM32__ $(call gcc_ok,-fno-stack-protector,)
+LNXCFLAGS  = -W -Wall -O -g -I../libutil/include $(call gcc_ok,-fno-stack-protector,)
 LNXSFLAGS  = -g
 LNXLDFLAGS = -g
 SFLAGS     = -D__COM32__ -march=i386
--- syslinux/com32/samples/Makefile
+++ syslinux/com32/samples/Makefile
@@ -25,8 +25,8 @@ LD         = ld -m elf_i386
 AR	   = ar
 NASM	   = nasm
 RANLIB	   = ranlib
-CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -W -Wall -march=i386 -Os -fomit-frame-pointer -I../libutil/include -I../include  -D__COM32__
-LNXCFLAGS  = -W -Wall -O -g -I../libutil/include
+CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -W -Wall -march=i386 -Os -fomit-frame-pointer -I../libutil/include -I../include  -D__COM32__ $(call gcc_ok,-fno-stack-protector,)
+LNXCFLAGS  = -W -Wall -O -g -I../libutil/include $(call gcc_ok,-fno-stack-protector,)
 LNXSFLAGS  = -g
 LNXLDFLAGS = -g
 SFLAGS     = -D__COM32__ -march=i386
--- syslinux/dos/Makefile
+++ syslinux/dos/Makefile
@@ -3,7 +3,7 @@ LD       = ld -m elf_i386
 OBJCOPY  = objcopy
 OPTFLAGS = -g -Os -march=i386 -falign-functions=0 -falign-jumps=0 -falign-loops=0 -fomit-frame-pointer
 INCLUDES = -include code16.h -I. -I.. -I../libfat
-CFLAGS	 = -W -Wall -ffreestanding -msoft-float $(OPTFLAGS) $(INCLUDES)
+CFLAGS	 = -W -Wall -fno-stack-protector -ffreestanding -msoft-float $(OPTFLAGS) $(INCLUDES)
 LDFLAGS	 = -T com16.ld
 AR       = ar
 RANLIB   = ranlib
--- syslinux/extlinux/Makefile
+++ syslinux/extlinux/Makefile
@@ -1,7 +1,7 @@
 CC	 = gcc
 OPTFLAGS = -g -Os
 INCLUDES = -I. -I.. -I../libfat
-CFLAGS	 = -W -Wall -Wno-sign-compare -D_FILE_OFFSET_BITS=64 $(OPTFLAGS) $(INCLUDES)
+CFLAGS	 = -W -Wall -fno-stack-protector -Wno-sign-compare -D_FILE_OFFSET_BITS=64 $(OPTFLAGS) $(INCLUDES)
 LDFLAGS	 = -s
 
 SRCS     = extlinux.c ../extlinux_bss_bin.c ../extlinux_sys_bin.c
--- syslinux/memdisk/Makefile
+++ syslinux/memdisk/Makefile
@@ -20,7 +20,7 @@ M32     := $(call gcc_ok,-m32,)
 FREE    := $(call gcc_ok,-ffreestanding,)
 
 CC       = gcc $(M32) $(FREE)
-CFLAGS   = -g -W -Wall -Wno-sign-compare \
+CFLAGS   = -g -W -Wall -Wno-sign-compare $(call gcc_ok,-fno-stack-protector,) \
 	   -Os -fomit-frame-pointer -march=i386 $(ALIGN) \
 	   -DVERSION='"$(VERSION)"' -DDATE='"$(DATE)"'
 LDFLAGS  = -g
--- syslinux/menu/Makefile
+++ syslinux/menu/Makefile
@@ -5,7 +5,7 @@ M32     := $(call gcc_ok,-m32,)
 COM32DIR   = ../com32
 LUDIR      = $(COM32DIR)/libutil
 LDIR       = $(COM32DIR)/lib
-CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -W -Wall -march=i386 -Os -fomit-frame-pointer -I$(LUDIR)/include -I$(COM32DIR)/include  -Ilibmenu -D__COM32__
+CFLAGS     = $(M32) -mregparm=3 -DREGPARM=3 -W -Wall -march=i386 -Os -fomit-frame-pointer -I$(LUDIR)/include -I$(COM32DIR)/include  -Ilibmenu -D__COM32__ $(call gcc_ok,-fno-stack-protector,)
 SFLAGS     = -D__COM32__ -march=i386
 LDFLAGS    = -T $(LDIR)/com32.ld
 OBJCOPY    = objcopy
--- syslinux/mtools/Makefile
+++ syslinux/mtools/Makefile
@@ -1,7 +1,7 @@
 CC	 = gcc
 OPTFLAGS = -g -Os
 INCLUDES = -I. -I.. -I../libfat
-CFLAGS	 = -W -Wall -D_FILE_OFFSET_BITS=64 $(OPTFLAGS) $(INCLUDES)
+CFLAGS	 = -W -Wall -fno-stack-protector -D_FILE_OFFSET_BITS=64 $(OPTFLAGS) $(INCLUDES)
 LDFLAGS	 = -s
 
 SRCS     = syslinux.c ../syslxmod.c ../bootsect_bin.c ../ldlinux_bin.c $(wildcard ../libfat/*.c)
--- syslinux/sample/Makefile
+++ syslinux/sample/Makefile
@@ -25,7 +25,7 @@ LD         = ld -m elf_i386
 AR	   = ar
 NASM	   = nasm
 RANLIB	   = ranlib
-CFLAGS     = -W -Wall -march=i386 -Os -fomit-frame-pointer -I../com32/include
+CFLAGS     = -W -Wall -march=i386 -Os -fomit-frame-pointer -I../com32/include $(call gcc_ok,-fno-stack-protector,)
 SFLAGS     = -march=i386
 LDFLAGS    = -s
 OBJCOPY    = objcopy
--- syslinux/unix/Makefile
+++ syslinux/unix/Makefile
@@ -1,7 +1,7 @@
 CC	 = gcc
 OPTFLAGS = -g -Os
 INCLUDES = -I. -I.. -I../libfat
-CFLAGS	 = -W -Wall -D_FILE_OFFSET_BITS=64 $(OPTFLAGS) $(INCLUDES)
+CFLAGS	 = -W -Wall -fno-stack-protector -D_FILE_OFFSET_BITS=64 $(OPTFLAGS) $(INCLUDES)
 LDFLAGS	 = -s
 
 SRCS     = syslinux.c ../syslxmod.c ../bootsect_bin.c ../ldlinux_bin.c $(wildcard ../libfat/*.c)
--- syslinux/win32/Makefile
+++ syslinux/win32/Makefile
@@ -23,7 +23,7 @@ ifeq ($(findstring CYGWIN,$(OSTYPE)),CYG
 CC       = gcc
 AR	 = ar
 RANLIB   = ranlib
-CFLAGS   = -mno-cygwin -W -Wall -Os -fomit-frame-pointer -D_FILE_OFFSET_BITS=64
+CFLAGS   = -mno-cygwin -W -Wall -fno-stack-protector -Os -fomit-frame-pointer -D_FILE_OFFSET_BITS=64
 PIC	 =
 LDFLAGS  = -mno-cygwin -Os -s
 else
@@ -37,7 +37,7 @@ AR	 = mingw-ar
 RANLIB   = mingw-ranlib
 endif
 
-CFLAGS   = -W -Wall -Wno-sign-compare -Os -fomit-frame-pointer -D_FILE_OFFSET_BITS=64
+CFLAGS   = -W -Wall -fno-stack-protector -Wno-sign-compare -Os -fomit-frame-pointer -D_FILE_OFFSET_BITS=64
 PIC      =
 LDFLAGS  = -Os -s
 endif
