borrowed from fedora
fix for CAN-2005-1279

--- tcpdump-3.8.2/print-bgp.c.old	2004-03-24 01:01:00.000000000 +0100
+++ tcpdump-3.8.2/print-bgp.c		2005-04-28 12:47:12.000000000 +0200
@@ -1089,7 +1089,7 @@
 			printf(", no SNPA");
                 }
 
-		while (len - (tptr - pptr) > 0) {
+		while ((tptr - pptr) > 0 &&  len - (tptr - pptr) > 0) {
 			switch (af) {
 			case AFNUM_INET:
                             switch (safi) {
