--- Makefile.bak	2006-10-24 00:56:23.000000000 +0000
+++ Makefile	2006-10-23 19:57:49.000000000 +0000
@@ -32,7 +32,7 @@
 	./gen_wc8bit `cat codesets.txt`
 
 gen_wctype: gen_wctype.c
-	$(HOSTCC) $(HOSTCFLAGS) $(NATIVE_LDFLAGS) -D__UCLIBC_GEN_LOCALE $^ -o $@
+	$(HOSTCC) $(HOSTCFLAGS) $(NATIVE_LDFLAGS) -O1 -D__UCLIBC_GEN_LOCALE $^ -o $@
 
 # Warning! Beware tr_TR toupper/tolower exceptions!
 wctables.h: gen_wctype
