diff -Naur udev-config-6.2.org/60-persistent-storage.rules udev-config-6.2/60-persistent-storage.rules
--- udev-config-6.2.org/60-persistent-storage.rules	2006-05-13 01:03:13.000000000 +0200
+++ udev-config-6.2/60-persistent-storage.rules	2010-08-14 11:08:35.000000000 +0200
@@ -9,7 +9,7 @@
 
 # never access removable ide devices, the drivers are causing event loops on open()
 KERNEL=="hd*[!0-9]", SYSFS{removable}=="1", DRIVER=="ide-cs|ide-floppy", GOTO="persistent_storage_end"
-KERNEL=="hd*[0-9]", SYSFS{../removable}=="1", GOTO="persistent_storage_end"
+KERNEL=="hd*[0-9]", SYSFS{removable}=="1", GOTO="persistent_storage_end"
 
 # for partitions import parent information
 KERNEL=="*[0-9]", IMPORT{parent}=="ID_*"
