diff -Naur systemd-208.org/src/libudev/libudev-util.c systemd-208/src/libudev/libudev-util.c
--- systemd-208.org/src/libudev/libudev-util.c	2013-09-23 21:31:55.000000000 +0200
+++ systemd-208/src/libudev/libudev-util.c	2013-11-19 23:18:10.326350131 +0100
@@ -481,7 +481,7 @@
         if (fd < 0)
                 return -errno;
 
-        len = snprintf(text, sizeof(text), "<30>systemd-udevd[%u]: ", getpid());
+        len = snprintf(text, sizeof(text), "udevd[%u]: ", getpid());
 
         va_start(ap, fmt);
         len += vsnprintf(text + len, sizeof(text) - len, fmt, ap);
