diff -Naur v4l-dvb-abd3aac6644e.org/v4l/compat.h v4l-dvb-abd3aac6644e/v4l/compat.h
--- v4l-dvb-abd3aac6644e.org/v4l/compat.h	2010-07-02 05:38:54.000000000 +0200
+++ v4l-dvb-abd3aac6644e/v4l/compat.h	2010-12-04 12:19:09.531056981 +0100
@@ -14,8 +14,6 @@
 #define INIT_DELAYED_WORK(a,b,c)	INIT_WORK(a,b,c)
 #endif
 
-#define EXPERIMENTAL_TREE
-
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 35)
 #define usb_buffer_alloc(dev, size, mem_flags, dma) usb_alloc_coherent(dev, size, mem_flags, dma)
 #define usb_buffer_free(dev, size, addr, dma) usb_free_coherent(dev, size, addr, dma)
