diff -Naur v4l-dvb-6292/linux/drivers/media/video/videobuf-dvb.c v4l-dvb/linux/drivers/media/video/videobuf-dvb.c
--- v4l-dvb-6292/linux/drivers/media/video/videobuf-dvb.c	2008-05-07 23:01:28.000000000 +0200
+++ v4l-dvb/linux/drivers/media/video/videobuf-dvb.c	2008-05-08 18:45:42.000000000 +0200
@@ -28,7 +28,7 @@
 #endif
 
 #include <media/videobuf-dma-sg.h>
-#include <media/video-buf-dvb.h>
+#include <media/videobuf-dvb.h>
 #include "compat.h"
 
 /* ------------------------------------------------------------------ */
