diff -up epgsearch-1.0.1.beta2/epgsearch.c~ epgsearch-1.0.1.beta2/epgsearch.c
--- epgsearch-1.0.1.beta2/epgsearch.c~	2012-06-11 20:01:54.000000000 +0300
+++ epgsearch-1.0.1.beta2/epgsearch.c	2012-06-12 22:20:38.740086339 +0300
@@ -446,16 +446,16 @@ bool cPluginEpgsearch::Start(void)
    LoadConfD();
    LoadUserVars();
 
-   ChannelGroups.Load(AddDirectory(CONFIGDIR, "epgsearchchangrps.conf"), true);
-   Blacklists.Load(AddDirectory(CONFIGDIR, "epgsearchblacklists.conf"));
-   SearchExts.Load(AddDirectory(CONFIGDIR, "epgsearch.conf"));
-   SearchTemplates.Load(AddDirectory(CONFIGDIR, "epgsearchtemplates.conf"));
-   RecsDone.Load(AddDirectory(CONFIGDIR, "epgsearchdone.data"));
-   SwitchTimers.Load(AddDirectory(CONFIGDIR, "epgsearchswitchtimers.conf"));
-   NoAnnounces.Load(AddDirectory(CONFIGDIR, "noannounce.conf"));
-   DefTimerCheckModes.Load(AddDirectory(CONFIGDIR, "deftimerchkmodes.conf"));
-   TimersDone.Load(AddDirectory(CONFIGDIR, "timersdone.conf"));
-   PendingNotifications.Load(AddDirectory(CONFIGDIR, "pendingnotifications.conf"));
+   ChannelGroups.Load("/var/lib/vdr/data/epgsearch/epgsearchchangrps.conf", true);
+   Blacklists.Load("/var/lib/vdr/data/epgsearch/epgsearchblacklists.conf");
+   SearchExts.Load("/var/lib/vdr/data/epgsearch/epgsearch.conf");
+   SearchTemplates.Load("/var/lib/vdr/data/epgsearch/epgsearchtemplates.conf");
+   RecsDone.Load("/var/lib/vdr/data/epgsearch/epgsearchdone.data");
+   SwitchTimers.Load("/var/lib/vdr/data/epgsearch/epgsearchswitchtimers.conf");
+   NoAnnounces.Load("/var/lib/vdr/data/epgsearch/noannounce.conf");
+   DefTimerCheckModes.Load("/var/lib/vdr/data/epgsearch/deftimerchkmodes.conf");
+   TimersDone.Load("/var/lib/vdr/data/epgsearch/timersdone.conf");
+   PendingNotifications.Load("/var/lib/vdr/data/epgsearch/pendingnotifications.conf");
 
    cSearchTimerThread::Init(this);
    cSwitchTimerThread::Init();
diff -up epgsearch-1.0.1.beta2/epgsearchcfg.c~ epgsearch-1.0.1.beta2/epgsearchcfg.c
--- epgsearch-1.0.1.beta2/epgsearchcfg.c~	2012-06-11 19:43:52.000000000 +0300
+++ epgsearch-1.0.1.beta2/epgsearchcfg.c	2012-06-12 22:20:39.169055465 +0300
@@ -80,7 +80,7 @@ cEPGSearchConfig::cEPGSearchConfig(void)
   strcpy(MailAuthUser, "");
   strcpy(MailAuthPass, "");
   strcpy(LastMailConflicts, "");
-  mailViaScript = 1;
+  mailViaScript = 0;
   manualTimerCheckDefault = 0;
   noAnnounceWhileReplay = 0;
   TimerProgRepeat = 0;
diff -up epgsearch-1.0.1.beta2/scripts/recordingdone.sh~ epgsearch-1.0.1.beta2/scripts/recordingdone.sh
--- epgsearch-1.0.1.beta2/scripts/recordingdone.sh~	2012-06-12 22:19:28.570878733 +0300
+++ epgsearch-1.0.1.beta2/scripts/recordingdone.sh	2012-06-12 22:20:39.536029167 +0300
@@ -60,10 +60,10 @@ CTIME_FROM_RECORDING=yes
 SVDRPSEND=svdrpsend
 
 # Home of EPGsearch
-EPGSEARCH_HOME="/etc/vdr/plugins"
+EPGSEARCH_HOME="/var/lib/vdr/data/epgsearch"
 
 # Video root
-VIDEO_ROOT="/video"
+VIDEO_ROOT="/var/video"
 
 # do not edit below this line
 #------------------------------------------------------------------------------
diff -up epgsearch-1.0.1.beta2/scripts/recrep.sh~ epgsearch-1.0.1.beta2/scripts/recrep.sh
--- epgsearch-1.0.1.beta2/scripts/recrep.sh~	2012-06-11 19:43:51.000000000 +0300
+++ epgsearch-1.0.1.beta2/scripts/recrep.sh	2012-06-12 22:20:39.914002061 +0300
@@ -15,7 +15,7 @@
 # adjust the following lines to your config
 
 # your plugins config dir
-PLUGINCONFDIR=/var/lib/vdr/plugins/epgsearch
+PLUGINCONFDIR=/etc/vdr/plugins/epgsearch
 # path to svdrpsend
 SVDRPSEND=svdrpsend
 # the key used to call epgsearch
diff -up epgsearch-1.0.1.beta2/scripts/timercmds-auxinfo.sh~ epgsearch-1.0.1.beta2/scripts/timercmds-auxinfo.sh
--- epgsearch-1.0.1.beta2/scripts/timercmds-auxinfo.sh~	2012-06-11 19:43:51.000000000 +0300
+++ epgsearch-1.0.1.beta2/scripts/timercmds-auxinfo.sh	2012-06-12 22:20:40.723944345 +0300
@@ -15,7 +15,7 @@
 #<Configuration>
 
 # Your timers.conf
-TIMERS="/video/timers.conf"
+TIMERS="/etc/vdr/timers.conf"
 
 #</Configuration>
 
