diff -Naur vdr-streamdev-10db11ac.org/server/menuHTTP.c vdr-streamdev-10db11ac/server/menuHTTP.c
--- vdr-streamdev-10db11ac.org/server/menuHTTP.c	2013-02-18 12:43:02.000000000 +0100
+++ vdr-streamdev-10db11ac/server/menuHTTP.c	2013-04-29 13:34:11.818857417 +0200
@@ -28,7 +28,7 @@
 {
 	struct stat st;
 	if (stat(current->FileName(), &st) == 0)
-		return cString::sprintf("%lu:%lu.rec", st.st_dev, st.st_ino);
+		return cString::sprintf("%lu:%lu.rec", (unsigned long) st.st_dev, (unsigned long) st.st_ino);
 	return "";
 }
 
