diff --git a/src/hs_valid_platform.c b/src/hs_valid_platform.c
index 0af36b6c..12ae5d9a 100644
--- a/src/hs_valid_platform.c
+++ b/src/hs_valid_platform.c
@@ -37,9 +37,9 @@
 
 HS_PUBLIC_API
 hs_error_t HS_CDECL hs_valid_platform(void) {
-    /* Hyperscan requires SSSE3, anything else is a bonus */
+    /* Vectorscan requires SSE4.2, anything else is a bonus */
 #if defined(ARCH_IA32) || defined(ARCH_X86_64)
-    if (check_ssse3()) {
+    if (check_sse42()) {
         return HS_SUCCESS;
     } else {
         return HS_ARCH_ERROR;
