diff -Naur w_scan_start.sh w_scan_start.sh
--- w_scan_start.sh	2008-01-05 19:58:44.000000000 +0100
+++ w_scan_start.sh	2008-05-14 18:29:21.000000000 +0200
@@ -6,11 +6,11 @@
 # - 20060812: first version
 # - 20060822: many improvements
 # cu@vdr-portal
-#
+# - 20080514: patches for ipfire by arne_f@ipfire.org
 #######################################################
 
 umask 022
-export PATH="$(pwd):/bin:/usr/bin:/sbin:/usr/sbin"
+export PATH="$(pwd):/bin:/usr/bin:/sbin:/usr/sbin:/opt/vdr/bin"
 DATE="$(date +%Y)$(date +%m)$(date +%d)" # $DATE == "YYYYMMDD", i.e. 20060710
 W_SCAN=$(which w_scan 2>/dev/null)
 UDEVSTART=$(which udevstart)
@@ -224,7 +224,7 @@
 		echo "Checking for ct-vdr style startscript.."
 		if [ -e /etc/init.d/vdr ]; then			# c't-vdr
 			echo_yes
-			$(/etc/init.d/vdr stop)
+			/etc/init.d/vdr stop
 			stopped=true
 		else
 			echo_no
@@ -375,7 +375,7 @@
 
 	# which generic dvb modules should be always loaded?
 	MODULES_TO_LOAD="videodev v4l1_compat \
-			v4l2_common video_buf dvb_core ves1820 \
+			v4l2_common dvb_core ves1820 \
 			dvb_ttpci \
 			budget_ci budget_av budget_ci budget "
 
@@ -661,9 +661,11 @@
 	echo -e -n $NORMAL
 	echo -e -n "\\033[1;34m"
 	if [ ${WHAT_TO_SCAN:0:2} != 0 ]; then
+		echo ":->W_SCAN $DATE DVB-T" >> channels.conf-$DATE
 		$W_SCAN >> channels.conf-$DATE
 	fi
-	if [ ${WHAT_TO_SCAN:3:2} != 0 ]; then 
+	if [ ${WHAT_TO_SCAN:3:2} != 0 ]; then
+		echo ":->W_SCAN $DATE DVB-C" >> channels.conf-$DATE 
 		$W_SCAN -fc >> channels.conf-$DATE
 	fi 
 	echo -e -n $NORMAL
@@ -671,55 +673,34 @@
 
 function CopyConf ()
 {
-	if [ -s channels.conf-$DATE ]; then
-		clear
-		echo -e -n $SUCCESS
-		echo "****************************************************************"
-		echo "*                                                              *"
-		echo "* w_scan is done.                                              *"
-		echo -e "* The output file is called "$FAILURE"channels.conf-$DATE"$SUCCESS".            *"
-		echo "* Should I copy this to default location for you? [y/n]        *"
-		echo "*                                                              *"
-		echo "****************************************************************"
-		echo -e -n $NORMAL && tput sgr0
-		read RESPONSE
-		echo -e -n $NORMAL
-		if [ "$RESPONSE" != "y" ]; then
-			echo "Copying channels.conf skipped - okay."
-			echo "The channels.conf generated can be found at"
-			echo -n -e "\\033[1;34m"
-			echo "$(pwd)/channels.conf-$DATE"
-			echo ""
-			echo -e -n $NORMAL
-		else
-			if [ -d /etc/vdr ]; then
-				mv /etc/vdr/channels.conf /etc/vdr/channels.conf-backup$DATE
-				cp channels.conf-$DATE /etc/vdr/channels.conf
-			fi
-			if [ -d /var/lib/vdr ]; then
-				mv /var/lib/vdr/channels.conf /var/lib/vdr/channels.conf-backup$DATE
-				cp channels.conf-$DATE /var/lib/vdr/channels.conf
-			fi			
-		fi
-        else
-		clear
-		echo -e -n $FAILURE
-		echo "****************************************************************"
-		echo "*                                                              *"
-		echo "* Something went wrong. The channels.conf generated is *EMPTY*!*"
-		echo "* To analyze your problem I recommend the command              *"
-		echo "*    './w_scan 2>&1 | tee w_scan.log-dvbt &&                   *"
-		echo "*     ./w_scan -fc 2>&1 | tee w_scan.log-dvbc'                 *"
-		echo "*                                                              *"
-		echo "* w_scan will report the problem into the logfiles             *"
-		echo "* w_scan.log-dvbt and w_scan.log-dvbc.                         *"
-		echo "*                                                              *"
-		echo "* PRESS ANY KEY TO CONTINUE                                    *"
-		echo "****************************************************************"
+	echo -e -n $SUCCESS
+	echo "****************************************************************"
+	echo "*                                                              *"
+	echo "* w_scan is done.                                              *"
+	echo -e "* The output file is called "$FAILURE"channels.conf-$DATE"$SUCCESS".            *"
+	echo "* Should I copy this to default location for you? [y/n]        *"
+	echo "*                                                              *"
+	echo "****************************************************************"
+	echo -e -n $NORMAL && tput sgr0
+	read RESPONSE
+	echo -e -n $NORMAL
+	if [ "$RESPONSE" != "y" ]; then
+		echo "Copying channels.conf skipped - okay."
+		echo "The channels.conf generated can be found at"
+		echo -n -e "\\033[1;34m"
+		echo "$(pwd)/channels.conf-$DATE"
+		echo ""
 		echo -e -n $NORMAL
-		read
+	else
+		if [ -d /opt/vdr/etc ]; then
+			mv /opt/vdr/etc/channels.conf /opt/vdr/etc/channels.conf-backup$DATE
+			cp channels.conf-$DATE /opt/vdr/etc/channels.conf
+		fi
+		if [ -d /var/lib/vdr ]; then
+			mv /var/lib/vdr/channels.conf /var/lib/vdr/channels.conf-backup$DATE
+			cp channels.conf-$DATE /var/lib/vdr/channels.conf
+		fi			
 	fi
-
 }
 
 # the program itself.
