--- a/xen/arch/x86/platform_hypercall.c	2009-07-23 07:26:18.000000000 +0000
+++ b/xen/arch/x86/platform_hypercall.c	2009-07-23 07:28:30.000000000 +0000
@@ -313,7 +313,6 @@
     {
         uint32_t cpu;
         uint64_t idletime, now = NOW();
-        struct vcpu *v;
         struct xenctl_cpumap ctlmap;
         cpumask_t cpumap;
         XEN_GUEST_HANDLE(uint8) cpumap_bitmap;
@@ -336,7 +335,7 @@
 
         for_each_cpu_mask ( cpu, cpumap )
         {
-            if ( (v = idle_vcpu[cpu]) != NULL )
+            if ( !idle_vcpu[cpu] )
                 cpu_clear(cpu, cpumap);
             idletime = get_cpu_idle_time(cpu);
 
