--- ipt_layer7.c~	2007-05-25 13:57:38.000000000 +0200
+++ ipt_layer7.c	2007-05-25 13:58:43.000000000 +0200
@@ -368,13 +368,13 @@
 		return (pattern_result ^ info->invert);
 	}
 
-	if(skb_is_nonlinear(skb)){
-		if(skb_linearize(skb, GFP_ATOMIC) != 0){
+//	if(skb_is_nonlinear(skb)){
+		if(skb_linearize(skb/*, GFP_ATOMIC */) != 0){
 			if (net_ratelimit()) 
 				printk(KERN_ERR "layer7: failed to linearize packet, bailing.\n");
 			return info->invert;
 		}
-	}
+//	}
 	
 	/* now that the skb is linearized, it's safe to set these. */
 	app_data = skb->data + app_data_offset(skb);
