--- a/tools/examples/xend-config.sxp	2009-05-18 11:05:38.000000000 +0000
+++ b/tools/examples/xend-config.sxp	2009-07-23 08:00:23.000000000 +0000
@@ -14,10 +14,6 @@
 #(logfile /var/log/xen/xend.log)
 #(loglevel DEBUG)
 
-# Uncomment the line below.  Set the value to flask, acm, or dummy to 
-# select a security module.
-
-#(xsm_module_name dummy)
 
 # The Xen-API server configuration.
 #
@@ -58,21 +54,20 @@
 
 
 #(xend-http-server no)
-#(xend-unix-server no)
-#(xend-tcp-xmlrpc-server no)
-#(xend-unix-xmlrpc-server yes)
+(xend-unix-server yes)
+(xend-tcp-xmlrpc-server yes)
+(xend-unix-xmlrpc-server yes)
 #(xend-relocation-server no)
 (xend-relocation-server yes)
 #(xend-relocation-ssl-server no)
-#(xend-udev-event-server no)
 
 #(xend-unix-path /var/lib/xend/xend-socket)
 
 
 # Address and port xend should use for the legacy TCP XMLRPC interface, 
 # if xend-tcp-xmlrpc-server is set.
-#(xend-tcp-xmlrpc-server-address 'localhost')
-#(xend-tcp-xmlrpc-server-port 8006)
+(xend-tcp-xmlrpc-server-address 'localhost')
+(xend-tcp-xmlrpc-server-port 8006)
 
 # SSL key and certificate to use for the legacy TCP XMLRPC interface.
 # Setting these will mean that this port serves only SSL connections as
@@ -82,11 +77,11 @@
 
 
 # Port xend should use for the HTTP interface, if xend-http-server is set.
-#(xend-port            8000)
+(xend-port            8000)
 
 # Port xend should use for the relocation interface, if xend-relocation-server
 # is set.
-#(xend-relocation-port 8002)
+(xend-relocation-port 8002)
 
 # Port xend should use for the ssl relocation interface, if
 # xend-relocation-ssl-server is set.
@@ -154,7 +149,7 @@
 # two fake interfaces per guest domain.  To do things like this, write
 # yourself a wrapper script, and call network-bridge from it, as appropriate.
 #
-(network-script network-bridge)
+(network-script network-dummy)
 
 # The script used to control virtual interfaces.  This can be overridden on a
 # per-vif basis when creating a domain or a configuring a new vif.  The
@@ -169,7 +164,7 @@
 # If you are using only one bridge, the vif-bridge script will discover that,
 # so there is no need to specify it explicitly.
 #
-(vif-script vif-bridge)
+#(vif-script vif-bridge)
 
 
 ## Use the following if network traffic is routed, as an alternative to the
@@ -194,7 +189,7 @@
 
 # In SMP system, dom0 will use dom0-cpus # of CPUS
 # If dom0-cpus = 0, dom0 will take all cpus available
-(dom0-cpus 0)
+#(dom0-cpus 0)
 
 # Whether to enable core-dumps when domains crash.
 #(enable-dump no)
@@ -205,7 +200,7 @@
 # The interface for VNC servers to listen on. Defaults
 # to 127.0.0.1  To restore old 'listen everywhere' behaviour
 # set this to 0.0.0.0
-#(vnc-listen '127.0.0.1')
+(vnc-listen '0.0.0.0')
 
 # The default password for VNC console on HVM domain.
 # Empty string is no authentication.
@@ -243,7 +238,7 @@
 
 # The default keymap to use for the VM's virtual keyboard
 # when not specififed in VM's configuration
-#(keymap 'en-us')
+(keymap 'de')
 
 # Script to run when the label of a resource has changed.
 #(resource-label-change-script '')
