diff --git a/geoip/xt_geoip_build b/geoip/xt_geoip_build
index 3b15875..7bc42f3 100755
--- a/geoip/xt_geoip_build
+++ b/geoip/xt_geoip_build
@@ -259,7 +259,12 @@ sub writeCountry
 		my ($start, $end) = split('-', $range);
 		$start = inet_pton($family, $start);
 		$end = inet_pton($family, $end);
-		print $fh $start, $end;
+
+		if ($family == AF_INET) {
+			print $fh substr($start, 0, 4), substr($end, 0, 4);
+		} else {
+			print $fh $start, $end;
+		}
 	}
 	close $fh;
 }
