diff -Naur dracut-056.prg/dracut.sh dracut-056/dracut.sh
--- dracut-056.prg/dracut.sh	2022-02-18 12:32:53.000000000 +0100
+++ dracut-056/dracut.sh	2023-12-19 23:30:27.720703043 +0100
@@ -1385,22 +1385,21 @@
 if [[ $early_microcode == yes ]]; then
     if [[ $hostonly ]]; then
         if [[ $(get_cpu_vendor) == "AMD" ]]; then
-            check_kernel_config CONFIG_MICROCODE_AMD || unset early_microcode
+            check_kernel_config CONFIG_MICROCODE || unset early_microcode
         elif [[ $(get_cpu_vendor) == "Intel" ]]; then
-            check_kernel_config CONFIG_MICROCODE_INTEL || unset early_microcode
+            check_kernel_config CONFIG_MICROCODE || unset early_microcode
         else
             unset early_microcode
         fi
     else
-        ! check_kernel_config CONFIG_MICROCODE_AMD \
-            && ! check_kernel_config CONFIG_MICROCODE_INTEL \
+        ! check_kernel_config CONFIG_MICROCODE \
             && unset early_microcode
     fi
     # Do not complain on non-x86 architectures as it makes no sense
     case $(uname -m) in
         x86_64 | i?86)
             [[ $early_microcode != yes ]] \
-                && dwarn "Disabling early microcode, because kernel does not support it. CONFIG_MICROCODE_[AMD|INTEL]!=y"
+                && dwarn "Disabling early microcode, because kernel does not support it. CONFIG_MICROCODE!=y"
             ;;
         *) ;;
     esac
