diff -Naur ccache-4.3.org/src/third_party/xxhash.h ccache-4.3/src/third_party/xxhash.h
--- ccache-4.3.org/src/third_party/xxhash.h	2021-05-09 18:39:46.000000000 +0000
+++ ccache-4.3/src/third_party/xxhash.h	2021-06-26 11:48:18.384900728 +0000
@@ -806,9 +806,7 @@
  * Prefer these methods in priority order (0 > 1 > 2 > 3)
  */
 #ifndef XXH_FORCE_MEMORY_ACCESS   /* can be defined externally, on command line for example */
-#  if !defined(__clang__) && defined(__GNUC__) && defined(__ARM_FEATURE_UNALIGNED) && defined(__ARM_ARCH) && (__ARM_ARCH == 6)
-#    define XXH_FORCE_MEMORY_ACCESS 2
-#  elif !defined(__clang__) && ((defined(__INTEL_COMPILER) && !defined(_WIN32)) || \
+#  if !defined(__clang__) && ((defined(__INTEL_COMPILER) && !defined(_WIN32)) || \
   (defined(__GNUC__) && (defined(__ARM_ARCH) && __ARM_ARCH >= 7)))
 #    define XXH_FORCE_MEMORY_ACCESS 1
 #  endif
