diff -Naur linux-4.9.58.org/mm/filemap.c linux-4.9.58/mm/filemap.c
--- linux-4.9.58.org/mm/filemap.c	2017-10-21 17:21:39.000000000 +0200
+++ linux-4.9.58/mm/filemap.c	2017-10-25 13:53:57.083152487 +0200
@@ -1688,9 +1688,9 @@
 	unsigned int prev_offset;
 	int error = 0;
 
-	if (unlikely(*ppos >= inode->i_sb->s_maxbytes))
-		return 0;
-	iov_iter_truncate(iter, inode->i_sb->s_maxbytes);
+//	if (unlikely(*ppos >= inode->i_sb->s_maxbytes))
+//		return 0;
+//	iov_iter_truncate(iter, inode->i_sb->s_maxbytes);
 
 	index = *ppos >> PAGE_SHIFT;
 	prev_index = ra->prev_pos >> PAGE_SHIFT;
