diff -Naur linux-4.9.8.org/drivers/net/wireless/ralink/rt2x00/rt2800usb.c linux-4.9.8/drivers/net/wireless/ralink/rt2x00/rt2800usb.c
--- linux-4.9.8.org/drivers/net/wireless/ralink/rt2x00/rt2800usb.c	2017-02-04 09:47:29.000000000 +0100
+++ linux-4.9.8/drivers/net/wireless/ralink/rt2x00/rt2800usb.c	2017-02-11 10:50:04.909859370 +0100
@@ -448,7 +448,7 @@
 
 	rt2x00usb_register_read(rt2x00dev, TXRXQ_PCNT, &reg);
 	if (rt2x00_get_field32(reg, TXRXQ_PCNT_TX0Q)) {
-		rt2x00_warn(rt2x00dev, "TX HW queue 0 timed out, invoke forced kick\n");
+		rt2x00_dbg(rt2x00dev, "TX HW queue 0 timed out, invoke forced kick\n");
 
 		rt2x00usb_register_write(rt2x00dev, PBF_CFG, 0xf40012);
 
@@ -463,7 +463,7 @@
 
 	rt2x00usb_register_read(rt2x00dev, TXRXQ_PCNT, &reg);
 	if (rt2x00_get_field32(reg, TXRXQ_PCNT_TX1Q)) {
-		rt2x00_warn(rt2x00dev, "TX HW queue 1 timed out, invoke forced kick\n");
+		rt2x00_dbg(rt2x00dev, "TX HW queue 1 timed out, invoke forced kick\n");
 
 		rt2x00usb_register_write(rt2x00dev, PBF_CFG, 0xf4000a);
 
@@ -613,7 +613,7 @@
 
 		if (unlikely(test_bit(ENTRY_OWNER_DEVICE_DATA, &entry->flags) ||
 			     !test_bit(ENTRY_DATA_STATUS_PENDING, &entry->flags))) {
-			rt2x00_warn(rt2x00dev, "Data pending for entry %u in queue %u\n",
+			rt2x00_dbg(rt2x00dev, "Data pending for entry %u in queue %u\n",
 				    entry->entry_idx, qid);
 			break;
 		}
diff -Naur linux-4.9.8.org/drivers/net/wireless/ralink/rt2x00/rt2x00usb.c linux-4.9.8/drivers/net/wireless/ralink/rt2x00/rt2x00usb.c
--- linux-4.9.8.org/drivers/net/wireless/ralink/rt2x00/rt2x00usb.c	2017-02-04 09:47:29.000000000 +0100
+++ linux-4.9.8/drivers/net/wireless/ralink/rt2x00/rt2x00usb.c	2017-02-11 10:50:04.909859370 +0100
@@ -524,7 +524,7 @@
 
 static void rt2x00usb_watchdog_tx_dma(struct data_queue *queue)
 {
-	rt2x00_warn(queue->rt2x00dev, "TX queue %d DMA timed out, invoke forced forced reset\n",
+	rt2x00_dbg(queue->rt2x00dev, "TX queue %d DMA timed out, invoke forced reset\n",
 		    queue->qid);
 
 	rt2x00queue_stop_queue(queue);
