diff -Naur linux-6.6.32.org/drivers/net/usb/ax88179_178a.c linux-6.6.32/drivers/net/usb/ax88179_178a.c
--- linux-6.6.32.org/drivers/net/usb/ax88179_178a.c	2024-05-25 16:22:56.000000000 +0200
+++ linux-6.6.32/drivers/net/usb/ax88179_178a.c	2024-06-04 10:45:18.113453100 +0200
@@ -1678,10 +1678,10 @@
 {
 	struct ax88179_data *ax179_data = dev->driver_priv;
 
-	if (ax179_data->initialized)
+//	if (ax179_data->initialized)
 		ax88179_reset(dev);
-	else
-		ax179_data->initialized = 1;
+//	else
+//		ax179_data->initialized = 1;
 
 	return 0;
 }
