diff U3 src/ap/wpa_auth.c src/ap/wpa_auth.c
--- a/src/ap/wpa_auth.c	Wed Aug  7 15:25:25 2019
+++ b/src/ap/wpa_auth.c	Fri Sep 20 17:35:23 2019
@@ -68,9 +68,9 @@
 static int ieee80211w_kde_len(struct wpa_state_machine *sm);
 static u8 * ieee80211w_kde_add(struct wpa_state_machine *sm, u8 *pos);
 
-static const u32 eapol_key_timeout_first = 100; /* ms */
-static const u32 eapol_key_timeout_subseq = 1000; /* ms */
-static const u32 eapol_key_timeout_first_group = 500; /* ms */
+static const u32 eapol_key_timeout_first = 300; /* ms */
+static const u32 eapol_key_timeout_subseq = 3000; /* ms */
+static const u32 eapol_key_timeout_first_group = 1500; /* ms */
 static const u32 eapol_key_timeout_no_retrans = 4000; /* ms */
 
 /* TODO: make these configurable */
