From d58f8ef75a29dd6f8968084b5383ce0f39c75666 Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Wed, 12 Aug 2015 10:50:42 +0100
Subject: [PATCH] Filter all IDs that only consist of 0xff

Signed-off-by: Michael Tremer <michael.tremer@ipfire.org>
---
 src/fireinfo/system.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/fireinfo/system.py b/src/fireinfo/system.py
index 4148c66..edf7359 100644
--- a/src/fireinfo/system.py
+++ b/src/fireinfo/system.py
@@ -255,6 +255,10 @@ class System(object):
 						id = None
 						break
 
+			# Check if the string only contains 0xff
+			if all((e == "\xff" for e in id)):
+				id = None
+
 			if id:
 				_ids.append(id)
 
-- 
2.4.3

