From deafec982e4c8f2e6ffa3bf70b0a94fa30158e9a Mon Sep 17 00:00:00 2001
From: Michael Tremer <michael.tremer@ipfire.org>
Date: Wed, 9 Sep 2015 15:04:43 +0100
Subject: [PATCH 5/5] Fix crash if there is id has already been reset to None

Signed-off-by: Michael Tremer <michael.tremer@ipfire.org>
---
 src/fireinfo/system.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/fireinfo/system.py b/src/fireinfo/system.py
index edf7359a17e6..c2ba12e818f0 100644
--- a/src/fireinfo/system.py
+++ b/src/fireinfo/system.py
@@ -256,7 +256,7 @@ class System(object):
 						break
 
 			# Check if the string only contains 0xff
-			if all((e == "\xff" for e in id)):
+			if id and all((e == "\xff" for e in id)):
 				id = None
 
 			if id:
-- 
2.4.3

