source: https://github.com/vim/vim/issues/5581

diff -Naur vim91/src/structs.h.orig vim91/src/structs.h

--- vim91/src/structs.h.orig	2023-12-18 08:47:12.000000000 +0100
+++ vim91/src/structs.h	2024-05-21 13:18:22.389336943 +0200
@@ -1727,7 +1727,7 @@
 {
     typval_T	di_tv;		// type and value of the variable
     char_u	di_flags;	// DI_FLAGS_ flags (only used for variable)
-    char_u	di_key[1];	// key (actually longer!)
+    char_u	di_key[];	// key (actually longer!)
 };
 typedef struct dictitem_S dictitem_T;
 
