source: https://github.com/vim/vim/issues/5581

diff -Naur vim-9.1.0886/src/structs.h.orig vim-9.1.0886/src/structs.h

--- vim-9.1.0886/src/structs.h.orig	2024-11-24 14:34:01.000000000 +0100
+++ vim-9.1.0886/src/structs.h	2024-11-27 13:10:17.019201106 +0100
@@ -1764,7 +1764,7 @@
 {
     typval_T	di_tv;		// type and value of the variable
     char_u	di_flags;	// DI_FLAGS_ flags (only used for variable)
-    char_u	di_key[1];	// key (actually longer!)
+    char_u	di_key[];	// key (actually longer!)
 };
 typedef struct dictitem_S dictitem_T;
 
