diff -rup a/iconvdata/gconv-modules b/iconvdata/gconv-modules
--- a/iconvdata/gconv-modules	2010-05-04 05:27:23.000000000 -0600
+++ b/iconvdata/gconv-modules	2012-01-26 10:58:24.181895489 -0700
@@ -1954,3 +1954,6 @@ alias	HPGREEK8//		HP-GREEK8//
 alias	OSF10010004//		HP-GREEK8//
 module	HP-GREEK8//		INTERNAL		HP-GREEK8	1
 module	INTERNAL		HP-GREEK8//		HP-GREEK8	1
+
+alias	ISO-10646-UCS-2//	UNICODE//
+alias	ISO-10646-UCS-2//	ISO-10646/UTF8/
