diff -rup a/resolv/res_init.c b/resolv/res_init.c
--- a/resolv/res_init.c	2010-05-04 05:27:23.000000000 -0600
+++ b/resolv/res_init.c	2012-02-10 10:20:24.923578396 -0700
@@ -325,7 +325,7 @@ __res_vinit(res_state statp, int preinit
                         struct in6_addr a6;
                         char *el;
 
-                        if ((el = strchr(cp, '\n')) != NULL)
+                        if ((el = strpbrk(cp, " \t\n")) != NULL)
                             *el = '\0';
 			if ((el = strchr(cp, SCOPE_DELIMITER)) != NULL)
 			    *el = '\0';
