diff -Nrup a/locale/loadlocale.c b/locale/loadlocale.c
--- a/locale/loadlocale.c	2010-05-04 07:27:23.000000000 -0400
+++ b/locale/loadlocale.c	2012-08-05 17:19:47.761384155 -0400
@@ -170,7 +170,6 @@ _nl_load_locale (struct loaded_l10nfile 
   int save_err;
   int alloc = ld_mapped;
 
-  file->decided = 1;
   file->data = NULL;
 
   fd = open_not_cancel_2 (file->filename, O_RDONLY);
@@ -279,6 +278,7 @@ _nl_load_locale (struct loaded_l10nfile 
   newdata->alloc = alloc;
 
   file->data = newdata;
+  file->decided = 1;
 }
 
 void
