diff -Naur u-boot-2018.03.org/board/raspberrypi/rpi/rpi.c u-boot-2018.03/board/raspberrypi/rpi/rpi.c
--- u-boot-2018.03.org/board/raspberrypi/rpi/rpi.c	2018-03-13 13:02:19.000000000 +0100
+++ u-boot-2018.03/board/raspberrypi/rpi/rpi.c	2018-05-14 13:40:38.418660145 +0200
@@ -111,6 +111,11 @@
 		DTB_DIR "bcm2835-rpi-zero-w.dtb",
 		false,
 	},
+	[0xD] = {
+		"3 Model B+",
+		DTB_DIR "bcm2837-rpi-3-b-plus.dtb",
+		true,
+	},
 };
 
 static const struct rpi_model rpi_models_old_scheme[] = {
