diff -Naur dtc-1.7.1.orig/Makefile dtc-1.7.1/Makefile
--- dtc-1.7.1.orig/Makefile	2024-08-17 10:22:16.000000000 +0200
+++ dtc-1.7.1/Makefile	2024-09-03 11:41:19.125603922 +0200
@@ -21,7 +21,7 @@
 WARNINGS = -Wall -Wpointer-arith -Wcast-qual -Wnested-externs -Wsign-compare \
 	-Wstrict-prototypes -Wmissing-prototypes -Wredundant-decls -Wshadow \
 	-Wsuggest-attribute=format -Wwrite-strings
-CFLAGS = -g -Os $(SHAREDLIB_CFLAGS) -Werror $(WARNINGS) $(EXTRA_CFLAGS)
+CFLAGS = -g -Os $(SHAREDLIB_CFLAGS) $(EXTRA_CFLAGS)
 
 BISON = bison
 LEX = flex
