diff -Naur dtc-1.6.1.org/Makefile dtc-1.6.1/Makefile
--- dtc-1.6.1.org/Makefile	2021-06-08 07:00:49.000000000 +0000
+++ dtc-1.6.1/Makefile	2021-10-01 07:57:58.344551140 +0000
@@ -23,7 +23,7 @@
 CPPFLAGS = -I libfdt -I . -DFDT_ASSUME_MASK=$(ASSUME_MASK)
 WARNINGS = -Wall -Wpointer-arith -Wcast-qual -Wnested-externs \
 	-Wstrict-prototypes -Wmissing-prototypes -Wredundant-decls -Wshadow
-CFLAGS = -g -Os $(SHAREDLIB_CFLAGS) -Werror $(WARNINGS) $(EXTRA_CFLAGS)
+CFLAGS = -g -Os $(SHAREDLIB_CFLAGS) $(EXTRA_CFLAGS)
 
 BISON = bison
 LEX = flex
