From 359104632fa556e3c5c78e4016c2585896225716 Mon Sep 17 00:00:00 2001
From: Robert Nelson <robertcnelson@gmail.com>
Date: Mon, 1 Apr 2013 12:17:50 -0500
Subject: [PATCH 4/4] HACK: PandaES: disable cpufreq so board will boot

Signed-off-by: Robert Nelson <robertcnelson@gmail.com>
---
 drivers/cpufreq/omap-cpufreq.c |    7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/cpufreq/omap-cpufreq.c b/drivers/cpufreq/omap-cpufreq.c
index 0279d18..0a8ac09 100644
--- a/drivers/cpufreq/omap-cpufreq.c
+++ b/drivers/cpufreq/omap-cpufreq.c
@@ -31,6 +31,8 @@
 #include <asm/smp_plat.h>
 #include <asm/cpu.h>
 
+#include "../../arch/arm/mach-omap2/soc.h"
+
 /* OPP tolerance in percentage */
 #define	OPP_TOLERANCE	4
 
@@ -246,6 +248,11 @@ static struct cpufreq_driver omap_driver = {
 
 static int omap_cpufreq_probe(struct platform_device *pdev)
 {
+	if (cpu_is_omap446x()) {
+		pr_err("%s: unsupported Silicon?\n", __func__);
+		return -EINVAL;
+	}
+
 	mpu_dev = get_cpu_device(0);
 	if (!mpu_dev) {
 		pr_warning("%s: unable to get the mpu device\n", __func__);
-- 
1.7.10.4

