From f2658275b25ebfe691cdcb9fede85a3088cca168 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Thu, 25 Sep 2014 21:51:25 +0100
Subject: [PATCH 01/55] Add newline at the end of example config file.

---
 dnsmasq.conf.example | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dnsmasq.conf.example b/dnsmasq.conf.example
index 83924fc4a9b4..6b00bd34fbb5 100644
--- a/dnsmasq.conf.example
+++ b/dnsmasq.conf.example
@@ -645,4 +645,4 @@
 #conf-dir=/etc/dnsmasq.d,.bak
 
 # Include all files in a directory which end in .conf
-#conf-dir=/etc/dnsmasq.d/*.conf
\ No newline at end of file
+#conf-dir=/etc/dnsmasq.d/*.conf
-- 
2.1.0

