From bd9520b7ade7098ee423acc38965376aa57feb07 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 16 Dec 2014 20:41:29 +0000
Subject: [PATCH 014/113] Remove redundant IN6_IS_ADDR_ULA(a) macro defn.

---
 src/network.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/src/network.c b/src/network.c
index 5067007c5cbc..99419f57951e 100644
--- a/src/network.c
+++ b/src/network.c
@@ -16,10 +16,6 @@
 
 #include "dnsmasq.h"
 
-#ifndef IN6_IS_ADDR_ULA
-#define IN6_IS_ADDR_ULA(a) ((((__const uint32_t *) (a))[0] & htonl (0xfe00000)) == htonl (0xfc000000))
-#endif
-
 #ifdef HAVE_LINUX_NETWORK
 
 int indextoname(int fd, int index, char *name)
-- 
2.1.0

