From d310ab7ecbffce79d3d90debba621e0222f9bced Mon Sep 17 00:00:00 2001
From: Matthias Andree <matthias.andree@gmx.de>
Date: Sat, 27 Dec 2014 15:36:38 +0000
Subject: [PATCH 23/71] Fix build failure in new inotify code on BSD.

---
 src/inotify.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/inotify.c b/src/inotify.c
index 960bf5efb41f..83730008c11b 100644
--- a/src/inotify.c
+++ b/src/inotify.c
@@ -15,10 +15,10 @@
 */
 
 #include "dnsmasq.h"
-#include <sys/inotify.h>
-
 #ifdef HAVE_LINUX_NETWORK
 
+#include <sys/inotify.h>
+
 /* the strategy is to set a inotify on the directories containing
    resolv files, for any files in the directory which are close-write 
    or moved into the directory.
-- 
2.1.0

