From 3a3965ac21b1b759eab8799b6edb09195b671306 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Sun, 9 Aug 2015 17:45:06 +0100
Subject: [PATCH] Don't answer non-auth queries for auth zones locally when
 --localise-queries set.

---
 src/forward.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/forward.c b/src/forward.c
index 2731b90..b76a974 100644
--- a/src/forward.c
+++ b/src/forward.c
@@ -1365,7 +1365,7 @@ void receive_query(struct listener *listen, time_t now)
 
 #ifdef HAVE_AUTH
       /* find queries for zones we're authoritative for, and answer them directly */
-      if (!auth_dns)
+      if (!auth_dns && !option_bool(OPT_LOCALISE))
 	for (zone = daemon->auth_zones; zone; zone = zone->next)
 	  if (in_zone(zone, daemon->namebuff, NULL))
 	    {
@@ -1904,7 +1904,7 @@ unsigned char *tcp_request(int confd, time_t now,
 	  
 #ifdef HAVE_AUTH
 	  /* find queries for zones we're authoritative for, and answer them directly */
-	  if (!auth_dns)
+	  if (!auth_dns && !option_bool(OPT_LOCALISE))
 	    for (zone = daemon->auth_zones; zone; zone = zone->next)
 	      if (in_zone(zone, daemon->namebuff, NULL))
 		{
-- 
1.7.10.4
