From 8ff70de618eb7de9147dbfbd4deca4a2dd62f0cb Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Sat, 14 Feb 2015 20:02:37 +0000
Subject: [PATCH 49/98] Typos.

---
 src/inotify.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/inotify.c b/src/inotify.c
index 44ce0c9af051..9422066257f5 100644
--- a/src/inotify.c
+++ b/src/inotify.c
@@ -104,7 +104,8 @@ void set_dynamic_inotify(int flag, int total_size, struct crec **rhash, int revh
 	   ah->wd = inotify_add_watch(daemon->inotifyfd, ah->fname, IN_CLOSE_WRITE | IN_MOVED_TO);
 	   ah->flags |= AH_WD_DONE;
 	 }
-       /* Read contents of dir _after_ calling add_watch, in the ho[e of avoiding
+
+       /* Read contents of dir _after_ calling add_watch, in the hope of avoiding
 	  a race which misses files being added as we start */
        if (ah->wd == -1 || !(dir_stream = opendir(ah->fname)))
 	 {
-- 
2.1.0

