From a006eb7e1486023480ea40244720ef7aab51de71 Mon Sep 17 00:00:00 2001
From: Moshe Levi <moshele@mellanox.com>
Date: Sun, 19 Apr 2015 22:10:40 +0100
Subject: [PATCH 079/113] Check IP address command line arg in dhcp_release.c

---
 contrib/wrt/dhcp_release.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/contrib/wrt/dhcp_release.c b/contrib/wrt/dhcp_release.c
index 53f47dda3aec..a51f04b30cab 100644
--- a/contrib/wrt/dhcp_release.c
+++ b/contrib/wrt/dhcp_release.c
@@ -277,6 +277,11 @@ int main(int argc, char **argv)
       exit(1);
     }
   
+  if (inet_addr(argv[2]) == INADDR_NONE)
+    {
+      perror("invalid ip address");
+      exit(1);
+    }
   
   lease.s_addr = inet_addr(argv[2]);
   server = find_interface(lease, nl, if_nametoindex(argv[1]));
-- 
2.1.0

