From 102208df695e886a3086754d32bf7f8c541fbe46 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Thu, 10 Sep 2015 21:50:00 +0100
Subject: [PATCH] DHCPv6 option 56 does not hold an address list. (RFC 5908).

---
 src/dhcp-common.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/dhcp-common.c b/src/dhcp-common.c
index bc48f41..8fc171a 100644
--- a/src/dhcp-common.c
+++ b/src/dhcp-common.c
@@ -599,7 +599,7 @@ static const struct opttab_t opttab6[] = {
   { "sntp-server", 31,  OT_ADDR_LIST },
   { "information-refresh-time", 32, OT_TIME },
   { "FQDN", 39, OT_INTERNAL | OT_RFC1035_NAME },
-  { "ntp-server", 56,  OT_ADDR_LIST },
+  { "ntp-server", 56,  0 },
   { "bootfile-url", 59, OT_NAME },
   { "bootfile-param", 60, OT_CSTRING },
   { NULL, 0, 0 }
-- 
1.7.10.4

