From a5ae1f85873829efe473075ad77806cc02792622 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Sat, 25 Apr 2015 21:46:10 +0100
Subject: [PATCH 83/87] Logs in DHCPv6 not suppressed by dhcp6-quiet.

---
 CHANGELOG     | 6 +++++-
 src/rfc3315.c | 4 ++--
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/CHANGELOG b/CHANGELOG
index 7f2b1e002e9e..af2b22cf8f73 100644
--- a/CHANGELOG
+++ b/CHANGELOG
@@ -103,7 +103,11 @@ version 2.73
 	    Previously we provided correct answers to PTR queries
 	    in such zones (including NS and SOA) but not direct
 	    NS and SOA queries. Thanks to Johnny S. Lee for 
-	    pointing out the problem.
+ 	    pointing out the problem.
+
+	    Fix logging of DHCPREPLY which should be suppressed 
+	    by quiet-dhcp6. Thanks to J. Pablo Abonia for 
+	    spotting the problem.
 
 	
 version 2.72
diff --git a/src/rfc3315.c b/src/rfc3315.c
index c45116a40a09..b4f5dd2db61f 100644
--- a/src/rfc3315.c
+++ b/src/rfc3315.c
@@ -1047,9 +1047,9 @@ static int dhcp6_no_relay(struct state *state, int msg_type, void *inbuff, size_
 		  {
 		    preferred_time = valid_time = 0;
 		    message = _("address invalid");
-		  }
+		  } 
 
-		if (message)
+		if (message && (message != state->hostname))
 		  log6_packet(state, "DHCPREPLY", req_addr, message);	
 		else
 		  log6_quiet(state, "DHCPREPLY", req_addr, message);
-- 
2.1.0

