From 77607cbea0ad0f876dfb79c8b2c121ee400d57d0 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Thu, 10 Sep 2015 23:08:43 +0100
Subject: [PATCH] Respect the --no-resolv flag in inotify code.

---
 CHANGELOG        |    7 ++++++-
 debian/changelog |    6 ++++++
 src/inotify.c    |    3 +++
 3 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/CHANGELOG b/CHANGELOG
index bbc2834..d6e309f 100644
--- a/CHANGELOG
+++ b/CHANGELOG
@@ -7,8 +7,13 @@ version 2.76
 
 	    Enhance --add-subnet to allow arbitrary subnet addresses.
             Thanks to Ed Barsley for the patch.
+
+	    Respect the --no-resolv flag in inotify code. Fixes bug
+	    which caused dnsmasq to fail to start if a resolv-file 
+	    was a dangling symbolic link, even of --no-resolv set.
+	    Thanks to Alexander Kurtz for spotting the problem.
+
 	
-	    
 version 2.75
             Fix reversion on 2.74 which caused 100% CPU use when a 
 	    dhcp-script is configured. Thanks to Adrian Davey for
diff --git a/src/inotify.c b/src/inotify.c
index 52d412f..ef05c58 100644
--- a/src/inotify.c
+++ b/src/inotify.c
@@ -90,6 +90,9 @@ void inotify_dnsmasq_init()
   
   if (daemon->inotifyfd == -1)
     die(_("failed to create inotify: %s"), NULL, EC_MISC);
+
+  if (option_bool(OPT_NO_RESOLV))
+    return;
   
   for (res = daemon->resolv_files; res; res = res->next)
     {
-- 
1.7.10.4

