From 7c0f2543a7e761d1ec82738374556beeb8a35bef Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Thu, 14 May 2015 21:16:18 +0100
Subject: [PATCH 93/98] Tweak last commit.

---
 src/rfc2131.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/rfc2131.c b/src/rfc2131.c
index a10e499ef768..b95f9beadf59 100644
--- a/src/rfc2131.c
+++ b/src/rfc2131.c
@@ -2275,16 +2275,16 @@ static void do_options(struct dhcp_context *context,
 	    t2val = hval;
 	}
        	  
+       /* ensure T1 is still < T2 */
+       if (t2val <= t1val)
+	 t1val = t2val - 1; 
+
        while (fuzz > (t1val/8))
 	 fuzz = fuzz/2;
 	 
        t1val -= fuzz;
        t2val -= fuzz;
        
-       /* ensure T1 is still < T2 */
-       if (t2val <= t1val)
-	 t1val = t2val - 1; 
-
        option_put(mess, end, OPTION_T1, 4, t1val);
        option_put(mess, end, OPTION_T2, 4, t2val);
     }
-- 
2.1.0

