From 7f8565b94ca52dde31f7688a9f9a0cc611d9dae3 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 19 May 2015 23:01:27 +0100
Subject: [PATCH 97/98] Select correct DHCP context when in PXE bootserver
 mode.

---
 src/rfc2131.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/rfc2131.c b/src/rfc2131.c
index 70d1e59530ad..e602a21585c9 100644
--- a/src/rfc2131.c
+++ b/src/rfc2131.c
@@ -805,7 +805,7 @@ size_t dhcp_reply(struct dhcp_context *context, char *iface_name, int int_index,
 	    if (service->type == type)
 	      break;
 	  
-	  if (!service || !service->basename)
+	  if (!service || !service->basename || !(context = narrow_context(context, mess->ciaddr, tagif_netid)))
 	    return 0;
 	  
 	  clear_packet(mess, end);
-- 
2.1.0

