From 27b78d990b7cd901866ad6f1a17b9d633a95fdce Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Sat, 26 Sep 2015 21:40:45 +0100
Subject: [PATCH] Rationalise 5e3e464ac4022ee0b3794513abe510817e2cf3ca

---
 src/rfc3315.c |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/rfc3315.c b/src/rfc3315.c
index 3f1f9ee..3ed8623 100644
--- a/src/rfc3315.c
+++ b/src/rfc3315.c
@@ -1324,8 +1324,7 @@ static struct dhcp_netid *add_options(struct state *state, int do_refresh)
       if (opt_cfg->opt == OPTION6_DNS_SERVER)
 	done_dns = 1;
       
-      /* Empty DNS_SERVER option will not set DHOPT_ADDR6 */
-      if ((opt_cfg->flags & DHOPT_ADDR6) || opt_cfg->opt == OPTION6_DNS_SERVER)
+      if (opt_cfg->flags & DHOPT_ADDR6)
 	{
 	  int len, j;
 	  struct in6_addr *a;
-- 
1.7.10.4

