From 98079ea89851da1df4966dfdfa1852a98da02912 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 13 Oct 2015 20:30:32 +0100
Subject: [PATCH] Catch errors from sendmsg in DHCP code.  Logs, eg,  iptables
 DROPS of dest 255.255.255.255

---
 src/dhcp.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/dhcp.c b/src/dhcp.c
index e6fceb1..1c85e42 100644
--- a/src/dhcp.c
+++ b/src/dhcp.c
@@ -452,8 +452,13 @@ void dhcp_packet(time_t now, int pxe_fd)
 #endif
   
   while(retry_send(sendmsg(fd, &msg, 0)));
+
+  /* This can fail when, eg, iptables DROPS destination 255.255.255.255 */
+  if (errno != 0)
+    my_syslog(MS_DHCP | LOG_WARNING, _("Error sending DHCP packet to %s: %s"),
+	      inet_ntoa(dest.sin_addr), strerror(errno));
 }
- 
+
 /* check against secondary interface addresses */
 static int check_listen_addrs(struct in_addr local, int if_index, char *label,
 			      struct in_addr netmask, struct in_addr broadcast, void *vparam)
-- 
1.7.10.4

