From a86fdf437ecc29398f9715ceb5240442a17ac014 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Sun, 20 Dec 2015 21:19:20 +0000
Subject: [PATCH] Minor tweak to previous commit.

---
 src/dnssec.c |    6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/src/dnssec.c b/src/dnssec.c
index 29848e1..9fa64b6 100644
--- a/src/dnssec.c
+++ b/src/dnssec.c
@@ -1889,8 +1889,6 @@ static int zone_status(char *name, int class, char *keyname, time_t now)
 	}
       else
 	{
-	  int gotone = 0;
-
 	  /* If all the DS records have digest and/or sig algos we don't support,
 	     then the zone is insecure. Note that if an algo
 	     appears in the DS, then RRSIGs for that algo MUST
@@ -1904,11 +1902,11 @@ static int zone_status(char *name, int class, char *keyname, time_t now)
 	      if (crecp->uid == (unsigned int)class &&
 		  hash_find(ds_digest_name(crecp->addr.ds.digest)) &&
 		  verify_func(crecp->addr.ds.algo))
-		gotone = 1;
+		break;
 	    }
 	  while ((crecp = cache_find_by_name(crecp, keyname, now, F_DS)));
 
-	  if (!gotone)
+	  if (!crecp)
 	    return STAT_INSECURE;
 	}
 
-- 
1.7.10.4

