diff -up dhcp-4.3.1b1/client/Makefile.am.brGmwh dhcp-4.3.1b1/client/Makefile.am
--- dhcp-4.3.1b1/client/Makefile.am.brGmwh	2014-07-02 19:58:38.000000000 +0200
+++ dhcp-4.3.1b1/client/Makefile.am	2014-07-10 17:36:30.484250976 +0200
@@ -10,8 +10,8 @@ dhclient_SOURCES = clparse.c dhclient.c
 		   scripts/bsdos scripts/freebsd scripts/linux scripts/macos \
 		   scripts/netbsd scripts/nextstep scripts/openbsd \
 		   scripts/solaris scripts/openwrt
-dhclient_LDADD = ../common/libdhcp.a ../omapip/libomapi.a ../bind/lib/libirs.a \
-		 ../bind/lib/libdns.a ../bind/lib/libisccfg.a ../bind/lib/libisc.a
+dhclient_LDADD = ../common/libdhcp.a ../omapip/libomapi.a \
+		 $(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
 man_MANS = dhclient.8 dhclient-script.8 dhclient.conf.5 dhclient.leases.5
 EXTRA_DIST = $(man_MANS)
 
diff -up dhcp-4.3.1b1/common/tests/Makefile.am.brGmwh dhcp-4.3.1b1/common/tests/Makefile.am
--- dhcp-4.3.1b1/common/tests/Makefile.am.brGmwh	2014-07-10 17:36:30.485250962 +0200
+++ dhcp-4.3.1b1/common/tests/Makefile.am	2014-07-10 17:38:04.010924566 +0200
@@ -13,21 +13,20 @@ ATF_TESTS += alloc_unittest dns_unittest
 alloc_unittest_SOURCES = test_alloc.c $(top_srcdir)/tests/t_api_dhcp.c
 alloc_unittest_LDADD = $(ATF_LDFLAGS)
 alloc_unittest_LDADD += ../libdhcp.a  \
-	../../omapip/libomapi.a ../../bind/lib/libirs.a \
-	../../bind/lib/libdns.a ../../bind/lib/libisccfg.a  ../../bind/lib/libisc.a
+	../../omapip/libomapi.a \
+	$(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
 
 dns_unittest_SOURCES = dns_unittest.c $(top_srcdir)/tests/t_api_dhcp.c
 dns_unittest_LDADD = $(ATF_LDFLAGS)
 dns_unittest_LDADD += ../libdhcp.a  \
-	../../omapip/libomapi.a ../../bind/lib/libirs.a \
-	../../bind/lib/libdns.a ../../bind/lib/libisccfg.a  ../../bind/lib/libisc.a
+	../../omapip/libomapi.a \
+	$(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
 
 misc_unittest_SOURCES = misc_unittest.c $(top_srcdir)/tests/t_api_dhcp.c
 misc_unittest_LDADD = $(ATF_LDFLAGS)
 misc_unittest_LDADD += ../libdhcp.a  \
-	../../omapip/libomapi.a ../../bind/lib/libirs.a \
-	../../bind/lib/libdns.a ../../bind/lib/libisccfg.a  ../../bind/lib/libisc.a
-
+	../../omapip/libomapi.a \
+	$(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
 check: $(ATF_TESTS)
 	atf-run | atf-report
 
diff -up dhcp-4.3.1b1/configure.ac.brGmwh dhcp-4.3.1b1/configure.ac
--- dhcp-4.3.1b1/configure.ac.brGmwh	2014-07-02 20:01:26.000000000 +0200
+++ dhcp-4.3.1b1/configure.ac	2014-07-10 17:36:30.485250962 +0200
@@ -562,20 +562,37 @@ AC_CHECK_MEMBER(struct msghdr.msg_contro
 
 libbind=
 AC_ARG_WITH(libbind,
-	AS_HELP_STRING([--with-libbind=PATH],[bind includes and libraries are in PATH 
-		        (default is ./bind)]),
+	AS_HELP_STRING([--with-libbind=PATH],[bind includes are in PATH 
+		        (default is ./bind/includes)]),
 	use_libbind="$withval", use_libbind="no")
 case "$use_libbind" in 
+yes|no)
+	libbind="\${top_srcdir}/bind/include"
+	;;
+*)
+	libbind="$use_libbind"
+	;;
+esac
+
+BIND9_LIBDIR='-L$(top_builddir)/bind/lib'
+AC_ARG_WITH(libbind-libs,
+	AC_HELP_STRING([--with-libbind-libs=PATH],
+		       [bind9 export libraries are in PATH]),
+		       [libbind_libs="$withval"], [libbind_libs='no'])
+case "$libbind_libs" in
 yes)
-	libbind="\${top_srcdir}/bind"
+	AC_MSG_ERROR([Specify path to bind9 libraries])
 	;;
 no)
-	libbind="\${top_srcdir}/bind"
+	BUNDLED_BIND=yes
 	;;
 *)
-	libbind="$use_libbind"
+	BIND9_LIBDIR="-L$libbind_libs"
+	BUNDLED_BIND=no
 	;;
 esac
+AM_CONDITIONAL([BUNDLED_BIND], [test "$BUNDLED_BIND" = yes])
+AC_SUBST([BIND9_LIBDIR])
 
 # OpenLDAP support.
 AC_ARG_WITH(ldap,
@@ -610,7 +627,7 @@ fi
 CFLAGS="$CFLAGS $STD_CWARNINGS"
 
 # Try to add the bind include directory
-CFLAGS="$CFLAGS -I$libbind/include"
+CFLAGS="$CFLAGS -I$libbind"
 
 case "$host" in
 *-darwin*)
diff -up dhcp-4.3.1b1/dhcpctl/Makefile.am.brGmwh dhcp-4.3.1b1/dhcpctl/Makefile.am
--- dhcp-4.3.1b1/dhcpctl/Makefile.am.brGmwh	2014-07-02 19:58:38.000000000 +0200
+++ dhcp-4.3.1b1/dhcpctl/Makefile.am	2014-07-10 17:36:30.485250962 +0200
@@ -6,12 +6,9 @@ EXTRA_DIST = $(man_MANS)
 
 omshell_SOURCES = omshell.c
 omshell_LDADD = libdhcpctl.a ../common/libdhcp.a ../omapip/libomapi.a \
-	        ../bind/lib/libirs.a ../bind/lib/libdns.a \
-	        ../bind/lib/libisccfg.a ../bind/lib/libisc.a
-
+	        $(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
 libdhcpctl_a_SOURCES = dhcpctl.c callback.c remote.c
 
 cltest_SOURCES = cltest.c
 cltest_LDADD = libdhcpctl.a ../common/libdhcp.a ../omapip/libomapi.a \
-	       ../bind/lib/libirs.a ../bind/lib/libdns.a \
-               ../bind/lib/libisccfg.a ../bind/lib/libisc.a
+	       $(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
diff -up dhcp-4.3.1b1/Makefile.am.brGmwh dhcp-4.3.1b1/Makefile.am
--- dhcp-4.3.1b1/Makefile.am.brGmwh	2014-07-02 19:58:38.000000000 +0200
+++ dhcp-4.3.1b1/Makefile.am	2014-07-10 17:36:30.484250976 +0200
@@ -25,7 +25,13 @@ EXTRA_DIST = RELNOTES LICENSE \
 	     bind/Makefile bind/bind.tar.gz bind/version.tmp \
 	     common/tests/Atffile server/tests/Atffile
 
-SUBDIRS = bind includes tests common dst omapip client dhcpctl relay server
+if BUNDLED_BIND
+SUBDIRS = bind
+else
+SUBDIRS = 
+endif
+
+SUBDIRS += includes tests common dst omapip client dhcpctl relay server
 
 nobase_include_HEADERS = dhcpctl/dhcpctl.h
 
diff -up dhcp-4.3.1b1/omapip/Makefile.am.brGmwh dhcp-4.3.1b1/omapip/Makefile.am
--- dhcp-4.3.1b1/omapip/Makefile.am.brGmwh	2014-07-02 19:58:39.000000000 +0200
+++ dhcp-4.3.1b1/omapip/Makefile.am	2014-07-10 17:36:30.486250948 +0200
@@ -10,6 +10,5 @@ man_MANS = omapi.3
 EXTRA_DIST = $(man_MANS)
 
 svtest_SOURCES = test.c
-svtest_LDADD = libomapi.a ../bind/lib/libirs.a ../bind/lib/libdns.a \
-		../bind/lib/libisccfg.a ../bind/lib/libisc.a
-
+svtest_LDADD = libomapi.a \
+	       $(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
diff -up dhcp-4.3.1b1/relay/Makefile.am.brGmwh dhcp-4.3.1b1/relay/Makefile.am
--- dhcp-4.3.1b1/relay/Makefile.am.brGmwh	2014-07-02 19:58:39.000000000 +0200
+++ dhcp-4.3.1b1/relay/Makefile.am	2014-07-10 17:36:30.486250948 +0200
@@ -3,8 +3,7 @@ AM_CPPFLAGS = -DLOCALSTATEDIR='"@localst
 sbin_PROGRAMS = dhcrelay
 dhcrelay_SOURCES = dhcrelay.c
 dhcrelay_LDADD = ../common/libdhcp.a ../omapip/libomapi.a \
-		 ../bind/lib/libirs.a ../bind/lib/libdns.a \
-		 ../bind/lib/libisccfg.a ../bind/lib/libisc.a
+		 $(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
 man_MANS = dhcrelay.8
 EXTRA_DIST = $(man_MANS)
 
diff -up dhcp-4.3.1b1/server/Makefile.am.brGmwh dhcp-4.3.1b1/server/Makefile.am
--- dhcp-4.3.1b1/server/Makefile.am.brGmwh	2014-07-02 19:58:39.000000000 +0200
+++ dhcp-4.3.1b1/server/Makefile.am	2014-07-10 17:36:30.486250948 +0200
@@ -13,10 +13,8 @@ dhcpd_SOURCES = dhcpd.c dhcp.c bootp.c c
 		dhcpv6.c mdb6.c ldap.c ldap_casa.c
 
 dhcpd_CFLAGS = $(LDAP_CFLAGS)
-dhcpd_LDADD = ../common/libdhcp.a ../omapip/libomapi.a \
-	      ../dhcpctl/libdhcpctl.a ../bind/lib/libirs.a \
-	      ../bind/lib/libdns.a ../bind/lib/libisccfg.a ../bind/lib/libisc.a
-
+dhcpd_LDADD = ../common/libdhcp.a ../omapip/libomapi.a ../dhcpctl/libdhcpctl.a \
+	      $(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
 man_MANS = dhcpd.8 dhcpd.conf.5 dhcpd.leases.5
 EXTRA_DIST = $(man_MANS)
 
diff -up dhcp-4.3.1b1/server/tests/Makefile.am.brGmwh dhcp-4.3.1b1/server/tests/Makefile.am
--- dhcp-4.3.1b1/server/tests/Makefile.am.brGmwh	2014-07-02 19:58:40.000000000 +0200
+++ dhcp-4.3.1b1/server/tests/Makefile.am	2014-07-10 17:36:30.486250948 +0200
@@ -18,9 +18,8 @@ DHCPSRC = ../dhcp.c ../bootp.c ../confpa
           ../ldap.c ../ldap_casa.c ../dhcpd.c
 
 DHCPLIBS = $(top_builddir)/common/libdhcp.a $(top_builddir)/omapip/libomapi.a    \
-          $(top_builddir)/dhcpctl/libdhcpctl.a $(top_builddir)/bind/lib/libirs.a \
-	  $(top_builddir)/bind/lib/libdns.a $(top_builddir)/bind/lib/libisccfg.a \
-	  $(top_builddir)/bind/lib/libisc.a
+          $(top_builddir)/dhcpctl/libdhcpctl.a \
+          $(BIND9_LIBDIR) -lirs -ldns -lisccfg -lisc
 
 ATF_TESTS =
 TESTS = 
