diff -up dhcp-4.3.1b1/client/dhclient.c.dlTsyN dhcp-4.3.1b1/client/dhclient.c
--- dhcp-4.3.1b1/client/dhclient.c.dlTsyN	2014-07-10 17:49:49.882925843 +0200
+++ dhcp-4.3.1b1/client/dhclient.c	2014-07-10 17:50:26.922402550 +0200
@@ -1748,11 +1748,6 @@ int write_host (host)
 	return 0;
 }
 
-void db_startup (testp)
-	int testp;
-{
-}
-
 void bootp (packet)
 	struct packet *packet;
 {
diff -up dhcp-4.3.1b1/includes/dhcpd.h.dlTsyN dhcp-4.3.1b1/includes/dhcpd.h
--- dhcp-4.3.1b1/includes/dhcpd.h.dlTsyN	2014-07-10 17:48:03.798424601 +0200
+++ dhcp-4.3.1b1/includes/dhcpd.h	2014-07-10 17:50:26.923402536 +0200
@@ -2866,7 +2866,11 @@ void commit_leases_timeout (void *);
 void commit_leases_readerdry(void *);
 int commit_leases (void);
 int commit_leases_timed (void);
+#if defined (PARANOIA)
+void db_startup (int, uid_t, gid_t);
+#else
 void db_startup (int);
+#endif /* PARANOIA */
 int new_lease_file (void);
 int group_writer (struct group_object *);
 int write_ia(const struct ia_xx *);
diff -up dhcp-4.3.1b1/server/confpars.c.dlTsyN dhcp-4.3.1b1/server/confpars.c
--- dhcp-4.3.1b1/server/confpars.c.dlTsyN	2014-07-10 17:39:25.801764596 +0200
+++ dhcp-4.3.1b1/server/confpars.c	2014-07-10 17:50:26.924402522 +0200
@@ -219,7 +219,11 @@ void trace_conf_input (trace_type_t *tty
 	}
 
 	if (!leaseconf_initialized && ttype == trace_readleases_type) {
+#if defined (PARANOIA)
+		db_startup (0, 0, 0);
+#else
 		db_startup (0);
+#endif /* PARANOIA */
 		leaseconf_initialized = 1;
 		postdb_startup ();
 	}
diff -up dhcp-4.3.1b1/server/db.c.dlTsyN dhcp-4.3.1b1/server/db.c
--- dhcp-4.3.1b1/server/db.c.dlTsyN	2014-07-10 17:39:25.801764596 +0200
+++ dhcp-4.3.1b1/server/db.c	2014-07-10 17:50:26.925402508 +0200
@@ -42,6 +42,10 @@ static int counting = 0;
 static int count = 0;
 TIME write_time;
 int lease_file_is_corrupt = 0;
+#if defined (PARANOIA)
+uid_t global_set_uid = 0;
+gid_t global_set_gid = 0;
+#endif /* PARANOIA */
 
 /* Write a single binding scope value in parsable format.
  */
@@ -1046,8 +1050,11 @@ int commit_leases_timed()
 	return (1);
 }
 
-void db_startup (testp)
-	int testp;
+#if defined (PARANOIA)
+void db_startup (int testp, uid_t set_uid, gid_t set_gid)
+#else
+void db_startup (int testp)
+#endif /* PARANOIA */
 {
 	isc_result_t status;
 
@@ -1066,6 +1073,11 @@ void db_startup (testp)
 	}
 #endif
 
+#if defined (PARANOIA)
+	global_set_uid = set_uid;
+	global_set_gid = set_gid;
+#endif /* PARANOIA */
+
 #if defined (TRACING)
 	/* If we're playing back, there is no lease file, so we can't
 	   append it, so we create one immediately (maybe this isn't
@@ -1128,6 +1140,17 @@ int new_lease_file ()
 		log_error ("Can't create new lease file: %m");
 		return 0;
 	}
+
+#if defined (PARANOIA)
+	if (global_set_uid && !geteuid() &&
+	    global_set_gid && !getegid())
+		if (fchown(db_fd, global_set_uid, global_set_gid)) {
+			log_fatal ("Can't chown new lease file: %m");
+			close(db_fd);
+			goto fdfail;
+	}
+#endif /* PARANOIA */
+
 	if ((new_db_file = fdopen(db_fd, "we")) == NULL) {
 		log_error("Can't fdopen new lease file: %m");
 		close(db_fd);
diff -up dhcp-4.3.1b1/server/dhcpd.8.dlTsyN dhcp-4.3.1b1/server/dhcpd.8
--- dhcp-4.3.1b1/server/dhcpd.8.dlTsyN	2014-07-02 19:58:39.000000000 +0200
+++ dhcp-4.3.1b1/server/dhcpd.8	2014-07-10 17:50:26.925402508 +0200
@@ -82,6 +82,18 @@ dhcpd - Dynamic Host Configuration Proto
 .I trace-output-file
 ]
 [
+.B -user
+.I user
+]
+[
+.B -group
+.I group
+]
+[
+.B -chroot
+.I dir
+]
+[
 .B -play
 .I trace-playback-file
 ]
@@ -269,6 +281,15 @@ lease file.
 .TP
 .BI --version
 Print version number and exit.
+.TP
+.BI \-user \ user
+Setuid to user after completing privileged operations, such as creating sockets that listen on privileged ports.
+.TP
+.BI \-group \ group
+Setgid to group after completing privileged operations, such as creating sockets that listen on privileged ports.
+.TP
+.BI \-chroot \ dir
+Chroot to directory after processing the command line arguments, but before reading the configuration file.
 .PP
 .I Modifying default file locations:
 The following options can be used to modify the locations 
diff -up dhcp-4.3.1b1/server/dhcpd.c.dlTsyN dhcp-4.3.1b1/server/dhcpd.c
--- dhcp-4.3.1b1/server/dhcpd.c.dlTsyN	2014-07-10 17:39:25.802764582 +0200
+++ dhcp-4.3.1b1/server/dhcpd.c	2014-07-10 17:52:35.341588248 +0200
@@ -628,7 +628,11 @@ main(int argc, char **argv) {
 	group_write_hook = group_writer;
 
 	/* Start up the database... */
+#if defined (PARANOIA)
+	db_startup (lftest, set_uid, set_gid);
+#else
 	db_startup (lftest);
+#endif /* PARANOIA */
 
 	if (lftest)
 		exit (0);
