diff -U 3 a/scripts/services/clamav b/scripts/services/clamav
--- a/scripts/services/clamav	Fri Sep 21 00:46:14 2020
+++ b/scripts/services/clamav	Mon Dec 25 09:26:26 2020
@@ -77,6 +77,7 @@
        ( $ThisLine =~ /Bytecode: Security mode set to /) or
        ( $ThisLine =~ /^No stats for Database check/ ) or
        ( $ThisLine =~ /^Received \d+ file descriptor\(s\) from systemd\.$/) or
+       ( $ThisLine =~ /^Activating the newly loaded database.../ ) or
        0 # This line prevents blame shifting as lines are added above
        ) {
        # We do not care about these.
