diff -U 3 a/conf/services/mdadm.conf b/conf/services/mdadm.conf
--- a/conf/services/mdadm.conf	Sat Jan 22 01:00:00 2022
+++ b/conf/services/mdadm.conf	Sun Apr 10 10:48:21 2022
@@ -13,7 +13,7 @@
 # Logwatch will try to find md devices in /etc/mdadm.conf or 
 # /etc/mdadm/mdadm.conf. If none of these files exist it can scan actively
 # for md devices. Set to 'Yes' to enable active scanning:
-$mdadm_enable_scan = No
+$mdadm_enable_scan = Yes
 
 # Logwatch will emit an error for md devices listed in /etc/mdadm.conf
 # that are not present.  If you do not want this (e.g. raid devices may come
diff -U 3 a/scripts/services/mdadm b/scripts/services/mdadm
--- a/scripts/services/mdadm	Sat Jan 22 01:00:00 2022
+++ b/scripts/services/mdadm	Sun Apr 10 10:38:19 2022
@@ -36,7 +36,7 @@
 if (
     open($mdadm, "<", "/etc/mdadm.conf") or
     open($mdadm, "<", "/etc/mdadm/mdadm.conf") or
-    open($mdadm, "<", "mdadm --detail --scan 2>/dev/null|")) {
+    open($mdadm, "<", "sudo mdadm --detail --scan 2>/dev/null|")) {
    while (<$mdadm>) {
        if (/^ARRAY/) {
          push(@devices,(split())[1]);
@@ -52,7 +52,7 @@
     next;
   }
 
-  open(MDADM,"mdadm --misc --detail $dev 2>&1 |");
+  open(MDADM,"sudo mdadm --misc --detail $dev 2>&1 |");
   while (<MDADM>) {
     if ($_ =~ /cannot open .*: No such file or directory/) {
       print $_ unless $ignore_missing;
@@ -75,7 +75,11 @@
 
   if ($Detail <= 4) {
     if (lc($mdhash{'state'}) =~ /clean|active/) {
-      print "$dev : $mdhash{'state'}\n" if $Detail;
+      if (lc($mdhash{'state'}) =~ /degraded/) {
+        print "$dev : $mdhash{'state'}\n";
+      } else {
+        print "$dev : $mdhash{'state'}\n" if $Detail;
+      }
     } else {
       print "$dev : $mdhash{'state'}\n";
       if (defined($mdhash{'middle devices'})) {
