diff -Naur logwatch-7.5.5-orig/conf/services/mdadm.conf logwatch-7.5.5/conf/services/mdadm.conf
--- logwatch-7.5.5-orig/conf/services/mdadm.conf	2021-01-22 21:59:40.000000000 +0100
+++ logwatch-7.5.5/conf/services/mdadm.conf	2021-10-04 13:52:30.850057355 +0200
@@ -13,7 +13,7 @@
 # Logwatch will try to find md devices in /etc/mdadm.conf or 
 # /etc/mdadm/mdadm.conf. If none of these files exist it can scan actively
 # for md devices. Set to 'Yes' to enable active scanning:
-$mdadm_enable_scan = No
+$mdadm_enable_scan = Yes
 
 # Logwatch will emit an error for md devices listed in /etc/mdadm.conf
 # that are not present.  If you do not want this (e.g. raid devices may come
diff -Naur logwatch-7.5.5-orig/scripts/services/mdadm logwatch-7.5.5/scripts/services/mdadm
--- logwatch-7.5.5-orig/scripts/services/mdadm	2021-01-22 21:59:40.000000000 +0100
+++ logwatch-7.5.5/scripts/services/mdadm	2021-10-06 11:41:14.800307603 +0200
@@ -35,7 +35,7 @@
 } elsif ( -f "/etc/mdadm/mdadm.conf" ) {
   open(MDADM,"< /etc/mdadm/mdadm.conf");
 } elsif ($enable_scan) {
-  open(MDADM,"mdadm --detail --scan 2>/dev/null|");
+  open(MDADM,"sudo mdadm --detail --scan 2>/dev/null|");
 }
 while (<MDADM>) {
   if (/^ARRAY/) {
@@ -51,7 +51,7 @@
     next;
   }
 
-  open(MDADM,"mdadm --misc --detail $dev 2>&1 |");
+  open(MDADM,"sudo mdadm --misc --detail $dev 2>&1 |");
   while (<MDADM>) {
     if ($_ =~ /cannot open .*: No such file or directory/) {
       print $_ unless $ignore_missing;
@@ -74,7 +74,11 @@
 
   if ($Detail <= 4) {
     if (lc($mdhash{'state'}) =~ /clean|active/) {
-      print "$dev : $mdhash{'state'}\n" if $Detail;
+      if (lc($mdhash{'state'}) =~ /degraded/) {
+        print "$dev : $mdhash{'state'}\n";
+      } else {
+        print "$dev : $mdhash{'state'}\n" if $Detail;
+      }
     } else {
       print "$dev : $mdhash{'state'}\n";
       if (defined($mdhash{'middle devices'})) {
