From: olh@suse.de
Subject: force speed to fix autodetection on pegasos2
Patch-mainline: never

---
 arch/powerpc/platforms/chrp/setup.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/powerpc/platforms/chrp/setup.c
+++ b/arch/powerpc/platforms/chrp/setup.c
@@ -295,7 +295,7 @@ static void chrp_init_early(void)
 	if (!property)
 		goto out_put;
 	if (!strcmp(property, "failsafe") || !strcmp(property, "serial"))
-		add_preferred_console("ttyS", 0, NULL);
+		add_preferred_console("ttyS", 0, "115200");
 out_put:
 	of_node_put(node);
 }
